// iwer/lib/depth/XRDepthInformation.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { P_DEPTH_INFO } from '../private.js';
export class XRCPUDepthInformation {
    constructor(data, width, height, normDepthBufferFromNormView, rawValueToMeters, dataFormat) {
        this[P_DEPTH_INFO] = {
            data,
            width,
            height,
            normDepthBufferFromNormView,
            rawValueToMeters,
            dataFormat,
            view: dataFormat === 'float32'
                ? new Float32Array(data)
                : new Uint8Array(data),
        };
    }
    get data() {
        return this[P_DEPTH_INFO].data;
    }
    get width() {
        return this[P_DEPTH_INFO].width;
    }
    get height() {
        return this[P_DEPTH_INFO].height;
    }
    get normDepthBufferFromNormView() {
        return this[P_DEPTH_INFO].normDepthBufferFromNormView;
    }
    get rawValueToMeters() {
        return this[P_DEPTH_INFO].rawValueToMeters;
    }
    get dataFormat() {
        return this[P_DEPTH_INFO].dataFormat;
    }
    getDepthInMeters(x, y) {
        const { width, height, rawValueToMeters, dataFormat, view } = this[P_DEPTH_INFO];
        if (x < 0 || x >= 1 || y < 0 || y >= 1) {
            throw new RangeError('Normalized coordinates must be in [0, 1) range.');
        }
        const col = Math.floor(x * width);
        const row = Math.floor(y * height);
        if (dataFormat === 'float32') {
            const floatView = view;
            const index = row * width + col;
            return floatView[index] * rawValueToMeters;
        }
        else {
            // luminance-alpha: 16-bit unsigned int packed as two bytes
            const byteView = view;
            const index = (row * width + col) * 2;
            const rawValue = byteView[index] + byteView[index + 1] * 256;
            return rawValue * rawValueToMeters;
        }
    }
}
