// iwer/lib/depth/XRWebGLBinding.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { P_FRAME, P_SESSION, P_VIEW } from '../private.js';
export class XRWebGLDepthInformation {
    constructor(texture, width, height, normDepthBufferFromNormView, rawValueToMeters, depthNear, depthFar, textureType, imageIndex) {
        this.texture = texture;
        this.width = width;
        this.height = height;
        this.normDepthBufferFromNormView = normDepthBufferFromNormView;
        this.rawValueToMeters = rawValueToMeters;
        this.depthNear = depthNear;
        this.depthFar = depthFar;
        this.textureType = textureType;
        this.imageIndex = imageIndex;
        this.isValid = true;
    }
}
export class XRWebGLBinding {
    constructor(session, _context) {
        this._session = session;
    }
    getDepthInformation(view) {
        var _a;
        const frame = this._session[P_SESSION].activeFrame;
        if (!frame || !frame[P_FRAME].active) {
            throw new DOMException('XRFrame access outside the callback that produced it is invalid.', 'InvalidStateError');
        }
        if (!this._session[P_SESSION].enabledFeatures.includes('depth-sensing')) {
            throw new DOMException('depth-sensing feature is not enabled on this session.', 'InvalidStateError');
        }
        if (this._session[P_SESSION].depthSensingUsage !== 'gpu-optimized') {
            throw new DOMException('getDepthInformation on XRWebGLBinding requires gpu-optimized depth sensing usage.', 'InvalidStateError');
        }
        const eye = view[P_VIEW].eye;
        return (_a = frame[P_FRAME].gpuDepthDataMap.get(eye)) !== null && _a !== void 0 ? _a : null;
    }
}
