// @iwsdk/vite-plugin-dev/dist/index.js
import { randomUUID } from 'crypto';
import * as fs from 'fs';
import { realpathSync, existsSync, readFileSync } from 'fs';
import * as path from 'path';
import path__default from 'path';
import { WebSocketServer, WebSocket } from 'ws';
import { spawn } from 'child_process';
import * as os from 'os';
import { chromium } from 'playwright';
import sharp from 'sharp';
import { readFile, rm, mkdir, writeFile, rename } from 'fs/promises';
import { fileURLToPath } from 'url';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const IWSDK_RUNTIME_SESSION_PATH = '.iwsdk/runtime/session.json';
const IWSDK_RUNTIME_STATE_SCHEMA_VERSION = 2;
const INTERNAL_BROWSER_PROBE_METHOD = '__iwsdk_browser_probe';
const RUNTIME_MCP_TOOLS = [
    // =============================================================================
    // Session Management
    // =============================================================================
    {
        name: 'xr_get_session_status',
        description: 'Get XR session and device status',
        inputSchema: {
            type: 'object',
            properties: {},
            required: [],
        },
    },
    {
        name: 'xr_accept_session',
        description: 'Accept an offered XR session (equivalent to clicking "Enter XR" button)',
        inputSchema: {
            type: 'object',
            properties: {},
            required: [],
        },
    },
    {
        name: 'xr_end_session',
        description: 'End the current active XR session',
        inputSchema: {
            type: 'object',
            properties: {},
            required: [],
        },
    },
    // =============================================================================
    // Transform Control
    // =============================================================================
    {
        name: 'xr_get_transform',
        description: 'Get position and orientation of a tracked device',
        inputSchema: {
            type: 'object',
            properties: {
                device: {
                    type: 'string',
                    enum: [
                        'headset',
                        'controller-left',
                        'controller-right',
                        'hand-left',
                        'hand-right',
                    ],
                    description: 'The device to query',
                },
            },
            required: ['device'],
        },
    },
    {
        name: 'xr_set_transform',
        description: 'Set position and/or orientation of a tracked device. Position is in meters, orientation can be quaternion or euler angles (degrees).',
        inputSchema: {
            type: 'object',
            properties: {
                device: {
                    type: 'string',
                    enum: [
                        'headset',
                        'controller-left',
                        'controller-right',
                        'hand-left',
                        'hand-right',
                    ],
                    description: 'The device to move',
                },
                position: {
                    type: 'object',
                    description: 'World position in meters',
                    properties: {
                        x: { type: 'number', description: 'X position (left/right)' },
                        y: {
                            type: 'number',
                            description: 'Y position (up/down, 1.6 is standing height)',
                        },
                        z: {
                            type: 'number',
                            description: 'Z position (forward/back, negative is forward)',
                        },
                    },
                },
                orientation: {
                    type: 'object',
                    description: 'Rotation as quaternion {x,y,z,w} or euler angles {pitch,yaw,roll} in degrees',
                    properties: {
                        x: { type: 'number', description: 'Quaternion X component' },
                        y: { type: 'number', description: 'Quaternion Y component' },
                        z: { type: 'number', description: 'Quaternion Z component' },
                        w: { type: 'number', description: 'Quaternion W component' },
                        pitch: {
                            type: 'number',
                            description: 'Pitch in degrees (X rotation)',
                        },
                        yaw: { type: 'number', description: 'Yaw in degrees (Y rotation)' },
                        roll: {
                            type: 'number',
                            description: 'Roll in degrees (Z rotation)',
                        },
                    },
                },
            },
            required: ['device'],
        },
    },
    {
        name: 'xr_look_at',
        description: 'Orient a device to look at a specific world position',
        inputSchema: {
            type: 'object',
            properties: {
                device: {
                    type: 'string',
                    enum: [
                        'headset',
                        'controller-left',
                        'controller-right',
                        'hand-left',
                        'hand-right',
                    ],
                    description: 'The device to orient',
                },
                target: {
                    type: 'object',
                    description: 'World position to look at',
                    properties: {
                        x: { type: 'number' },
                        y: { type: 'number' },
                        z: { type: 'number' },
                    },
                    required: ['x', 'y', 'z'],
                },
                moveToDistance: {
                    type: 'number',
                    description: 'Optional: move device to this distance from target',
                },
            },
            required: ['device', 'target'],
        },
    },
    {
        name: 'xr_animate_to',
        description: 'Smoothly animate a device to a new position/orientation over time',
        inputSchema: {
            type: 'object',
            properties: {
                device: {
                    type: 'string',
                    enum: [
                        'headset',
                        'controller-left',
                        'controller-right',
                        'hand-left',
                        'hand-right',
                    ],
                    description: 'The device to animate',
                },
                position: {
                    type: 'object',
                    description: 'Target world position in meters',
                    properties: {
                        x: { type: 'number' },
                        y: { type: 'number' },
                        z: { type: 'number' },
                    },
                },
                orientation: {
                    type: 'object',
                    description: 'Target rotation as quaternion {x,y,z,w} or euler angles {pitch,yaw,roll} in degrees',
                    properties: {
                        x: { type: 'number', description: 'Quaternion X component' },
                        y: { type: 'number', description: 'Quaternion Y component' },
                        z: { type: 'number', description: 'Quaternion Z component' },
                        w: { type: 'number', description: 'Quaternion W component' },
                        pitch: {
                            type: 'number',
                            description: 'Pitch in degrees (X rotation)',
                        },
                        yaw: { type: 'number', description: 'Yaw in degrees (Y rotation)' },
                        roll: {
                            type: 'number',
                            description: 'Roll in degrees (Z rotation)',
                        },
                    },
                },
                duration: {
                    type: 'number',
                    description: 'Animation duration in seconds (default: 0.5)',
                },
            },
            required: ['device'],
        },
    },
    // =============================================================================
    // Input Mode
    // =============================================================================
    {
        name: 'xr_set_input_mode',
        description: 'Switch between controller and hand tracking input modes',
        inputSchema: {
            type: 'object',
            properties: {
                mode: {
                    type: 'string',
                    enum: ['controller', 'hand'],
                    description: 'Input mode to switch to',
                },
            },
            required: ['mode'],
        },
    },
    {
        name: 'xr_set_connected',
        description: 'Connect or disconnect an input device',
        inputSchema: {
            type: 'object',
            properties: {
                device: {
                    type: 'string',
                    enum: [
                        'controller-left',
                        'controller-right',
                        'hand-left',
                        'hand-right',
                    ],
                    description: 'The input device',
                },
                connected: {
                    type: 'boolean',
                    description: 'Whether the device should be connected',
                },
            },
            required: ['device', 'connected'],
        },
    },
    // =============================================================================
    // Select/Trigger Input
    // =============================================================================
    {
        name: 'xr_get_select_value',
        description: 'Get the current select (trigger/pinch) value for an input device',
        inputSchema: {
            type: 'object',
            properties: {
                device: {
                    type: 'string',
                    enum: [
                        'controller-left',
                        'controller-right',
                        'hand-left',
                        'hand-right',
                    ],
                    description: 'The input device',
                },
            },
            required: ['device'],
        },
    },
    {
        name: 'xr_set_select_value',
        description: 'Set the select (trigger/pinch) value for an input device. Use for grab-move-release patterns.',
        inputSchema: {
            type: 'object',
            properties: {
                device: {
                    type: 'string',
                    enum: [
                        'controller-left',
                        'controller-right',
                        'hand-left',
                        'hand-right',
                    ],
                    description: 'The input device',
                },
                value: {
                    type: 'number',
                    minimum: 0,
                    maximum: 1,
                    description: 'Select value (0=released, 1=fully pressed/pinched)',
                },
            },
            required: ['device', 'value'],
        },
    },
    {
        name: 'xr_select',
        description: 'Perform a complete select action (press and release). Dispatches selectstart, select, selectend events.',
        inputSchema: {
            type: 'object',
            properties: {
                device: {
                    type: 'string',
                    enum: [
                        'controller-left',
                        'controller-right',
                        'hand-left',
                        'hand-right',
                    ],
                    description: 'The input device',
                },
                duration: {
                    type: 'number',
                    description: 'How long to hold in seconds (default: 0.15)',
                },
            },
            required: ['device'],
        },
    },
    // =============================================================================
    // Gamepad State (Controllers only)
    // =============================================================================
    {
        name: 'xr_get_gamepad_state',
        description: 'Get full gamepad state including all buttons and axes. Button indices in the result: 0=trigger, 1=squeeze, 2=thumbstick, 3=A/X, 4=B/Y, 5=thumbrest.',
        inputSchema: {
            type: 'object',
            properties: {
                device: {
                    type: 'string',
                    enum: ['controller-left', 'controller-right'],
                    description: 'The controller',
                },
            },
            required: ['device'],
        },
    },
    {
        name: 'xr_set_gamepad_state',
        description: 'Set gamepad button and axis values by index',
        inputSchema: {
            type: 'object',
            properties: {
                device: {
                    type: 'string',
                    enum: ['controller-left', 'controller-right'],
                    description: 'The controller',
                },
                buttons: {
                    type: 'array',
                    description: 'Button states to set',
                    items: {
                        type: 'object',
                        properties: {
                            index: {
                                type: 'number',
                                description: 'Button index (0=trigger, 1=squeeze, 2=thumbstick, 3=A/X, 4=B/Y, 5=thumbrest)',
                            },
                            value: { type: 'number', description: 'Button value 0-1' },
                            touched: {
                                type: 'boolean',
                                description: 'Whether button is touched',
                            },
                        },
                        required: ['index', 'value'],
                    },
                },
                axes: {
                    type: 'array',
                    description: 'Axis values to set',
                    items: {
                        type: 'object',
                        properties: {
                            index: {
                                type: 'number',
                                description: 'Axis index (0=thumbstick X, 1=thumbstick Y)',
                            },
                            value: { type: 'number', description: 'Axis value -1 to 1' },
                        },
                        required: ['index', 'value'],
                    },
                },
            },
            required: ['device'],
        },
    },
    // =============================================================================
    // Screenshot
    // =============================================================================
    {
        name: 'browser_screenshot',
        description: 'Capture a screenshot of the browser. Returns the image directly.',
        inputSchema: {
            type: 'object',
            properties: {},
        },
    },
    // =============================================================================
    // Device State
    // =============================================================================
    {
        name: 'xr_get_device_state',
        description: 'Get comprehensive state of the XR device including headset, controllers, and hands',
        inputSchema: {
            type: 'object',
            properties: {},
            required: [],
        },
    },
    {
        name: 'xr_set_device_state',
        description: 'Set device state. When called with no state, resets everything to defaults.',
        inputSchema: {
            type: 'object',
            properties: {
                state: {
                    type: 'object',
                    description: 'Partial device state to apply. Omit to reset to defaults.',
                    properties: {
                        headset: {
                            type: 'object',
                            description: 'Headset transform',
                            properties: {
                                position: {
                                    type: 'object',
                                    properties: {
                                        x: { type: 'number' },
                                        y: { type: 'number' },
                                        z: { type: 'number' },
                                    },
                                },
                                orientation: {
                                    type: 'object',
                                    properties: {
                                        x: { type: 'number' },
                                        y: { type: 'number' },
                                        z: { type: 'number' },
                                        w: { type: 'number' },
                                    },
                                },
                            },
                        },
                        inputMode: {
                            type: 'string',
                            enum: ['controller', 'hand'],
                            description: 'Input mode',
                        },
                        stereoEnabled: {
                            type: 'boolean',
                            description: 'Whether stereo rendering is enabled',
                        },
                        fov: {
                            type: 'number',
                            description: 'Field of view in degrees',
                        },
                        controllers: {
                            type: 'object',
                            description: 'Controller states',
                            properties: {
                                left: {
                                    type: 'object',
                                    properties: {
                                        position: { type: 'object' },
                                        orientation: { type: 'object' },
                                        connected: { type: 'boolean' },
                                    },
                                },
                                right: {
                                    type: 'object',
                                    properties: {
                                        position: { type: 'object' },
                                        orientation: { type: 'object' },
                                        connected: { type: 'boolean' },
                                    },
                                },
                            },
                        },
                        hands: {
                            type: 'object',
                            description: 'Hand states',
                            properties: {
                                left: {
                                    type: 'object',
                                    properties: {
                                        position: { type: 'object' },
                                        orientation: { type: 'object' },
                                        connected: { type: 'boolean' },
                                    },
                                },
                                right: {
                                    type: 'object',
                                    properties: {
                                        position: { type: 'object' },
                                        orientation: { type: 'object' },
                                        connected: { type: 'boolean' },
                                    },
                                },
                            },
                        },
                    },
                },
            },
        },
    },
    // =============================================================================
    // Console Logs (Plugin-specific, not in IWER)
    // =============================================================================
    {
        name: 'browser_get_console_logs',
        description: 'Get console logs from the browser with optional filtering. By default excludes debug level logs.',
        inputSchema: {
            type: 'object',
            properties: {
                count: {
                    type: 'number',
                    description: 'Maximum number of logs to return (most recent N)',
                },
                level: {
                    oneOf: [
                        {
                            type: 'string',
                            enum: ['log', 'info', 'warn', 'error', 'debug'],
                        },
                        {
                            type: 'array',
                            items: {
                                type: 'string',
                                enum: ['log', 'info', 'warn', 'error', 'debug'],
                            },
                        },
                    ],
                    description: 'Filter by log level(s). Default: ["log", "info", "warn", "error"] (excludes debug)',
                },
                pattern: {
                    type: 'string',
                    description: 'Regex pattern to filter log messages',
                },
                since: {
                    type: 'number',
                    description: 'Return logs since this timestamp (ms since epoch)',
                },
            },
        },
    },
    {
        name: 'browser_reload_page',
        description: 'Reload the browser page to reset application state. Use when the app enters an unrecoverable state or to apply code changes.',
        inputSchema: {
            type: 'object',
            properties: {},
        },
    },
    // =============================================================================
    // Framework-Specific Tools (IWSDK or any framework with FRAMEWORK_MCP_RUNTIME)
    // =============================================================================
    {
        name: 'scene_get_hierarchy',
        description: 'Get the Three.js scene hierarchy as a JSON tree. Returns object names, UUIDs, and entity indices where available. Requires IWSDK or a framework that provides FRAMEWORK_MCP_RUNTIME.',
        inputSchema: {
            type: 'object',
            properties: {
                parentId: {
                    type: 'string',
                    description: 'UUID of parent Object3D to start from. Defaults to scene root if omitted.',
                },
                maxDepth: {
                    type: 'number',
                    description: 'Maximum depth to traverse (default: 5). Use to limit context size.',
                },
            },
        },
    },
    {
        name: 'scene_get_object_transform',
        description: 'Get local and global transforms of an Object3D. Includes positionRelativeToXROrigin which can be used directly with xr_look_at tool. Requires IWSDK or a framework that provides FRAMEWORK_MCP_RUNTIME.',
        inputSchema: {
            type: 'object',
            properties: {
                uuid: {
                    type: 'string',
                    description: 'UUID of the Object3D (get this from scene_get_hierarchy)',
                },
            },
            required: ['uuid'],
        },
    },
    // =============================================================================
    // ECS Debugging (IWSDK — requires FRAMEWORK_MCP_RUNTIME)
    // =============================================================================
    {
        name: 'ecs_pause',
        description: 'Pause ECS system updates. The render loop continues (XR session stays alive, screenshots still work) but no systems tick. Use ecs_step to advance individual frames while paused. Requires FRAMEWORK_MCP_RUNTIME.',
        inputSchema: {
            type: 'object',
            properties: {},
        },
    },
    {
        name: 'ecs_resume',
        description: 'Resume ECS system updates after pausing. The first frame after resume uses a capped delta to avoid physics explosions. Requires FRAMEWORK_MCP_RUNTIME.',
        inputSchema: {
            type: 'object',
            properties: {},
        },
    },
    {
        name: 'ecs_step',
        description: 'Advance N ECS frames with a fixed timestep while paused. Must call ecs_pause first. Useful for frame-by-frame debugging. Requires FRAMEWORK_MCP_RUNTIME.',
        inputSchema: {
            type: 'object',
            properties: {
                count: {
                    type: 'number',
                    description: 'Number of frames to advance (1-120, default: 1)',
                },
                delta: {
                    type: 'number',
                    description: 'Fixed timestep in seconds for each frame (default: 1/72 ≈ 0.0139, matching Quest refresh rate)',
                },
            },
        },
    },
    {
        name: 'ecs_query_entity',
        description: 'Get all component data for an entity. Use entityIndex from scene_get_hierarchy or ecs_find_entities. Returns serialized component values including vectors, entity refs, and Object3D references. Requires FRAMEWORK_MCP_RUNTIME.',
        inputSchema: {
            type: 'object',
            properties: {
                entityIndex: {
                    type: 'number',
                    description: 'Entity index (get this from scene_get_hierarchy or ecs_find_entities)',
                },
                components: {
                    type: 'array',
                    items: { type: 'string' },
                    description: 'Optional list of component IDs to include. If omitted, returns all components on the entity.',
                },
            },
            required: ['entityIndex'],
        },
    },
    {
        name: 'ecs_find_entities',
        description: 'Find entities by component composition and/or name. Returns entity indices and component lists. Use the returned entityIndex values with ecs_query_entity for detailed inspection. Requires FRAMEWORK_MCP_RUNTIME.',
        inputSchema: {
            type: 'object',
            properties: {
                withComponents: {
                    type: 'array',
                    items: { type: 'string' },
                    description: 'Component IDs that entities must have (AND logic)',
                },
                withoutComponents: {
                    type: 'array',
                    items: { type: 'string' },
                    description: 'Component IDs that entities must NOT have',
                },
                namePattern: {
                    type: 'string',
                    description: 'Regex pattern to match against entity Object3D name (case-insensitive)',
                },
                limit: {
                    type: 'number',
                    description: 'Maximum number of results (1-50, default: 50)',
                },
            },
        },
    },
    {
        name: 'ecs_list_systems',
        description: 'List all registered ECS systems with name, priority, pause state, config keys, and query entity counts. Requires FRAMEWORK_MCP_RUNTIME.',
        inputSchema: {
            type: 'object',
            properties: {},
        },
    },
    {
        name: 'ecs_list_components',
        description: 'List all registered ECS components with their field schemas (type, default). Requires FRAMEWORK_MCP_RUNTIME.',
        inputSchema: {
            type: 'object',
            properties: {},
        },
    },
    {
        name: 'ecs_toggle_system',
        description: 'Pause or resume a specific ECS system by name. Use ecs_list_systems to discover system names. Requires FRAMEWORK_MCP_RUNTIME.',
        inputSchema: {
            type: 'object',
            properties: {
                name: {
                    type: 'string',
                    description: "System class name (e.g. 'OrbSystem', 'RobotSystem')",
                },
                paused: {
                    type: 'boolean',
                    description: 'Set to true to pause, false to resume. If omitted, toggles current state.',
                },
            },
            required: ['name'],
        },
    },
    {
        name: 'ecs_set_component',
        description: 'Set a component field value on an entity. Scalars use setValue (with validation), vectors accept arrays. Use ecs_query_entity to inspect current values first. Requires FRAMEWORK_MCP_RUNTIME.',
        inputSchema: {
            type: 'object',
            properties: {
                entityIndex: {
                    type: 'number',
                    description: 'Entity index (from ecs_find_entities or scene_get_hierarchy)',
                },
                componentId: {
                    type: 'string',
                    description: "Component ID (e.g. 'Orb', 'RobotMood', 'Transform')",
                },
                field: {
                    type: 'string',
                    description: "Field name within the component (e.g. 'orbitSpeed', 'mood')",
                },
                value: {
                    description: 'New value. Scalars: number/string/boolean. Vectors: array of numbers (e.g. [1,2,3] for Vec3).',
                },
            },
            required: ['entityIndex', 'componentId', 'field', 'value'],
        },
    },
    {
        name: 'ecs_snapshot',
        description: 'Capture a snapshot of all ECS entity/component state. Stores up to 2 snapshots. Use with ecs_diff to compare. Requires FRAMEWORK_MCP_RUNTIME.',
        inputSchema: {
            type: 'object',
            properties: {
                label: {
                    type: 'string',
                    description: 'Label for this snapshot (default: auto-generated). Use to reference in ecs_diff.',
                },
            },
        },
    },
    {
        name: 'ecs_diff',
        description: 'Compare two ECS snapshots. Shows added/removed/changed entities and field-level diffs. Requires FRAMEWORK_MCP_RUNTIME.',
        inputSchema: {
            type: 'object',
            properties: {
                from: {
                    type: 'string',
                    description: "Label of the 'before' snapshot",
                },
                to: {
                    type: 'string',
                    description: "Label of the 'after' snapshot",
                },
            },
            required: ['from', 'to'],
        },
    },
];
const RUNTIME_TOOL_TO_METHOD = {
    xr_get_session_status: 'get_session_status',
    xr_accept_session: 'accept_session',
    xr_end_session: 'end_session',
    xr_get_transform: 'get_transform',
    xr_set_transform: 'set_transform',
    xr_look_at: 'look_at',
    xr_animate_to: 'animate_to',
    xr_set_input_mode: 'set_input_mode',
    xr_set_connected: 'set_connected',
    xr_get_select_value: 'get_select_value',
    xr_set_select_value: 'set_select_value',
    xr_select: 'select',
    xr_get_gamepad_state: 'get_gamepad_state',
    xr_set_gamepad_state: 'set_gamepad_state',
    xr_get_device_state: 'get_device_state',
    xr_set_device_state: 'set_device_state',
    browser_screenshot: 'screenshot',
    browser_get_console_logs: 'get_console_logs',
    browser_reload_page: 'reload_page',
    scene_get_hierarchy: 'get_scene_hierarchy',
    scene_get_object_transform: 'get_object_transform',
};
const RUNTIME_CLI_PATHS = {
    xr_get_session_status: ['xr', 'status'],
    xr_accept_session: ['xr', 'enter'],
    xr_end_session: ['xr', 'exit'],
    xr_get_transform: ['xr', 'get-transform'],
    xr_set_transform: ['xr', 'set-transform'],
    xr_look_at: ['xr', 'look-at'],
    xr_animate_to: ['xr', 'animate-to'],
    xr_set_input_mode: ['xr', 'set-input-mode'],
    xr_set_connected: ['xr', 'set-connected'],
    xr_get_select_value: ['xr', 'get-select-value'],
    xr_set_select_value: ['xr', 'set-select-value'],
    xr_select: ['xr', 'select'],
    xr_get_gamepad_state: ['xr', 'get-gamepad-state'],
    xr_set_gamepad_state: ['xr', 'set-gamepad-state'],
    browser_screenshot: ['browser', 'screenshot'],
    xr_get_device_state: ['xr', 'get-device-state'],
    xr_set_device_state: ['xr', 'set-device-state'],
    browser_get_console_logs: ['browser', 'logs'],
    browser_reload_page: ['browser', 'reload'],
    scene_get_hierarchy: ['scene', 'hierarchy'],
    scene_get_object_transform: ['scene', 'transform'],
    ecs_pause: ['ecs', 'pause'],
    ecs_resume: ['ecs', 'resume'],
    ecs_step: ['ecs', 'step'],
    ecs_query_entity: ['ecs', 'query'],
    ecs_find_entities: ['ecs', 'find'],
    ecs_list_systems: ['ecs', 'systems'],
    ecs_list_components: ['ecs', 'components'],
    ecs_toggle_system: ['ecs', 'toggle-system'],
    ecs_set_component: ['ecs', 'set-component'],
    ecs_snapshot: ['ecs', 'snapshot'],
    ecs_diff: ['ecs', 'diff'],
};
RUNTIME_MCP_TOOLS.map((tool) => {
    const cliPath = RUNTIME_CLI_PATHS[tool.name];
    return {
        id: cliPath ? cliPath.join('.') : tool.name,
        domain: cliPath?.[0] ?? 'misc',
        action: cliPath?.[1] ?? tool.name,
        cliPath: cliPath ?? ['misc', tool.name],
        mcpName: tool.name,
        wsMethod: RUNTIME_TOOL_TO_METHOD[tool.name] ?? tool.name,
        description: tool.description,
        inputSchema: tool.inputSchema,
    };
});

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const RUNTIME_RPC_ERROR_CODE = -32e3;
function createUnavailableBrowserRpcError(browser) {
    if (browser?.status === 'launch_failed') {
        return {
            code: RUNTIME_RPC_ERROR_CODE,
            message: browser.lastError?.message ?? 'Managed browser launch failed.',
            cause: browser.lastError?.cause ?? 'browser_launch_failed',
        };
    }
    if (browser?.status === 'disconnected') {
        return {
            code: RUNTIME_RPC_ERROR_CODE,
            message: browser.lastError?.message ??
                'Managed browser runtime disconnected from the MCP bridge.',
            cause: browser.lastError?.cause ?? 'connection_lost',
        };
    }
    if (browser?.connected && browser.commandReady === false) {
        return {
            code: RUNTIME_RPC_ERROR_CODE,
            message: browser.lastError?.message ??
                'Managed browser command path is still warming up.',
            cause: browser.lastError?.cause ?? 'browser_not_ready',
        };
    }
    return {
        code: RUNTIME_RPC_ERROR_CODE,
        message: 'Browser not ready',
        cause: browser?.lastError?.cause ?? 'browser_not_ready',
    };
}

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const MAX_LOGS = 1000;
const TRACE_PREFIX = '[IWSDK-MCP-TRACE]';
/** Map Playwright console message types to our LogLevel. */
const PLAYWRIGHT_TYPE_MAP = {
    log: 'log',
    info: 'info',
    warning: 'warn',
    error: 'error',
    debug: 'debug',
    trace: 'trace',
    assert: 'error',
};
/**
 * Server-side console capture that accumulates Playwright console events.
 */
class ServerSideConsoleCapture {
    constructor() {
        this.logs = [];
    }
    add(level, message) {
        // Log compaction: if the last entry has the same level + message,
        // increment repeatCount instead of adding a new entry.
        const last = this.logs[this.logs.length - 1];
        if (last && last.level === level && last.message === message) {
            last.repeatCount = (last.repeatCount ?? 1) + 1;
            last.timestamp = Date.now();
            return;
        }
        this.logs.push({
            timestamp: Date.now(),
            level,
            message,
            args: [message],
        });
        if (this.logs.length > MAX_LOGS) {
            this.logs.shift();
        }
    }
    query(options = {}) {
        let result = [...this.logs];
        if (options.level) {
            const levels = Array.isArray(options.level)
                ? options.level
                : [options.level];
            if (levels.length > 0) {
                result = result.filter((log) => levels.includes(log.level));
            }
        }
        if (options.since) {
            result = result.filter((log) => log.timestamp >= options.since);
        }
        if (options.until) {
            result = result.filter((log) => log.timestamp <= options.until);
        }
        if (options.pattern) {
            const regex = new RegExp(options.pattern, 'i');
            result = result.filter((log) => regex.test(log.message));
        }
        if (options.count && options.count > 0) {
            result = result.slice(-options.count);
        }
        return result;
    }
}
let chromiumInstalled = false;
let installPromise = null;
/**
 * Verify the Chromium binary exists and install it automatically if missing.
 * Uses a Promise guard so concurrent callers share one install attempt.
 * On install failure the flag stays unset so the next retry can try again.
 */
async function ensureChromiumInstalled() {
    if (chromiumInstalled) {
        return;
    }
    if (installPromise) {
        return installPromise;
    }
    const execPath = chromium.executablePath();
    if (fs.existsSync(execPath)) {
        chromiumInstalled = true;
        return;
    }
    installPromise = doChromiumInstall().finally(() => {
        installPromise = null;
    });
    return installPromise;
}
async function doChromiumInstall() {
    console.log('\n🔧 IWSDK: Chromium browser not found. Installing (first time only)...\n');
    await new Promise((resolve, reject) => {
        const child = spawn('npx', ['playwright', 'install', 'chromium'], {
            stdio: 'inherit',
            shell: true,
        });
        child.on('close', (code) => {
            if (code === 0) {
                chromiumInstalled = true;
                console.log('\n✅ IWSDK: Chromium installed successfully.\n');
                resolve();
            }
            else {
                reject(new Error(`Chromium installation failed (exit code ${code}). ` +
                    'Try running manually: npx playwright install chromium'));
            }
        });
        child.on('error', (err) => {
            reject(new Error(`Failed to start Chromium installer: ${err.message}. ` +
                'Try running manually: npx playwright install chromium'));
        });
    });
}
/**
 * Check whether a GPU device is available. On macOS/Windows this always
 * returns true (Metal/D3D11 require a GPU by definition). On Linux it
 * probes /dev/dri/ for render nodes, which works in Docker, VMs, and
 * bare-metal environments without requiring external tools.
 */
function hasGpuDevice() {
    if (os.platform() !== 'linux') {
        return true;
    }
    try {
        const entries = fs.readdirSync('/dev/dri');
        return entries.some((e) => e.startsWith('renderD') || e.startsWith('card'));
    }
    catch {
        return false;
    }
}
const VALID_GPU_VALUES = ['auto', 'gpu', 'swiftshader'];
/**
 * Resolve the GPU / ANGLE backend for Chromium.
 *
 * Override with IWSDK_GPU env var:
 *   - "auto"        (default) auto-detect; hardware GL when a GPU is present,
 *                    SwiftShader otherwise
 *   - "gpu"         force hardware GL (fails if no GPU)
 *   - "swiftshader" force CPU-based SwiftShader rendering
 */
function resolveGpuBackend() {
    const envOverride = process.env.IWSDK_GPU?.toLowerCase();
    if (envOverride !== undefined && !VALID_GPU_VALUES.includes(envOverride)) {
        console.warn(`⚠️  IWSDK: Unknown IWSDK_GPU value "${process.env.IWSDK_GPU}". ` +
            'Valid values: auto, gpu, swiftshader. Falling back to auto-detect.');
    }
    if (envOverride === 'swiftshader') {
        return {
            kind: 'swiftshader',
            useGl: 'angle',
            useAngle: 'swiftshader',
            reason: 'env-override',
        };
    }
    const platform = os.platform();
    if (platform === 'darwin') {
        return {
            kind: 'hardware',
            useGl: 'angle',
            useAngle: 'metal',
            reason: 'auto',
        };
    }
    if (platform === 'win32') {
        return {
            kind: 'hardware',
            useGl: 'angle',
            useAngle: 'd3d11',
            reason: 'auto',
        };
    }
    // Linux: honour explicit "gpu" override, otherwise auto-detect
    if (envOverride === 'gpu') {
        return {
            kind: 'hardware',
            useGl: 'angle',
            useAngle: 'gl',
            reason: 'env-override',
        };
    }
    if (hasGpuDevice()) {
        return { kind: 'hardware', useGl: 'angle', useAngle: 'gl', reason: 'auto' };
    }
    return {
        kind: 'swiftshader',
        useGl: 'angle',
        useAngle: 'swiftshader',
        reason: 'auto',
    };
}
async function launchManagedBrowser(url, headless, verbose, viewport = null, screenshotSize = {
    width: 800,
    height: 800,
}, traceMcp = false) {
    await ensureChromiumInstalled();
    const backend = resolveGpuBackend();
    console.log(backend.kind === 'swiftshader'
        ? backend.reason === 'env-override'
            ? '🖥️  IWSDK: Using SwiftShader (software rendering) — IWSDK_GPU=swiftshader'
            : '🖥️  IWSDK: Using SwiftShader (software rendering) — no GPU detected'
        : `🖥️  IWSDK: Using hardware GPU (${backend.useAngle})`);
    const browser = await chromium.launch({
        headless,
        args: [
            '--enable-webgl',
            `--use-gl=${backend.useGl}`,
            `--use-angle=${backend.useAngle}`,
            '--disable-background-timer-throttling',
            '--disable-renderer-backgrounding',
        ],
    });
    const context = await browser.newContext({
        ignoreHTTPSErrors: true, // Accept self-signed certs (e.g. from mkcert)
        viewport, // null = freely resizable; { width, height } = fixed viewport
    });
    const page = await context.newPage();
    // Server-side console capture — accumulates logs from Playwright via CDP
    const consoleCapture = new ServerSideConsoleCapture();
    page.on('console', (msg) => {
        const type = msg.type();
        const text = msg.text();
        const level = PLAYWRIGHT_TYPE_MAP[type];
        // Accumulate into server-side buffer
        if (level) {
            consoleCapture.add(level, text);
        }
        // Also forward to Node console for debugging
        if (type === 'error') {
            console.error('[browser]', text);
        }
        else if (verbose || text.startsWith(TRACE_PREFIX)) {
            console.log(`[browser:${type}]`, text);
        }
    });
    // Capture uncaught page errors (with full stack traces)
    page.on('pageerror', (err) => {
        // err.stack already includes "ErrorName: message" as its first line
        const text = err.stack || (err.name ? `${err.name}: ${err.message}` : err.message);
        consoleCapture.add('error', `[uncaught] ${text}`);
        console.error('[browser:pageerror]', text);
    });
    // Capture unhandled promise rejections by re-emitting as console.error
    // (which Playwright's page.on('console') already captures via CDP)
    await page.addInitScript(() => {
        window.addEventListener('unhandledrejection', (event) => {
            const reason = event.reason;
            // reason.stack already includes "ErrorName: message" as its first line
            const text = reason instanceof Error
                ? reason.stack || `${reason.name}: ${reason.message}`
                : String(reason);
            console.error(`[unhandledrejection] ${text}`);
        });
    });
    // Mark this tab as the Playwright-managed tab. The injection template
    // checks this flag and only initializes the MCP WebSocket client when
    // present, so manually-opened browser tabs are not remote-controlled.
    await page.addInitScript((traceEnabled) => {
        window.__IWER_MCP_MANAGED = true;
        window.__IWSDK_MCP_TRACE = traceEnabled;
    }, traceMcp);
    await page.goto(url, { waitUntil: 'commit' });
    // Wait for IWER to inject and XR device to be available
    await page.waitForFunction(() => window.IWER_DEVICE !== undefined, {
        timeout: 15000,
    });
    if (verbose) {
        console.log(headless
            ? '🖥️  IWSDK: Headless browser launched'
            : '🖥️  IWSDK: Browser launched');
    }
    // Track intentional vs unexpected closure
    let intentionalClose = false;
    let closeCallback = null;
    let closeFired = false;
    const fireCloseCallback = () => {
        if (!intentionalClose && closeCallback && !closeFired) {
            closeFired = true;
            closeCallback();
        }
    };
    page.on('close', fireCloseCallback);
    browser.on('disconnected', fireCloseCallback);
    return {
        close: async () => {
            intentionalClose = true;
            try {
                await context.close();
            }
            finally {
                await browser.close();
            }
        },
        page,
        queryLogs: (options) => consoleCapture.query(options),
        getTabMetadata: async () => page.evaluate(() => {
            const id = typeof sessionStorage !== 'undefined'
                ? sessionStorage.getItem('iwer-mcp-tab-id')
                : null;
            const rawGeneration = typeof sessionStorage !== 'undefined'
                ? sessionStorage.getItem('iwer-mcp-gen')
                : null;
            const generation = rawGeneration != null && rawGeneration !== ''
                ? Number(rawGeneration)
                : null;
            return {
                id,
                generation: Number.isFinite(generation) ? generation : null,
            };
        }),
        screenshot: async () => {
            const raw = await page.screenshot({ type: 'png' });
            // In non-agent modes (viewport is null / freely resizable), downscale
            // the screenshot to fit within screenshotSize bounds.
            if (viewport === null) {
                return sharp(raw)
                    .resize(screenshotSize.width, screenshotSize.height, {
                    fit: 'inside',
                    withoutEnlargement: true,
                })
                    .png()
                    .toBuffer();
            }
            return raw;
        },
        onClose: (callback) => {
            closeCallback = callback;
        },
        isClosed: () => page.isClosed(),
    };
}

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
/**
 * hzdb telemetry integration for IWSDK.
 *
 * Reports MCP tool calls, session lifecycle, and errors through hzdb's
 * DeveloperTelemetry pipeline via the `hzdb xxiwsdk` subcommand.
 * All calls are fire-and-forget — telemetry never blocks or breaks the tool.
 *
 * Uses `npx @meta-quest/hzdb` so we don't need to resolve the binary path.
 * If hzdb is not installed, npx fails and the error is silently swallowed.
 */
function hzdbTelemetry(args, clientVersion) {
    const globalArgs = ['@meta-quest/hzdb', 'xxiwsdk'];
    if (clientVersion) {
        globalArgs.push('--client-version', clientVersion);
    }
    const child = spawn('npx', [...globalArgs, ...args], {
        stdio: 'ignore',
    });
    child.on('error', () => { });
    child.unref();
}
function reportSessionStart(sessionId, opts) {
    const args = ['session-start', '--session-id', sessionId];
    if (opts?.iwsdkVersion) {
        args.push('--iwsdk-version', opts.iwsdkVersion);
    }
    if (opts?.device) {
        args.push('--device', opts.device);
    }
    if (opts?.port) {
        args.push('--port', String(opts.port));
    }
    hzdbTelemetry(args, opts?.clientVersion);
}
function reportSessionEnd(sessionId, opts) {
    const args = ['session-end', '--session-id', sessionId];
    if (opts?.durationMs != null) {
        args.push('--duration-ms', String(Math.round(opts.durationMs)));
    }
    if (opts?.reason) {
        args.push('--reason', opts.reason);
    }
    hzdbTelemetry(args, opts?.clientVersion);
}

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// Get __dirname equivalent for ES modules
const __filename$1 = fileURLToPath(import.meta.url);
const __dirname$1 = path.dirname(__filename$1);
/**
 * Load and customize the pre-built injection bundle
 */
async function buildInjectionBundle(options) {
    try {
        if (options.verbose) {
            console.log('🔄 Loading pre-built IWSDK injection bundle...');
        }
        // Load the pre-built injection bundle
        const bundlePath = path.resolve(__dirname$1, 'injection-bundle.js');
        let bundleCode = await readFile(bundlePath, 'utf8');
        // Create the configuration object to inject
        const config = {
            device: options.device,
            activation: options.activation instanceof RegExp
                ? `/${options.activation.source}/`
                : options.activation,
            verbose: options.verbose,
            sem: options.sem || null,
            ai: options.ai || null,
            userAgentException: options.userAgentException instanceof RegExp
                ? `/${options.userAgentException.source}/`
                : options.userAgentException,
        };
        // Replace the CONFIG token with actual configuration
        const configReplacement = JSON.stringify(config, null, 2);
        bundleCode = bundleCode.replace('"__IWSDK_DEV_CONFIG__"', configReplacement);
        const size = Buffer.byteLength(bundleCode, 'utf8');
        if (options.verbose) {
            console.log(`✅ IWSDK injection bundle loaded and configured (${(size / 1024).toFixed(1)}KB)`);
        }
        return { code: bundleCode, size };
    }
    catch (error) {
        console.error('❌ Failed to load IWSDK injection bundle:', error);
        console.error('Make sure to run "pnpm build:injection" first to create the pre-built bundle.');
        throw error;
    }
}

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
/** WebSocket OPEN readyState constant */
const WS_OPEN = 1;
/**
 * Create a relay handler that implements first-response-wins message routing.
 *
 * When multiple browser tabs are connected, a request from the MCP server is
 * broadcast to all tabs. Each tab processes it and responds. The relay
 * forwards only the FIRST response for each request ID and silently drops
 * duplicates.
 */
function createRelayHandler(options) {
    const verbose = options?.verbose ?? false;
    // Track pending request IDs for first-response-wins deduplication.
    const pendingRelayRequests = new Map();
    function onMessage(senderWs, data, clients) {
        let parsed = null;
        try {
            parsed = JSON.parse(data);
        }
        catch {
            // Not JSON — broadcast as-is for backward compatibility
        }
        if (parsed && typeof parsed.id === 'string') {
            const isRequest = typeof parsed.method === 'string';
            const isResponse = !parsed.method &&
                (parsed.result !== undefined || parsed.error !== undefined);
            if (isRequest) {
                // Track this request for deduplication
                pendingRelayRequests.set(parsed.id, {
                    timestamp: Date.now(),
                    sourceWs: senderWs,
                });
                // Broadcast request to all OTHER clients (all browser tabs)
                clients.forEach((client) => {
                    if (client !== senderWs && client.readyState === WS_OPEN) {
                        client.send(data);
                    }
                });
                return;
            }
            if (isResponse) {
                const pending = pendingRelayRequests.get(parsed.id);
                if (pending) {
                    // First response wins — forward to the original requester
                    pendingRelayRequests.delete(parsed.id);
                    if (pending.sourceWs.readyState === WS_OPEN) {
                        pending.sourceWs.send(data);
                    }
                    if (verbose) {
                        console.log(`[MCP-IWER] Response for ${parsed.id} forwarded (first-wins)`);
                    }
                }
                else if (verbose) {
                    console.log(`[MCP-IWER] Duplicate response for ${parsed.id} dropped`);
                }
                return;
            }
        }
        // Unknown message shape — broadcast for backward compatibility
        clients.forEach((client) => {
            if (client !== senderWs && client.readyState === WS_OPEN) {
                client.send(data);
            }
        });
    }
    function pendingCount() {
        return pendingRelayRequests.size;
    }
    function cleanStale(maxAgeMs) {
        const now = Date.now();
        for (const [id, entry] of pendingRelayRequests) {
            if (now - entry.timestamp > maxAgeMs) {
                pendingRelayRequests.delete(id);
            }
        }
    }
    return { onMessage, pendingCount, cleanStale };
}

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
function isTraceEnabled() {
    return process.env.IWSDK_RUNTIME_TRACE === '1';
}
function traceRuntimeSession(event, details = {}) {
    if (!isTraceEnabled()) {
        return;
    }
    console.error(`[IWSDK-RUNTIME-TRACE][session] ${event} ${JSON.stringify(details)}`);
}
function normalizeWorkspaceRoot(workspaceRoot) {
    const resolved = path__default.resolve(workspaceRoot);
    try {
        return existsSync(resolved) ? realpathSync.native(resolved) : resolved;
    }
    catch {
        return resolved;
    }
}
function getRuntimeSessionFilePath(workspaceRoot) {
    return path__default.join(normalizeWorkspaceRoot(workspaceRoot), IWSDK_RUNTIME_SESSION_PATH);
}
const sessionMutationQueues = new Map();
async function enqueueSessionMutation(workspaceRoot, mutate) {
    const normalizedWorkspaceRoot = normalizeWorkspaceRoot(workspaceRoot);
    const previous = sessionMutationQueues.get(normalizedWorkspaceRoot) ?? Promise.resolve();
    const next = previous
        .catch(() => { })
        .then(() => mutate(normalizedWorkspaceRoot));
    sessionMutationQueues.set(normalizedWorkspaceRoot, next);
    try {
        return await next;
    }
    finally {
        if (sessionMutationQueues.get(normalizedWorkspaceRoot) === next) {
            sessionMutationQueues.delete(normalizedWorkspaceRoot);
        }
    }
}
async function readRuntimeSession(filePath) {
    try {
        const raw = await readFile(filePath, 'utf8');
        return JSON.parse(raw);
    }
    catch {
        return null;
    }
}
async function writeRuntimeSession(workspaceRoot, session) {
    const filePath = getRuntimeSessionFilePath(workspaceRoot);
    await mkdir(path__default.dirname(filePath), { recursive: true });
    const tempPath = `${filePath}.${process.pid}.${Date.now()}.tmp`;
    try {
        await writeFile(tempPath, `${JSON.stringify(session, null, 2)}\n`, 'utf8');
        await rename(tempPath, filePath);
    }
    finally {
        await rm(tempPath, { force: true }).catch(() => { });
    }
    return session;
}
async function registerRuntimeSession(input) {
    return enqueueSessionMutation(input.workspaceRoot, async (workspaceRoot) => {
        const existing = await readRuntimeSession(getRuntimeSessionFilePath(workspaceRoot));
        const now = new Date().toISOString();
        const session = {
            schemaVersion: IWSDK_RUNTIME_STATE_SCHEMA_VERSION,
            sessionId: input.sessionId,
            workspaceRoot,
            pid: input.pid,
            port: input.port,
            localUrl: input.localUrl,
            networkUrls: input.networkUrls ?? [],
            aiMode: input.aiMode,
            browser: input.browser ?? existing?.browser,
            registeredAt: existing?.registeredAt ?? now,
            updatedAt: now,
        };
        traceRuntimeSession('register', {
            workspaceRoot,
            port: input.port,
            browserStatus: session.browser?.status ?? null,
            bridgeConnected: session.browser?.connected ?? false,
            commandReady: session.browser?.commandReady ?? false,
        });
        return writeRuntimeSession(workspaceRoot, session);
    });
}
async function setRuntimeSessionBrowserState(workspaceRoot, browser) {
    return enqueueSessionMutation(workspaceRoot, async (normalizedWorkspaceRoot) => {
        const existing = await readRuntimeSession(getRuntimeSessionFilePath(normalizedWorkspaceRoot));
        if (!existing) {
            return null;
        }
        const session = {
            ...existing,
            browser,
            updatedAt: new Date().toISOString(),
        };
        traceRuntimeSession('set_browser_state', {
            workspaceRoot: normalizedWorkspaceRoot,
            browserStatus: browser.status,
            bridgeConnected: browser.connected,
            commandReady: browser.commandReady,
            connectedClientCount: browser.connectedClientCount,
            lastError: browser.lastError ?? null,
        });
        return writeRuntimeSession(normalizedWorkspaceRoot, session);
    });
}
async function unregisterRuntimeSession(workspaceRoot) {
    await enqueueSessionMutation(workspaceRoot, async (normalizedWorkspaceRoot) => {
        traceRuntimeSession('unregister', {
            workspaceRoot: normalizedWorkspaceRoot,
        });
        await rm(getRuntimeSessionFilePath(normalizedWorkspaceRoot), {
            force: true,
        }).catch(() => { });
    });
}

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
/**
 * Derive internal headless / devUI / viewport settings from the AI mode.
 */
const MODE_SETTINGS = {
    agent: { headless: true, devUI: false, fixedViewport: true },
    oversight: { headless: false, devUI: false, fixedViewport: false },
    collaborate: { headless: false, devUI: true, fixedViewport: false },
};
/**
 * Process and normalize plugin options with defaults
 */
function processOptions(options = {}) {
    const emulator = options.emulator ?? {};
    const processed = {
        device: emulator.device || 'metaQuest3',
        injectOnBuild: emulator.injectOnBuild || false,
        activation: emulator.activation || 'localhost',
        verbose: options.verbose || false,
        userAgentException: emulator.userAgentException || new RegExp('OculusBrowser'),
    };
    // Process SEM options from emulator.environment
    if (emulator.environment) {
        processed.sem = {
            defaultScene: emulator.environment,
        };
    }
    // AI is opt-in: omit `ai` to disable entirely
    if (options.ai) {
        const mode = options.ai.mode ?? 'agent';
        const settings = MODE_SETTINGS[mode];
        if (!settings) {
            const valid = Object.keys(MODE_SETTINGS).join(', ');
            throw new Error(`[IWSDK] Invalid ai.mode "${mode}". Valid modes: ${valid}`);
        }
        const ssInput = options.ai.screenshotSize;
        const ssWidth = ssInput?.width;
        const ssHeight = ssInput?.height;
        const screenshotSize = {
            width: ssWidth ?? ssHeight ?? 800,
            height: ssHeight ?? ssWidth ?? 800,
        };
        processed.ai = {
            mode,
            headless: settings.headless,
            devUI: settings.devUI,
            viewport: settings.fixedViewport ? screenshotSize : null,
            screenshotSize,
        };
    }
    return processed;
}
/**
 * Vite plugin for IWSDK development — XR emulation, AI agent tooling, and Playwright browser
 */
function iwsdkDev(options = {}) {
    const pluginOptions = processOptions(options);
    let injectionBundle = null;
    let config;
    let mcpWss = null;
    let mcpClients = null;
    let managedBrowser = null;
    const VIRTUAL_ID = '/@iwer-injection-runtime';
    const RESOLVED_VIRTUAL_ID = '\0' + VIRTUAL_ID;
    return {
        name: 'iwsdk-dev',
        config(userConfig) {
            // In oversight/collaborate mode the Playwright window IS the visible
            // browser, so suppress Vite's auto-open to avoid a duplicate tab.
            if (pluginOptions.ai && !pluginOptions.ai.headless) {
                if (userConfig.server) {
                    userConfig.server.open = false;
                }
                else {
                    userConfig.server = { open: false };
                }
            }
        },
        configResolved(resolvedConfig) {
            config = resolvedConfig;
            if (pluginOptions.verbose) {
                console.log('🔧 IWSDK Dev Configuration:');
                console.log(`  - Device: ${pluginOptions.device}`);
                console.log(`  - SEM: ${pluginOptions.sem ? 'enabled (' + pluginOptions.sem.defaultScene + ')' : 'disabled'}`);
                console.log(`  - AI: ${pluginOptions.ai ? `enabled (${pluginOptions.ai.mode} mode)` : 'disabled'}`);
                console.log(`  - Activation: ${pluginOptions.activation}`);
                if (pluginOptions.userAgentException) {
                    console.log('  - UA exception: enabled');
                }
                console.log(`  - Inject on build: ${pluginOptions.injectOnBuild}`);
            }
        },
        configureServer(server) {
            if (!pluginOptions.ai) {
                return;
            }
            // Closure-scoped state for browser auto-recovery
            let browserLaunchPromise = null;
            let browserCommandReadyPromise = null;
            let browserRuntimeClients = null;
            let serverShuttingDown = false;
            let browserUrl = '';
            let consecutiveFailures = 0;
            let currentBrowserState = null;
            const MAX_LAUNCH_FAILURES = 3;
            const BRIDGE_READY_TIMEOUT_MS = 5000;
            const traceEnabled = process.env.IWSDK_RUNTIME_TRACE === '1';
            const wsConnectionIds = new WeakMap();
            const wsConnectionKinds = new WeakMap();
            const traceRuntime = (event, details = {}) => {
                if (!traceEnabled) {
                    return;
                }
                console.error(`[IWSDK-RUNTIME-TRACE][vite] ${event} ${JSON.stringify(details)}`);
            };
            const getConnectionId = (ws) => wsConnectionIds.get(ws) ?? 'unknown';
            const getConnectionKind = (ws) => wsConnectionKinds.get(ws) ?? 'command';
            const markConnectionKind = (ws, kind) => {
                wsConnectionKinds.set(ws, kind);
            };
            const createBrowserIssue = (cause, message) => ({
                cause,
                message,
                at: new Date().toISOString(),
            });
            const classifyBrowserLaunchFailure = (message) => /permission|not permitted|denied|sandbox|eacces|eperm/i.test(message)
                ? 'permission_denied'
                : 'browser_launch_failed';
            const createBrowserState = (status, options = {}, previous = currentBrowserState) => {
                const connectedClientCount = options.connectedClientCount ?? browserRuntimeClients?.size ?? 0;
                const connected = options.connected ?? status === 'connected';
                const commandReady = options.commandReady ?? false;
                return {
                    status,
                    connected,
                    commandReady,
                    connectedClientCount,
                    lastTransitionAt: new Date().toISOString(),
                    ...((options.lastBridgeConnectedAt ?? previous?.lastBridgeConnectedAt)
                        ? {
                            lastBridgeConnectedAt: options.lastBridgeConnectedAt ??
                                previous?.lastBridgeConnectedAt,
                        }
                        : {}),
                    ...((options.lastCommandReadyAt ?? previous?.lastCommandReadyAt)
                        ? {
                            lastCommandReadyAt: options.lastCommandReadyAt ?? previous?.lastCommandReadyAt,
                        }
                        : {}),
                    ...(options.lastError ? { lastError: options.lastError } : {}),
                };
            };
            currentBrowserState = createBrowserState('launching', { connected: false }, null);
            const publishBrowserState = (browser) => {
                currentBrowserState = browser;
                traceRuntime('browser_state', {
                    status: browser.status,
                    bridgeConnected: browser.connected,
                    commandReady: browser.commandReady,
                    connectedClientCount: browser.connectedClientCount,
                    lastError: browser.lastError ?? null,
                });
                void setRuntimeSessionBrowserState(config.root, browser).catch((error) => {
                    console.error('[IWSDK Dev] Failed to update browser state:', error);
                });
            };
            /**
             * Launch (or re-launch) the Playwright-managed browser.
             * Guards against concurrent launches via `browserLaunchPromise`.
             * Stops retrying after MAX_LAUNCH_FAILURES consecutive failures.
             */
            const launchBrowser = () => {
                if (browserLaunchPromise) {
                    return browserLaunchPromise;
                }
                browserLaunchPromise = (async () => {
                    publishBrowserState(createBrowserState('launching'));
                    traceRuntime('browser_launch_start', {
                        browserUrl,
                        headless: pluginOptions.ai.headless,
                    });
                    try {
                        const browser = await launchManagedBrowser(browserUrl, pluginOptions.ai.headless, pluginOptions.verbose, pluginOptions.ai.viewport, pluginOptions.ai.screenshotSize, traceEnabled);
                        managedBrowser = browser;
                        consecutiveFailures = 0;
                        traceRuntime('browser_launch_success', {
                            browserRuntimeClients: browserRuntimeClients?.size ?? 0,
                        });
                        publishBrowserState(createBrowserState((browserRuntimeClients?.size ?? 0) > 0
                            ? 'connected'
                            : 'waiting_for_connection', {
                            connected: (browserRuntimeClients?.size ?? 0) > 0,
                            commandReady: false,
                            lastBridgeConnectedAt: (browserRuntimeClients?.size ?? 0) > 0
                                ? new Date().toISOString()
                                : undefined,
                        }));
                        // On unexpected close, mark as null. The browser will be
                        // relaunched lazily on the next MCP request via ensureBrowser().
                        browser.onClose(() => {
                            managedBrowser = null;
                            browserCommandReadyPromise = null;
                            traceRuntime('browser_closed', {
                                serverShuttingDown,
                            });
                            publishBrowserState(createBrowserState('disconnected', {
                                connected: false,
                                commandReady: false,
                                lastError: createBrowserIssue('connection_lost', 'Managed browser closed unexpectedly. It will relaunch on the next MCP request.'),
                            }));
                            if (!serverShuttingDown) {
                                console.log('🔄 IWSDK: Browser closed. Will relaunch on next MCP request.');
                            }
                        });
                    }
                    catch (error) {
                        consecutiveFailures++;
                        const message = error instanceof Error ? error.message : String(error);
                        traceRuntime('browser_launch_failed', {
                            consecutiveFailures,
                            message,
                        });
                        publishBrowserState(createBrowserState('launch_failed', {
                            connected: false,
                            commandReady: false,
                            lastError: createBrowserIssue(classifyBrowserLaunchFailure(message), message),
                        }));
                        console.error('❌ IWSDK: Failed to launch browser:', error);
                        if (consecutiveFailures >= MAX_LAUNCH_FAILURES) {
                            console.error(`❌ IWSDK: ${MAX_LAUNCH_FAILURES} consecutive launch failures, giving up. ` +
                                'Restart the dev server to retry.');
                        }
                    }
                    finally {
                        browserLaunchPromise = null;
                    }
                })();
                return browserLaunchPromise;
            };
            /**
             * Return the current managed browser, re-launching if it was closed.
             * `relaunched` is true when the browser was just freshly launched
             * (meaning the previous page state was lost).
             */
            const ensureBrowser = async () => {
                const current = managedBrowser;
                if (current && !current.isClosed()) {
                    traceRuntime('ensure_browser_reuse', {
                        browserRuntimeClients: browserRuntimeClients?.size ?? 0,
                    });
                    return { browser: current, relaunched: false };
                }
                managedBrowser = null;
                if (consecutiveFailures >= MAX_LAUNCH_FAILURES) {
                    traceRuntime('ensure_browser_aborted', {
                        consecutiveFailures,
                    });
                    return { browser: null, relaunched: false };
                }
                await launchBrowser();
                traceRuntime('ensure_browser_relaunch_result', {
                    relaunched: managedBrowser !== null,
                    browserRuntimeClients: browserRuntimeClients?.size ?? 0,
                });
                return { browser: managedBrowser, relaunched: managedBrowser !== null };
            };
            const waitForBridgeConnection = async (timeoutMs, reason) => {
                const startedAt = Date.now();
                traceRuntime('bridge_wait_start', {
                    reason,
                    timeoutMs,
                    browserRuntimeClients: browserRuntimeClients?.size ?? 0,
                });
                while (Date.now() - startedAt < timeoutMs) {
                    if ((browserRuntimeClients?.size ?? 0) > 0) {
                        const waitedForBridgeMs = Date.now() - startedAt;
                        traceRuntime('bridge_wait_ready', {
                            reason,
                            waitedForBridgeMs,
                        });
                        return waitedForBridgeMs;
                    }
                    await new Promise((resolve) => setTimeout(resolve, 100));
                }
                const waitedForBridgeMs = Date.now() - startedAt;
                traceRuntime('bridge_wait_timeout', {
                    reason,
                    waitedForBridgeMs,
                });
                return waitedForBridgeMs;
            };
            const ensureBrowserCommandReady = async (reason) => {
                if (browserCommandReadyPromise) {
                    return browserCommandReadyPromise;
                }
                browserCommandReadyPromise = (async () => {
                    const { browser, relaunched } = await ensureBrowser();
                    if (!browser) {
                        return {
                            browser: null,
                            relaunched: false,
                            bridgeConnected: false,
                            waitedForBridgeMs: 0,
                        };
                    }
                    const needsBridgeWait = relaunched ||
                        !currentBrowserState?.connected ||
                        currentBrowserState?.commandReady === false;
                    const waitedForBridgeMs = needsBridgeWait
                        ? await waitForBridgeConnection(BRIDGE_READY_TIMEOUT_MS, reason)
                        : 0;
                    const bridgeConnected = (browserRuntimeClients?.size ?? 0) > 0;
                    if (bridgeConnected) {
                        publishBrowserState(createBrowserState('connected', {
                            connected: true,
                            commandReady: true,
                            connectedClientCount: browserRuntimeClients.size,
                            lastBridgeConnectedAt: currentBrowserState?.lastBridgeConnectedAt ??
                                new Date().toISOString(),
                            lastCommandReadyAt: new Date().toISOString(),
                        }));
                    }
                    else {
                        publishBrowserState(createBrowserState('waiting_for_connection', {
                            connected: false,
                            commandReady: false,
                            connectedClientCount: 0,
                            lastError: createBrowserIssue(relaunched ? 'browser_relaunched' : 'browser_not_ready', relaunched
                                ? 'Browser relaunched and is waiting for the runtime bridge to reconnect.'
                                : 'Managed browser bridge has not connected yet.'),
                        }));
                    }
                    traceRuntime('browser_command_ready_result', {
                        reason,
                        relaunched,
                        bridgeConnected,
                        waitedForBridgeMs,
                    });
                    return {
                        browser,
                        relaunched,
                        bridgeConnected,
                        waitedForBridgeMs,
                    };
                })().finally(() => {
                    browserCommandReadyPromise = null;
                });
                return browserCommandReadyPromise;
            };
            // Initialize WebSocket server and client tracking
            mcpClients = new Set();
            browserRuntimeClients = new Set();
            mcpWss = new WebSocketServer({ noServer: true });
            // First-response-wins relay handler (extracted for testability)
            const relay = createRelayHandler({
                verbose: pluginOptions.verbose,
            });
            // Clean up stale entries every 60 seconds
            const relayCleanupInterval = setInterval(() => {
                relay.cleanStale(60000);
            }, 60000);
            relayCleanupInterval.unref();
            const BROWSER_RELAUNCHED_RESULT = {
                status: 'browser_relaunched',
                message: 'Browser was closed and has been relaunched. ' +
                    'The page state has been reset — please retry your request.',
            };
            const sendWsJson = (ws, payload, context) => {
                if (ws.readyState !== WebSocket.OPEN) {
                    traceRuntime('ws_send_skipped', {
                        connectionId: getConnectionId(ws),
                        kind: getConnectionKind(ws),
                        context,
                        readyState: ws.readyState,
                    });
                    return;
                }
                traceRuntime('ws_send', {
                    connectionId: getConnectionId(ws),
                    kind: getConnectionKind(ws),
                    context,
                    id: payload.id ?? null,
                });
                ws.send(JSON.stringify(payload));
            };
            const sendUnavailableBrowser = (ws, requestId, context) => {
                sendWsJson(ws, {
                    id: requestId,
                    error: createUnavailableBrowserRpcError(currentBrowserState),
                }, context);
            };
            const sendInternalError = (ws, requestId, error, context) => {
                sendWsJson(ws, {
                    id: requestId,
                    error: {
                        code: -32e3,
                        message: error instanceof Error ? error.message : String(error),
                    },
                }, context);
            };
            const createBrowserProbeResult = (waitedForBridgeMs) => ({
                bridgeConnected: Boolean(currentBrowserState?.connected),
                commandReady: Boolean(currentBrowserState?.commandReady),
                waitedForBridgeMs,
                browser: currentBrowserState ??
                    createBrowserState('launching', {
                        connected: false,
                        commandReady: false,
                    }),
            });
            mcpWss.on('connection', (ws) => {
                const connectionId = randomUUID();
                wsConnectionIds.set(ws, connectionId);
                markConnectionKind(ws, 'command');
                mcpClients.add(ws);
                traceRuntime('ws_connected', {
                    connectionId,
                    kind: getConnectionKind(ws),
                });
                if (pluginOptions.verbose) {
                    console.log('[IWSDK-MCP] Client connected');
                }
                ws.on('message', async (data) => {
                    const message = data.toString();
                    traceRuntime('ws_message', {
                        connectionId,
                        kind: getConnectionKind(ws),
                        bytes: data.length,
                    });
                    if (pluginOptions.verbose) {
                        console.log('[IWSDK-MCP] Message received:', message.substring(0, 100));
                    }
                    let intercepted = false;
                    try {
                        const parsed = JSON.parse(message);
                        if (parsed?.type === 'iwsdk_browser_hello') {
                            intercepted = true;
                            markConnectionKind(ws, 'bridge');
                            if (!browserRuntimeClients.has(ws)) {
                                browserRuntimeClients.add(ws);
                                publishBrowserState(createBrowserState('connected', {
                                    connected: true,
                                    commandReady: false,
                                    connectedClientCount: browserRuntimeClients.size,
                                    lastBridgeConnectedAt: new Date().toISOString(),
                                }));
                            }
                            traceRuntime('bridge_hello', {
                                connectionId,
                                tabId: parsed.tabId ?? null,
                                tabGeneration: parsed.tabGeneration ?? null,
                                connectedClientCount: browserRuntimeClients.size,
                            });
                            return;
                        }
                        if (parsed.method === INTERNAL_BROWSER_PROBE_METHOD &&
                            typeof parsed.id === 'string') {
                            intercepted = true;
                            traceRuntime('probe_request', {
                                connectionId,
                                requestId: parsed.id,
                            });
                            try {
                                const readiness = await ensureBrowserCommandReady('internal_browser_probe');
                                if (!readiness.browser || !readiness.bridgeConnected) {
                                    sendUnavailableBrowser(ws, parsed.id, 'probe_unavailable');
                                }
                                else {
                                    sendWsJson(ws, {
                                        id: parsed.id,
                                        result: createBrowserProbeResult(readiness.waitedForBridgeMs),
                                    }, 'probe_ready');
                                }
                            }
                            catch (error) {
                                sendInternalError(ws, parsed.id, error, 'probe_error');
                            }
                            return;
                        }
                        if (parsed.method === 'get_console_logs' &&
                            typeof parsed.id === 'string') {
                            intercepted = true;
                            try {
                                const readiness = await ensureBrowserCommandReady('get_console_logs');
                                if (!readiness.browser || !readiness.bridgeConnected) {
                                    sendUnavailableBrowser(ws, parsed.id, 'console_logs_unavailable');
                                    return;
                                }
                                if (readiness.relaunched) {
                                    const tab = await readiness.browser.getTabMetadata();
                                    sendWsJson(ws, {
                                        id: parsed.id,
                                        result: BROWSER_RELAUNCHED_RESULT,
                                        ...(tab.id
                                            ? {
                                                _tabId: tab.id,
                                                _tabGeneration: tab.generation ?? undefined,
                                            }
                                            : {}),
                                    }, 'console_logs_relaunched');
                                    return;
                                }
                                const params = parsed.params ?? {};
                                if (!params.level) {
                                    params.level = ['log', 'info', 'warn', 'error'];
                                }
                                const tab = await readiness.browser.getTabMetadata();
                                sendWsJson(ws, {
                                    id: parsed.id,
                                    result: readiness.browser.queryLogs(params),
                                    ...(tab.id
                                        ? {
                                            _tabId: tab.id,
                                            _tabGeneration: tab.generation ?? undefined,
                                        }
                                        : {}),
                                }, 'console_logs_result');
                            }
                            catch (error) {
                                sendInternalError(ws, parsed.id, error, 'console_logs_error');
                            }
                            return;
                        }
                        if (parsed.method === 'screenshot' &&
                            typeof parsed.id === 'string') {
                            intercepted = true;
                            try {
                                const readiness = await ensureBrowserCommandReady('screenshot');
                                if (!readiness.browser || !readiness.bridgeConnected) {
                                    sendUnavailableBrowser(ws, parsed.id, 'screenshot_unavailable');
                                    return;
                                }
                                if (readiness.relaunched) {
                                    sendWsJson(ws, {
                                        id: parsed.id,
                                        result: BROWSER_RELAUNCHED_RESULT,
                                    }, 'screenshot_relaunched');
                                    return;
                                }
                                const buffer = await readiness.browser.screenshot();
                                const base64 = buffer.toString('base64');
                                sendWsJson(ws, {
                                    id: parsed.id,
                                    result: { imageData: base64, mimeType: 'image/png' },
                                }, 'screenshot_result');
                            }
                            catch (error) {
                                sendInternalError(ws, parsed.id, error, 'screenshot_error');
                            }
                            return;
                        }
                    }
                    catch (error) {
                        traceRuntime('ws_message_parse_fallthrough', {
                            connectionId,
                            kind: getConnectionKind(ws),
                            message: error instanceof Error ? error.message : 'non_json_message',
                        });
                    }
                    if (!intercepted) {
                        relay.onMessage(ws, message, mcpClients);
                    }
                });
                ws.on('close', (code, reasonBuffer) => {
                    const kind = getConnectionKind(ws);
                    const reason = typeof reasonBuffer === 'string'
                        ? reasonBuffer
                        : reasonBuffer.toString('utf8');
                    mcpClients.delete(ws);
                    const removedBridge = browserRuntimeClients.delete(ws);
                    if (removedBridge) {
                        browserCommandReadyPromise = null;
                        publishBrowserState(createBrowserState(browserRuntimeClients.size > 0 ? 'connected' : 'disconnected', {
                            connected: browserRuntimeClients.size > 0,
                            commandReady: false,
                            connectedClientCount: browserRuntimeClients.size,
                            lastError: browserRuntimeClients.size > 0
                                ? undefined
                                : createBrowserIssue('connection_lost', 'Managed browser runtime disconnected from the MCP bridge.'),
                        }));
                    }
                    traceRuntime('ws_closed', {
                        connectionId,
                        kind,
                        code,
                        reason,
                        removedBridge,
                        browserRuntimeClients: browserRuntimeClients.size,
                    });
                    if (pluginOptions.verbose) {
                        console.log('[IWSDK-MCP] Client disconnected');
                    }
                });
                ws.on('error', (error) => {
                    traceRuntime('ws_error', {
                        connectionId,
                        kind: getConnectionKind(ws),
                        message: error instanceof Error ? error.message : String(error),
                    });
                    if (pluginOptions.verbose) {
                        console.error('[IWSDK-MCP] WebSocket error:', error);
                    }
                });
            });
            // Set up WebSocket endpoint for MCP - handle upgrade requests
            server.httpServer?.on('upgrade', (request, socket, head) => {
                if (request.url !== '/__iwer_mcp') {
                    return;
                }
                traceRuntime('ws_upgrade', {
                    url: request.url,
                    remoteAddress: request.socket.remoteAddress ?? null,
                });
                if (pluginOptions.verbose) {
                    console.log('[IWSDK-MCP] WebSocket upgrade request received');
                }
                mcpWss.handleUpgrade(request, socket, head, (ws) => {
                    mcpWss.emit('connection', ws, request);
                });
            });
            if (pluginOptions.verbose) {
                console.log('🔌 IWSDK-MCP: WebSocket endpoint registered at /__iwer_mcp');
            }
            // Register the project-local runtime session after server start.
            // Waiting for 'listening' lets us record Vite's actual chosen port.
            // Resolve IWSDK version for telemetry attribution
            let iwsdkVersion;
            try {
                const pluginPkgPath = path.join(config.root, 'node_modules', '@iwsdk', 'vite-plugin-dev', 'package.json');
                const pluginPkg = JSON.parse(readFileSync(pluginPkgPath, 'utf-8'));
                iwsdkVersion = pluginPkg.version;
            }
            catch {
                // Version detection is best-effort
            }
            // Session tracking for telemetry
            const sessionId = randomUUID();
            const sessionStartTime = Date.now();
            // Wait for server to start listening to get the actual port
            server.httpServer?.on('listening', async () => {
                const address = server.httpServer?.address();
                const actualPort = typeof address === 'object' && address
                    ? address.port
                    : server.config.server.port || 5173;
                const protocol = server.config.server.https ? 'https' : 'http';
                browserUrl = `${protocol}://localhost:${actualPort}`;
                try {
                    await registerRuntimeSession({
                        sessionId,
                        workspaceRoot: config.root,
                        pid: process.pid,
                        port: actualPort,
                        localUrl: server.resolvedUrls?.local?.[0] ?? browserUrl,
                        networkUrls: server.resolvedUrls?.network ?? [],
                        aiMode: pluginOptions.ai?.mode,
                        browser: currentBrowserState ?? undefined,
                    });
                }
                catch (error) {
                    console.error('[IWSDK Dev] Failed to register runtime session:', error);
                }
                // Report session start to hzdb telemetry (fire-and-forget via npx)
                reportSessionStart(sessionId, {
                    iwsdkVersion,
                    clientVersion: iwsdkVersion,
                    port: actualPort,
                });
                // Launch Playwright-managed browser
                launchBrowser();
            });
            // Clean up WebSocket server and browser when Vite server closes.
            server.httpServer?.on('close', () => {
                serverShuttingDown = true;
                void unregisterRuntimeSession(config.root);
                reportSessionEnd(sessionId, {
                    durationMs: Date.now() - sessionStartTime,
                    reason: 'user_closed',
                    clientVersion: iwsdkVersion,
                });
                if (mcpWss) {
                    for (const client of mcpClients || []) {
                        client.close();
                    }
                    mcpClients?.clear();
                    mcpWss.close();
                    mcpWss = null;
                }
                if (managedBrowser) {
                    managedBrowser.close().catch(() => { });
                    managedBrowser = null;
                }
            });
        },
        resolveId(id) {
            if (id === VIRTUAL_ID) {
                return RESOLVED_VIRTUAL_ID;
            }
        },
        load(id) {
            if (id === RESOLVED_VIRTUAL_ID) {
                if (!injectionBundle) {
                    return 'console.warn("[IWSDK Dev] Runtime not available - injection bundle not loaded");';
                }
                return injectionBundle.code;
            }
        },
        async buildStart() {
            // Determine if we should generate injection script
            const shouldInject = config.command === 'serve' ||
                (config.command === 'build' && pluginOptions.injectOnBuild);
            if (!shouldInject) {
                if (pluginOptions.verbose && config.command === 'build') {
                    console.log('⏭️  IWSDK Dev: Skipping build injection (injectOnBuild: false)');
                }
                return;
            }
            try {
                if (pluginOptions.verbose) {
                    console.log('🚀 IWSDK Dev: Starting injection bundle generation...');
                }
                injectionBundle = await buildInjectionBundle(pluginOptions);
                if (pluginOptions.verbose) {
                    console.log('✅ IWSDK Dev: Injection bundle ready');
                }
            }
            catch (error) {
                console.error('❌ IWSDK Dev: Failed to generate injection bundle:', error);
                // Continue without injection rather than failing the build
            }
        },
        transformIndexHtml: {
            order: 'pre', // Run before other HTML transformations
            handler(html) {
                // Check if we should inject
                const shouldInject = config.command === 'serve' ||
                    (config.command === 'build' && pluginOptions.injectOnBuild);
                if (!shouldInject || !injectionBundle) {
                    return html;
                }
                if (pluginOptions.verbose) {
                    console.log('💉 IWSDK Dev: Injecting runtime script into HTML');
                }
                // Inject the script using Vite's tag API for robustness
                return {
                    tags: [
                        {
                            tag: 'script',
                            attrs: { type: 'module', src: VIRTUAL_ID },
                            injectTo: 'head',
                        },
                    ],
                };
            },
        },
        // Display summary at the end of build process
        closeBundle: {
            order: 'post',
            async handler() {
                // Only show summary when injection actually happened
                const shouldInject = config.command === 'serve' ||
                    (config.command === 'build' && pluginOptions.injectOnBuild);
                if (shouldInject && injectionBundle) {
                    const mode = config.command === 'serve' ? 'Development' : 'Build';
                    console.log(`\n🥽 IWSDK Dev Summary (${mode}):`);
                    console.log(`  - Device: ${pluginOptions.device}`);
                    console.log(`  - Runtime injected: ${(injectionBundle.size / 1024).toFixed(1)}KB`);
                    console.log(`  - Activation mode: ${pluginOptions.activation}`);
                    if (pluginOptions.sem) {
                        console.log(`  - SEM environment: ${pluginOptions.sem.defaultScene}`);
                    }
                    if (pluginOptions.ai) {
                        console.log(`  - AI: ${pluginOptions.ai.mode} mode (WebSocket at /__iwer_mcp)`);
                    }
                    if (pluginOptions.activation === 'localhost') {
                        console.log('  - Note: Runtime only activates on localhost/local networks');
                    }
                    console.log(''); // Extra line for spacing
                }
            },
        },
    };
}
/** @deprecated Use `iwsdkDev` instead */
const injectIWER = iwsdkDev;

export { injectIWER, iwsdkDev };
