// @iwsdk/core/dist/locomotion/locomotion-input-provider.js
import { InputActions } from '../input/input-actions.js';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const TURN_THRESHOLD = 0.8;
/**
 * Converts low-level input actions into locomotion intent.
 *
 * Locomotion should not care whether input came from XR controllers, keyboard,
 * browser gamepads, or an app-authored adapter. This provider is the narrow
 * bridge between generic actions and the existing locomotor systems.
 */
class ActionLocomotionInputProvider {
    constructor(world) {
        this.world = world;
    }
    enableBrowserControls(options) {
        if (!options) {
            return;
        }
        const bindingOptions = typeof options === 'object'
            ? {
                keyboard: options.keyboard,
                gamepad: options.gamepad,
            }
            : undefined;
        this.world.input.actions.enableDefaultBindings('browser-first-person', bindingOptions);
    }
    getMoveAxis(out) {
        const value = this.world.input.actions.getAxis2D(InputActions.LocomotionMove);
        return out.set(value.x, value.y);
    }
    getMovementReferenceQuaternion(out) {
        if (this.world.session) {
            this.world.player.head.getWorldQuaternion(out);
        }
        else {
            this.world.camera.getWorldQuaternion(out);
        }
        return out;
    }
    getJumpDown() {
        return this.world.input.actions.getButtonDown(InputActions.LocomotionJump);
    }
    getTurnAxis() {
        return this.world.input.actions.getAxis1D(InputActions.LocomotionTurn);
    }
    getTurnLeftDown(microGestureControlsEnabled) {
        var _a, _b;
        if (this.isHandMicroGestureMode(microGestureControlsEnabled)) {
            return (_b = (_a = this.world.input.xr.gamepads.right) === null || _a === void 0 ? void 0 : _a.getButtonDownByIdx(5)) !== null && _b !== void 0 ? _b : false;
        }
        return this.world.input.actions.getAxis1DEnteringNegative(InputActions.LocomotionTurn, TURN_THRESHOLD);
    }
    getTurnRightDown(microGestureControlsEnabled) {
        var _a, _b;
        if (this.isHandMicroGestureMode(microGestureControlsEnabled)) {
            return (_b = (_a = this.world.input.xr.gamepads.right) === null || _a === void 0 ? void 0 : _a.getButtonDownByIdx(6)) !== null && _b !== void 0 ? _b : false;
        }
        return this.world.input.actions.getAxis1DEnteringPositive(InputActions.LocomotionTurn, TURN_THRESHOLD);
    }
    shouldShowTurnSignals(microGestureControlsEnabled) {
        return this.isHandMicroGestureMode(microGestureControlsEnabled);
    }
    getTeleportActive(pointerBusy, microGestureControlsEnabled) {
        var _a, _b;
        if (pointerBusy) {
            return false;
        }
        if (this.world.input.xr.isPrimary('hand', 'right')) {
            return (microGestureControlsEnabled &&
                !((_b = (_a = this.world.input.xr.gamepads.right) === null || _a === void 0 ? void 0 : _a.getButtonDownByIdx(9)) !== null && _b !== void 0 ? _b : false));
        }
        return this.world.input.actions.getButtonPressed(InputActions.LocomotionTeleportAim);
    }
    getTeleportCommit(microGestureControlsEnabled) {
        var _a, _b;
        if (this.world.input.xr.isPrimary('hand', 'right')) {
            return (microGestureControlsEnabled &&
                ((_b = (_a = this.world.input.xr.gamepads.right) === null || _a === void 0 ? void 0 : _a.getButtonDownByIdx(9)) !== null && _b !== void 0 ? _b : false));
        }
        return this.world.input.actions.getButtonUp(InputActions.LocomotionTeleportAim);
    }
    getTeleportRay(origin, direction) {
        if (!this.world.session) {
            return false;
        }
        this.world.player.raySpaces.right.getWorldPosition(origin);
        this.world.player.raySpaces.right.getWorldDirection(direction);
        return true;
    }
    isHandMicroGestureMode(microGestureControlsEnabled) {
        return (microGestureControlsEnabled &&
            this.world.input.xr.isPrimary('hand', 'right'));
    }
}
function getRequiredInputProvider(systemName, value) {
    if (value instanceof ActionLocomotionInputProvider) {
        return value;
    }
    throw new Error(`${systemName} must be registered through LocomotionSystem so locomotion subsystems share one input provider.`);
}

export { ActionLocomotionInputProvider, getRequiredInputProvider };
