// @iwsdk/core/dist/environment-raycast/environment-raycast.js
import { createSystem } from '../ecs/system.js';
import '../runtime/three.js';
import '../transform/transform.js';
import { setWorldPosition, setWorldQuaternion } from '../transform/transform-utils.js';
import { EnvironmentRaycastTarget, RaycastSpace } from './raycast-target.js';
import { Matrix4, Vector3, Quaternion } from 'three';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// Reusable math objects to avoid allocations
const tempMatrix = new Matrix4();
const tempPosition = new Vector3();
const tempQuaternion = new Quaternion();
const tempWorldQuaternion = new Quaternion();
const tempScale = new Vector3();
const tempNormal = new Vector3();
const tempRayOrigin = new Vector3();
const tempDirection = new Vector3();
const upVector = new Vector3(0, 1, 0);
const defaultDirection = new Vector3(0, 0, -1);
/**
 * System that manages environment raycasting.
 * @category Environment Raycast
 */
class EnvironmentRaycastSystem extends createSystem({
    targets: { required: [EnvironmentRaycastTarget] },
}) {
    constructor() {
        super(...arguments);
        this.pendingControllerSources = new Map();
    }
    init() {
        this.xrManager.addEventListener('sessionstart', () => {
            this.updateEnabledFeatures(this.xrManager.getSession());
        });
        this.xrManager.addEventListener('sessionend', () => {
            this.queries.targets.entities.forEach((entity) => {
                this.cleanupRaycastSource(entity);
            });
            this.hitTestFeatureEnabled = undefined;
            this.pendingControllerSources.clear();
        });
        // Automatic cleanup when component is removed
        this.queries.targets.subscribe('disqualify', (entity) => {
            this.cleanupRaycastSource(entity);
        });
    }
    update(_delta, _time) {
        const frame = this.xrFrame;
        const session = this.xrManager.getSession();
        const referenceSpace = this.xrManager.getReferenceSpace();
        if (!this.hitTestFeatureEnabled || !frame || !session || !referenceSpace) {
            return;
        }
        this.checkPendingControllerSources(session, referenceSpace);
        this.queries.targets.entities.forEach((entity) => {
            const space = entity.getValue(EnvironmentRaycastTarget, 'space');
            if (space === RaycastSpace.Screen) {
                this.handleTransientInput(entity, session, frame, referenceSpace);
            }
            else {
                this.handleStandardRaycast(entity, session, frame, referenceSpace);
            }
        });
    }
    handleStandardRaycast(entity, session, frame, referenceSpace) {
        const source = entity.getValue(EnvironmentRaycastTarget, '_source');
        const sourceRequested = entity.getValue(EnvironmentRaycastTarget, '_sourceRequested');
        if (!source && !sourceRequested) {
            this.requestRaycastSource(entity, session, referenceSpace);
            return;
        }
        if (source) {
            this.updateRaycastResults(entity, frame, referenceSpace);
        }
    }
    handleTransientInput(entity, session, frame, referenceSpace) {
        const transientSource = entity.getValue(EnvironmentRaycastTarget, '_transientSource');
        const sourceRequested = entity.getValue(EnvironmentRaycastTarget, '_sourceRequested');
        if (!transientSource && !sourceRequested) {
            this.requestTransientRaycastSource(entity, session);
            return;
        }
        if (transientSource) {
            this.updateTransientRaycastResults(entity, frame, referenceSpace, transientSource);
        }
    }
    checkPendingControllerSources(session, referenceSpace) {
        if (this.pendingControllerSources.size === 0) {
            return;
        }
        for (const [entity, handedness] of this.pendingControllerSources) {
            const inputSource = this.input.xr.getPrimaryInputSource(handedness);
            if (inputSource) {
                this.pendingControllerSources.delete(entity);
                entity.setValue(EnvironmentRaycastTarget, '_sourceRequested', false);
                this.requestRaycastSource(entity, session, referenceSpace);
            }
        }
    }
    updateEnabledFeatures(xrSession) {
        var _a;
        if (!xrSession) {
            return;
        }
        this.hitTestFeatureEnabled =
            (_a = xrSession.enabledFeatures) === null || _a === void 0 ? void 0 : _a.includes('hit-test');
        if (!this.hitTestFeatureEnabled) {
            console.warn('[EnvironmentRaycastSystem] hit-test feature not enabled.');
        }
    }
    async requestRaycastSource(entity, session, _referenceSpace) {
        if (!session.requestHitTestSource) {
            return;
        }
        entity.setValue(EnvironmentRaycastTarget, '_sourceRequested', true);
        try {
            const space = entity.getValue(EnvironmentRaycastTarget, 'space');
            const xrSpace = await this.resolveXRSpace(entity, space, session);
            if (!xrSpace) {
                return;
            }
            // Build offsetRay from component properties
            const offsetRay = this.buildOffsetRay(entity);
            const source = await session.requestHitTestSource({
                space: xrSpace,
                offsetRay,
            });
            entity.setValue(EnvironmentRaycastTarget, '_source', source);
            entity.setValue(EnvironmentRaycastTarget, '_raySpace', xrSpace);
            entity.setValue(EnvironmentRaycastTarget, '_sourceRequested', false);
        }
        catch (error) {
            console.error('[EnvironmentRaycastSystem] Failed to create source:', error);
            entity.setValue(EnvironmentRaycastTarget, '_sourceRequested', false);
        }
    }
    buildOffsetRay(entity) {
        const offsetPosition = entity.getValue(EnvironmentRaycastTarget, 'offsetPosition');
        const offsetQuaternion = entity.getValue(EnvironmentRaycastTarget, 'offsetQuaternion');
        // Calculate direction from quaternion (apply to default -Z direction)
        if (offsetQuaternion) {
            tempDirection.copy(defaultDirection).applyQuaternion(offsetQuaternion);
        }
        else {
            tempDirection.copy(defaultDirection);
        }
        const origin = offsetPosition
            ? new DOMPointReadOnly(offsetPosition.x, offsetPosition.y, offsetPosition.z, 1)
            : new DOMPointReadOnly(0, 0, 0, 1);
        const direction = new DOMPointReadOnly(tempDirection.x, tempDirection.y, tempDirection.z, 0);
        return new XRRay(origin, direction);
    }
    async requestTransientRaycastSource(entity, session) {
        if (!session.requestHitTestSourceForTransientInput) {
            return;
        }
        entity.setValue(EnvironmentRaycastTarget, '_sourceRequested', true);
        try {
            const source = await session.requestHitTestSourceForTransientInput({
                profile: 'generic-touchscreen',
            });
            entity.setValue(EnvironmentRaycastTarget, '_transientSource', source);
            entity.setValue(EnvironmentRaycastTarget, '_sourceRequested', false);
        }
        catch (error) {
            console.error('[EnvironmentRaycastSystem] Failed to create transient source:', error);
            entity.setValue(EnvironmentRaycastTarget, '_sourceRequested', false);
        }
    }
    async resolveXRSpace(entity, space, session) {
        switch (space) {
            case RaycastSpace.Viewer:
                return await session.requestReferenceSpace('viewer');
            case RaycastSpace.Left:
            case RaycastSpace.Right: {
                const inputSource = this.input.xr.getPrimaryInputSource(space);
                if (inputSource) {
                    return inputSource.targetRaySpace;
                }
                this.pendingControllerSources.set(entity, space);
                return null;
            }
            default:
                return await session.requestReferenceSpace('viewer');
        }
    }
    updateRaycastResults(entity, frame, referenceSpace) {
        const source = entity.getValue(EnvironmentRaycastTarget, '_source');
        if (!source) {
            return;
        }
        const raySpace = entity.getValue(EnvironmentRaycastTarget, '_raySpace');
        const object3D = entity.object3D;
        const maxDistance = entity.getValue(EnvironmentRaycastTarget, 'maxDistance');
        try {
            const results = frame.getHitTestResults(source);
            if (results.length === 0) {
                this.clearHitResult(entity, object3D);
                return;
            }
            // Get ray origin for distance calculation
            const rayOrigin = this.getRayOrigin(frame, referenceSpace, raySpace);
            // Find first valid result within maxDistance
            const validResult = this.findValidHitResult(results, referenceSpace, maxDistance, rayOrigin);
            if (!validResult) {
                this.clearHitResult(entity, object3D);
                return;
            }
            this.applyHitResult(entity, object3D, validResult, referenceSpace, undefined);
        }
        catch (_a) {
            this.cleanupRaycastSource(entity);
        }
    }
    updateTransientRaycastResults(entity, frame, referenceSpace, transientSource) {
        const object3D = entity.object3D;
        const maxDistance = entity.getValue(EnvironmentRaycastTarget, 'maxDistance');
        try {
            const transientResults = frame.getHitTestResultsForTransientInput(transientSource);
            for (const transientResult of transientResults) {
                const inputSource = transientResult.inputSource;
                const rayOrigin = this.getRayOrigin(frame, referenceSpace, inputSource === null || inputSource === void 0 ? void 0 : inputSource.targetRaySpace);
                const validResult = this.findValidHitResult(transientResult.results, referenceSpace, maxDistance, rayOrigin);
                if (validResult) {
                    this.applyHitResult(entity, object3D, validResult, referenceSpace, inputSource);
                    return;
                }
            }
            // No valid hit found in any transient result
            this.clearHitResult(entity, object3D);
        }
        catch (_a) {
            this.cleanupRaycastSource(entity);
        }
    }
    getRayOrigin(frame, referenceSpace, raySpace) {
        if (!raySpace) {
            return null;
        }
        const rayPose = frame.getPose(raySpace, referenceSpace);
        if (!rayPose) {
            return null;
        }
        tempRayOrigin.set(rayPose.transform.position.x, rayPose.transform.position.y, rayPose.transform.position.z);
        return tempRayOrigin;
    }
    findValidHitResult(results, referenceSpace, maxDistance, rayOrigin) {
        for (const result of results) {
            const pose = result.getPose(referenceSpace);
            if (!pose) {
                continue;
            }
            tempMatrix.fromArray(pose.transform.matrix);
            tempMatrix.decompose(tempPosition, tempQuaternion, tempScale);
            // Only apply distance filter if we have a valid ray origin
            if (rayOrigin) {
                const distance = tempPosition.distanceTo(rayOrigin);
                if (distance > maxDistance) {
                    continue;
                }
            }
            return result;
        }
        return null;
    }
    applyHitResult(entity, object3D, result, referenceSpace, inputSource) {
        const pose = result.getPose(referenceSpace);
        if (!pose) {
            this.clearHitResult(entity, object3D);
            return;
        }
        tempMatrix.fromArray(pose.transform.matrix);
        tempMatrix.decompose(tempPosition, tempQuaternion, tempScale);
        tempNormal.set(0, 1, 0).applyQuaternion(tempQuaternion);
        if (object3D) {
            setWorldPosition(object3D, tempPosition);
            tempWorldQuaternion.setFromUnitVectors(upVector, tempNormal);
            setWorldQuaternion(object3D, tempWorldQuaternion);
            object3D.visible = true;
        }
        entity.setValue(EnvironmentRaycastTarget, 'xrHitTestResult', result);
        entity.setValue(EnvironmentRaycastTarget, 'inputSource', inputSource);
    }
    clearHitResult(entity, object3D) {
        entity.setValue(EnvironmentRaycastTarget, 'xrHitTestResult', undefined);
        entity.setValue(EnvironmentRaycastTarget, 'inputSource', undefined);
        if (object3D) {
            object3D.visible = false;
        }
    }
    cleanupRaycastSource(entity) {
        const source = entity.getValue(EnvironmentRaycastTarget, '_source');
        if (source) {
            try {
                source.cancel();
            }
            catch (_a) {
                /* ignore */
            }
        }
        const transientSource = entity.getValue(EnvironmentRaycastTarget, '_transientSource');
        if (transientSource) {
            try {
                transientSource.cancel();
            }
            catch (_b) {
                /* ignore */
            }
        }
        entity.setValue(EnvironmentRaycastTarget, '_source', undefined);
        entity.setValue(EnvironmentRaycastTarget, '_raySpace', undefined);
        entity.setValue(EnvironmentRaycastTarget, '_transientSource', undefined);
        entity.setValue(EnvironmentRaycastTarget, '_sourceRequested', false);
        entity.setValue(EnvironmentRaycastTarget, 'xrHitTestResult', undefined);
        entity.setValue(EnvironmentRaycastTarget, 'inputSource', undefined);
        this.pendingControllerSources.delete(entity);
        if (entity.object3D) {
            entity.object3D.visible = false;
        }
    }
}

export { EnvironmentRaycastSystem };
