// @iwer/devui/lib/prefs.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
export const IWE_PREFS_CHANNEL = 'iwer-prefs';
export function emitPrefsPatch(patch) {
    if (typeof window === 'undefined' || !patch)
        return;
    const targetOrigin = window.location.origin === 'null' ? '*' : window.location.origin;
    window.postMessage({
        channel: IWE_PREFS_CHANNEL,
        scope: 'origin',
        patch,
    }, targetOrigin);
}
// Trailing-edge debounce for high-frequency controls (e.g. the FOV slider drag)
// so a continuous gesture coalesces into a single persisted write instead of
// flooding the extension's storage layer. Patches are shallow-merged across the
// window; only top-level keys are emitted here so that is sufficient.
let debounceTimer = null;
let debouncedPatch = null;
export function emitPrefsPatchDebounced(patch, delayMs = 200) {
    if (typeof window === 'undefined' || !patch)
        return;
    debouncedPatch = { ...(debouncedPatch !== null && debouncedPatch !== void 0 ? debouncedPatch : {}), ...patch };
    if (debounceTimer != null)
        clearTimeout(debounceTimer);
    debounceTimer = setTimeout(() => {
        const pending = debouncedPatch;
        debouncedPatch = null;
        debounceTimer = null;
        if (pending)
            emitPrefsPatch(pending);
    }, delayMs);
}
let environmentLoader = null;
export function setEnvironmentLoader(loader) {
    environmentLoader = loader;
}
export function getEnvironmentLoader() {
    return environmentLoader;
}
