// @iwsdk/xr-input/dist/gamepad/stateful-gamepad.js
import { StatefulButtonAxesDevice } from './stateful-button-axes-device.js';
export { AxesState } from './stateful-button-axes-device.js';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var InputComponent;
(function (InputComponent) {
    InputComponent["Trigger"] = "xr-standard-trigger";
    InputComponent["Squeeze"] = "xr-standard-squeeze";
    InputComponent["Touchpad"] = "xr-standard-touchpad";
    InputComponent["Thumbstick"] = "xr-standard-thumbstick";
    InputComponent["A_Button"] = "a-button";
    InputComponent["B_Button"] = "b-button";
    InputComponent["X_Button"] = "x-button";
    InputComponent["Y_Button"] = "y-button";
    InputComponent["Thumbrest"] = "thumbrest";
    InputComponent["Menu"] = "menu";
})(InputComponent || (InputComponent = {}));
class StatefulGamepad extends StatefulButtonAxesDevice {
    constructor({ inputSource, layout }) {
        super(layout, inputSource.gamepad.buttons.length);
        this.handedness = inputSource.handedness;
        this.gamepad = inputSource.gamepad;
        this.inputSource = inputSource;
    }
    update() {
        super.update(this.gamepad.buttons, this.gamepad.axes);
    }
}

export { InputComponent, StatefulGamepad };
