// @iwsdk/xr-input/dist/pointer/cursor-visual.js
import { CanvasTexture, Vector3, Quaternion, Matrix3, Mesh, CircleGeometry, MeshBasicMaterial } from 'three';
import { lerp } from '../node_modules/.pnpm/super-three@0.181.0/node_modules/super-three/src/math/MathUtils.js';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// Create shared cursor texture
const cursorRes = 512;
const canvas = document.createElement('canvas');
canvas.width = cursorRes;
canvas.height = cursorRes;
const ctx = canvas.getContext('2d');
ctx.clearRect(0, 0, canvas.width, canvas.height);
ctx.fillStyle = 'white';
ctx.beginPath();
ctx.arc(cursorRes / 2, cursorRes / 2, (cursorRes / 16) * 7, 0, Math.PI * 2);
ctx.fill();
ctx.strokeStyle = 'gray';
ctx.lineWidth = 3;
ctx.beginPath();
ctx.arc(cursorRes / 2, cursorRes / 2, (cursorRes / 16) * 7, 0, Math.PI * 2);
ctx.stroke();
const cursorTexture = new CanvasTexture(canvas);
const ZAxis = new Vector3(0, 0, 1);
const offsetHelper = new Vector3();
const cursorPosition = new Vector3();
const quaternionHelper = new Quaternion();
const scratchNormal = new Vector3();
const scratchNormalMatrix = new Matrix3();
let cursorCount = 0;
/**
 * Shared cursor visual that can be updated from any pointer's intersection.
 * Renders as a circle mesh positioned at the intersection point, oriented to the surface normal.
 *
 * @category Pointer
 */
class CursorVisual {
    constructor(xrOrigin, pointerIndex) {
        this.xrOrigin = xrOrigin;
        this.focusAlpha = 0;
        this.cursor = new Mesh(new CircleGeometry(0.008), new MeshBasicMaterial({
            map: cursorTexture,
            transparent: true,
        }));
        this.cursor.renderOrder = Infinity;
        this.cursor.userData.attached = true;
        this.cursor.visible = false;
        this.zOffset = 0.004 + (pointerIndex + cursorCount++) * 0.001;
        xrOrigin.add(this.cursor);
    }
    /**
     * Update cursor position and orientation from an intersection.
     */
    updateFromIntersection(intersection, delta, focused) {
        var _a, _b;
        cursorPosition.copy(intersection.pointOnFace);
        this.focusAlpha = lerp(this.focusAlpha, focused ? 1 : 0, 30 * delta);
        const cursorScale = (Math.max(0, intersection.distance - 0.3) + 1) *
            lerp(1, 0.8, this.focusAlpha);
        this.cursor.material.opacity = lerp(0.7, 1, this.focusAlpha);
        this.cursor.scale.setScalar(cursorScale);
        const normal = (_a = intersection.normal) !== null && _a !== void 0 ? _a : (_b = intersection.face) === null || _b === void 0 ? void 0 : _b.normal;
        if (normal != null) {
            // Convert local-space normal to world-space using normal matrix to handle non-uniform scales
            scratchNormal.copy(normal);
            scratchNormalMatrix.getNormalMatrix(intersection.object.matrixWorld);
            scratchNormal.applyNormalMatrix(scratchNormalMatrix).normalize();
            // Build world-space orientation from +Z to world normal
            this.cursor.quaternion.setFromUnitVectors(ZAxis, scratchNormal);
            // Convert world orientation to xrOrigin local space
            quaternionHelper.copy(this.xrOrigin.quaternion).invert();
            this.cursor.quaternion.multiply(quaternionHelper);
            // Offset slightly along the oriented normal to avoid z-fighting
            offsetHelper.set(0, 0, this.zOffset);
            offsetHelper.applyQuaternion(this.cursor.quaternion);
            cursorPosition.add(offsetHelper);
        }
        else if (intersection.pointerQuaternion) {
            // Fallback: align cursor with pointer direction when no surface normal is available
            this.cursor.quaternion.copy(intersection.pointerQuaternion);
            quaternionHelper.copy(this.xrOrigin.quaternion).invert();
            this.cursor.quaternion.multiply(quaternionHelper);
            offsetHelper.set(0, 0, this.zOffset);
            offsetHelper.applyQuaternion(this.cursor.quaternion);
            cursorPosition.add(offsetHelper);
        }
        this.xrOrigin.worldToLocal(cursorPosition);
        this.cursor.position.copy(cursorPosition);
        this.cursor.updateMatrix();
    }
    /**
     * Set cursor visibility.
     */
    setVisible(visible) {
        this.cursor.visible = visible;
    }
    /**
     * Dispose of cursor resources.
     */
    dispose() {
        this.cursor.geometry.dispose();
        this.cursor.material.dispose();
        this.cursor.removeFromParent();
    }
}

export { CursorVisual };
