// @iwsdk/xr-input/dist/pointer/ray-pointer.js
import { createRayPointer } from '@pmndrs/pointer-events';
import { Color, Mesh, CylinderGeometry, ShaderMaterial } from 'three';
import { lerp } from '../node_modules/.pnpm/super-three@0.181.0/node_modules/super-three/src/math/MathUtils.js';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const vertexShader = `
  varying float vPosition;
  void main() {
    vPosition = (position.z + 1.0) / 1.0;
    gl_Position = projectionMatrix * modelViewMatrix * vec4(position, 1.0);
  }
`;
const fragmentShader = `
  uniform float endValue;
  uniform float opacity;
  uniform vec3 color;
  varying float vPosition;
  void main() {
    float alpha = vPosition < endValue ? smoothstep(endValue-0.05, endValue, vPosition) :
               vPosition < 0.97 ? 1.0 :
               1.0 - smoothstep(0.97, 1.0, vPosition);
    gl_FragColor = vec4(color, alpha * opacity);
  }
`;
const rayPressedColor = new Color(0x3383e6);
const rayDefaultColor = new Color(0xffffff);
var RayDisplayMode;
(function (RayDisplayMode) {
    RayDisplayMode[RayDisplayMode["Visible"] = 1] = "Visible";
    RayDisplayMode[RayDisplayMode["VisibleOnIntersection"] = 2] = "VisibleOnIntersection";
    RayDisplayMode[RayDisplayMode["Invisible"] = 3] = "Invisible";
})(RayDisplayMode || (RayDisplayMode = {}));
class RayPointer {
    constructor(camera, xrOrigin, handedness) {
        this.enabled = true;
        this.rayDisplayMode = RayDisplayMode.VisibleOnIntersection;
        this.pointer = createRayPointer(() => camera, { current: xrOrigin.raySpaces[handedness] }, {}, {
            // Disable contextmenu on button 2 (squeeze) to avoid spurious events
            contextMenuButton: -1,
        });
        // Optimize raycaster for BVH acceleration - only get first hit for better performance
        const raycaster = this.pointer.intersector.raycaster;
        if (raycaster) {
            raycaster.firstHitOnly = true;
        }
        this.ray = new Mesh(new CylinderGeometry(0.001, 0.001, 1, 6, 1, true)
            .translate(0, 0.5, 0)
            .rotateX(-Math.PI / 2), new ShaderMaterial({
            vertexShader: vertexShader,
            fragmentShader: fragmentShader,
            transparent: true,
            depthWrite: false,
            uniforms: {
                endValue: { value: 0.75 },
                color: { value: new Color().copy(rayDefaultColor) },
                opacity: { value: 0 },
            },
        }));
        this.ray.visible = false;
        this.ray.renderOrder = 0;
        xrOrigin.raySpaces[handedness].add(this.ray);
    }
    update(connected, delta, _time, selectStart, selectEnd, policy) {
        // CombinedPointer is responsible for moving/enabling; we only render visuals
        // CombinedPointer controls actual pointer enablement; reflect that in visuals
        const pointerEnabled = this.pointer.getEnabled();
        const active = pointerEnabled && connected && this.enabled;
        this.ray.visible = active && !(policy === null || policy === void 0 ? void 0 : policy.forceHideRay);
        if (active) {
            if (pointerEnabled && selectStart) {
                this.ray.material.uniforms.color.value.copy(rayPressedColor);
            }
            else if (pointerEnabled && selectEnd) {
                this.ray.material.uniforms.color.value.copy(rayDefaultColor);
            }
            // Movement is handled by the owning CombinedPointer aggregator
        }
        this.updatePointerRendering(active, delta);
    }
    updatePointerRendering(pointerActive, delta = 1) {
        let rayOpacityTarget = 0;
        if (pointerActive) {
            const intersection = this.pointer.getIntersection();
            const intersectionValid = !!(intersection && !intersection.object.isVoidObject);
            this.rayIntersection = intersectionValid ? intersection : undefined;
            switch (this.rayDisplayMode) {
                case RayDisplayMode.Visible:
                    rayOpacityTarget = 1;
                    break;
                case RayDisplayMode.Invisible:
                    rayOpacityTarget = 0;
                    break;
                default:
                    rayOpacityTarget = intersectionValid ? 1 : 0;
            }
            if (intersectionValid) {
                this.ray.material.uniforms.endValue.value =
                    1.05 - Math.min(0.3, intersection.distance);
            }
        }
        else {
            this.rayIntersection = undefined;
        }
        this.ray.material.uniforms.opacity.value = lerp(this.ray.material.uniforms.opacity.value, rayOpacityTarget, delta * 10);
    }
    get busy() {
        return !!this.rayIntersection;
    }
    /**
     * Dispose of ray pointer resources.
     */
    dispose() {
        this.ray.geometry.dispose();
        this.ray.material.dispose();
        this.ray.removeFromParent();
    }
}

export { RayDisplayMode, RayPointer };
