// @iwsdk/core/dist/environment/ibl-texture.js
import { createComponent, Types } from 'elics';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
/**
 * Image‑based lighting from a texture or the built‑in Room environment.
 *
 * @remarks
 * - `src`: "room" | HDR/EXR equirect | LDR equirect. PMREM processed to `scene.environment`.
 * - `rotation` controls `scene.environmentRotation` in radians.
 * - Does not set `scene.background`.
 *
 * @category Environment & Lighting
 */
const IBLTexture = createComponent('IBLTexture', {
    src: {
        type: Types.FilePath,
        default: 'room',
        fileTypes: '.png,.jpg,.ktx2,.hdr,.exr',
        subfolder: 'textures',
    },
    intensity: { type: Types.Float32, default: 1.0 },
    rotation: { type: Types.Vec3, default: [0, 0, 0] },
    _needsUpdate: { type: Types.Boolean, default: true },
}, 'Image‑based lighting from texture or Room environment');

export { IBLTexture };
