// @iwsdk/core/dist/depth/depth-sensing-system.js
import { createSystem } from '../ecs/system.js';
import { Types } from 'elics';
import '../runtime/three.js';
import { DepthOccludable, OcclusionShadersMode } from './depth-occludable.js';
import { DepthTextures } from './depth-textures.js';
import { DepthPreprocessingPass } from './occlusion/preprocessing-pass.js';
import { Vector2, Mesh } from 'three';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
/**
 * DepthSensingSystem - Manages WebXR depth sensing and occlusion.
 *
 * @remarks
 * - Automatically retrieves and processes depth data from WebXR sessions.
 * - Supports both CPU-optimized and GPU-optimized depth sensing.
 * - Enables occlusion of virtual objects behind real-world surfaces.
 * - Requires WebXR session feature: 'depth-sensing'.
 * NOTE: The depth occlusion feature may not be compatible with custom shaders.
 *
 * @example Basic depth sensing setup
 * ```ts
 * // Configure world with depth sensing
 * World.create(document.getElementById('scene-container'), {
 *   xr: {
 *     sessionMode: SessionMode.ImmersiveAR,
 *     features: {
 *       depthSensing: { usage: 'gpu-optimized', format: 'float32' }
 *     },
 *   }
 * })
 *
 * // Add the depth sensing system
 * world.registerSystem(DepthSensingSystem)
 * ```
 *
 * @example Enable occlusion for an object
 * ```ts
 * const cube = world.createTransformEntity(cubeObject)
 * cube.addComponent(DepthOccludable) // Will be occluded by real-world geometry
 * ```
 *
 * @category Depth Sensing
 * @see {@link DepthOccludable}
 */
class DepthSensingSystem extends createSystem({
    occludables: { required: [DepthOccludable] },
}, {
    enableOcclusion: { type: Types.Boolean, default: true },
    enableDepthTexture: { type: Types.Boolean, default: true },
    useFloat32: { type: Types.Boolean, default: true },
    blurRadius: { type: Types.Float32, default: 20.0 },
}) {
    constructor() {
        super(...arguments);
        // Depth data storage
        this.cpuDepthData = [];
        this.gpuDepthData = [];
        // Occlusion
        this.entityShaderMap = new Map();
        this.viewportSize = new Vector2();
        this.minMaxEntityCount = 0;
    }
    /**
     * Get the raw value to meters conversion factor.
     */
    get rawValueToMeters() {
        if (this.cpuDepthData.length) {
            return this.cpuDepthData[0].rawValueToMeters;
        }
        else if (this.gpuDepthData.length) {
            return this.gpuDepthData[0].rawValueToMeters;
        }
        return 0;
    }
    init() {
        this.xrManager.addEventListener('sessionstart', () => {
            this.updateEnabledFeatures(this.xrManager.getSession());
        });
        this.xrManager.addEventListener('sessionend', () => {
            this.cleanup();
        });
        // React to config changes
        this.config.enableDepthTexture.subscribe((enabled) => {
            if (enabled && !this.depthTextures) {
                this.initializeDepthTextures();
            }
        });
        this.queries.occludables.subscribe('qualify', (entity) => {
            this.attachOcclusionToEntity(entity);
            if (DepthOccludable.data.mode[entity.index] ===
                OcclusionShadersMode.MinMaxSoftOcclusion) {
                this.minMaxEntityCount++;
            }
        });
        this.queries.occludables.subscribe('disqualify', (entity) => {
            if (DepthOccludable.data.mode[entity.index] ===
                OcclusionShadersMode.MinMaxSoftOcclusion) {
                this.minMaxEntityCount--;
            }
            this.detachOcclusionFromEntity(entity);
        });
    }
    initializeDepthTextures() {
        this.depthTextures = new DepthTextures(this.config.useFloat32.value);
    }
    /**
     * Injects inline depth occlusion shader code into all materials of an entity.
     */
    attachOcclusionToEntity(entity) {
        const object3D = entity.object3D;
        if (!object3D) {
            return;
        }
        const entityUniforms = new Set();
        this.entityShaderMap.set(entity, entityUniforms);
        object3D.traverse((child) => {
            var _a;
            if (!(child instanceof Mesh)) {
                return;
            }
            const materials = Array.isArray(child.material)
                ? child.material
                : [child.material];
            for (const material of materials) {
                if (!material) {
                    continue;
                }
                material.transparent = true;
                const existingCallback = (_a = material.onBeforeCompile) === null || _a === void 0 ? void 0 : _a.bind(material);
                material.onBeforeCompile = (shader, renderer) => {
                    if (existingCallback) {
                        existingCallback(shader, renderer);
                    }
                    // Only inject occlusion if not already present
                    if (!shader.uniforms.occlusionEnabled) {
                        DepthSensingSystem.addOcclusionToShader(shader);
                    }
                    material.userData.shader = shader;
                    entityUniforms.add(shader.uniforms);
                };
                material.needsUpdate = true;
            }
        });
    }
    detachOcclusionFromEntity(entity) {
        this.entityShaderMap.delete(entity);
    }
    /**
     * Modifies a material's shader in-place to incorporate inline depth-based
     * occlusion. Compares the virtual fragment's view-space depth against the
     * real-world depth from the XR depth texture array.
     * Both GPU-optimized (ExternalTexture) and CPU-optimized (DataArrayTexture)
     * paths use sampler2DArray with VIEW_ID for correct stereo depth sampling.
     * @param shader - The shader object provided by onBeforeCompile.
     */
    static addOcclusionToShader(shader) {
        var _a;
        shader.uniforms.occlusionEnabled = { value: false };
        shader.uniforms.uXRDepthTextureArray = { value: null };
        shader.uniforms.uRawValueToMeters = { value: 0.001 };
        shader.uniforms.uIsGPUDepth = { value: false };
        shader.uniforms.uDepthNear = { value: 0 };
        shader.uniforms.uViewportSize = { value: new Vector2() };
        shader.uniforms.uOcclusionBlurRadius = { value: 20.0 };
        shader.uniforms.uOcclusionHardMode = { value: false };
        shader.uniforms.uOcclusionMinMaxMode = { value: false };
        shader.uniforms.uMinMaxTexture0 = { value: null };
        shader.uniforms.uMinMaxTexture1 = { value: null };
        shader.defines = {
            ...((_a = shader.defines) !== null && _a !== void 0 ? _a : {}),
            USE_UV: true,
        };
        // Vertex shader: compute view-space depth for occlusion comparison
        shader.vertexShader = shader.vertexShader
            .replace('#include <common>', ['varying float vOcclusionViewDepth;', '#include <common>'].join('\n'))
            .replace('#include <fog_vertex>', [
            '#include <fog_vertex>',
            'vec4 occlusion_view_pos = modelViewMatrix * vec4(position, 1.0);',
            'vOcclusionViewDepth = -occlusion_view_pos.z;',
        ].join('\n'));
        // Fragment shader: sample XR depth array and compare against virtual depth
        shader.fragmentShader = shader.fragmentShader
            .replace('uniform vec3 diffuse;', [
            'uniform vec3 diffuse;',
            'uniform bool occlusionEnabled;',
            'uniform float uRawValueToMeters;',
            'uniform bool uIsGPUDepth;',
            'uniform float uDepthNear;',
            'uniform vec2 uViewportSize;',
            'uniform float uOcclusionBlurRadius;',
            'uniform bool uOcclusionHardMode;',
            'uniform bool uOcclusionMinMaxMode;',
            'uniform sampler2D uMinMaxTexture0;',
            'uniform sampler2D uMinMaxTexture1;',
            'varying float vOcclusionViewDepth;',
            '',
            'uniform sampler2DArray uXRDepthTextureArray;',
            '',
            '// Fallback for non-multiview sessions',
            '#ifndef VIEW_ID',
            '#define VIEW_ID 0',
            '#endif',
            '',
            'float OcclusionDepthGetMeters(in vec2 uv) {',
            '  float textureValue = texture(uXRDepthTextureArray, vec3(uv.x, uv.y, float(VIEW_ID))).r;',
            '  if (uIsGPUDepth) {',
            '    return uRawValueToMeters * uDepthNear / (1.0 - textureValue);',
            '  }',
            '  return textureValue * uRawValueToMeters;',
            '}',
            '',
            'float OcclusionGetSample(in vec2 depthUV, in vec2 offset) {',
            '  float sampleDepth = OcclusionDepthGetMeters(depthUV + offset);',
            '  return smoothstep(0.0, 0.05, sampleDepth - vOcclusionViewDepth);',
            '}',
        ].join('\n'))
            .replace('vec4 diffuseColor = vec4( diffuse, opacity );', [
            'vec4 diffuseColor = vec4( diffuse, opacity );',
            'if (occlusionEnabled) {',
            '  vec2 screenUV = gl_FragCoord.xy / uViewportSize;',
            '  vec2 depthUV = uIsGPUDepth ? screenUV : vec2(screenUV.x, 1.0 - screenUV.y);',
            '  float occlusion_value;',
            '  if (uOcclusionMinMaxMode) {',
            '    // MinMax soft occlusion — two-cluster edge-aware blending',
            '    vec4 mmData;',
            '    if (uint(VIEW_ID) == 0u) {',
            '      mmData = texture2D(uMinMaxTexture0, depthUV);',
            '    } else {',
            '      mmData = texture2D(uMinMaxTexture1, depthUV);',
            '    }',
            '    float minAvgDepth = mmData.r;',
            '    float maxAvgDepth = mmData.g;',
            '    float midAvgDepth = mmData.r + mmData.b;',
            '    float fadeRange = vOcclusionViewDepth * 0.04;',
            '    float fadeRangeInv = 1.0 / max(fadeRange, 0.001);',
            '    vec3 envDepths = vec3(minAvgDepth, maxAvgDepth, midAvgDepth);',
            '    vec3 occAlphas = clamp((envDepths - vOcclusionViewDepth) * fadeRangeInv, 0.0, 1.0);',
            '    occlusion_value = occAlphas.z;',
            '    float alphaDiff = occAlphas.y - occAlphas.x;',
            '    if (alphaDiff > 0.03) {',
            '      float denom = mmData.a;',
            '      float interp = denom > 0.001 ? mmData.b / denom : 0.5;',
            '      occlusion_value = mix(occAlphas.x, occAlphas.y, smoothstep(0.2, 0.8, interp));',
            '    }',
            '  } else if (uOcclusionHardMode) {',
            '    occlusion_value = OcclusionGetSample(depthUV, vec2(0.0));',
            '  } else {',
            '   vec2 texelSize = uOcclusionBlurRadius / uViewportSize;',
            '   // 13-tap two-ring sampling pattern for smooth occlusion edges',
            '   // Center sample',
            '   occlusion_value = OcclusionGetSample(depthUV, vec2(0.0));',
            '   // Inner ring: 6 samples at 40% radius, 60 degree intervals',
            '   occlusion_value += OcclusionGetSample(depthUV, texelSize * vec2( 0.4,  0.0));',
            '   occlusion_value += OcclusionGetSample(depthUV, texelSize * vec2( 0.2,  0.346));',
            '   occlusion_value += OcclusionGetSample(depthUV, texelSize * vec2(-0.2,  0.346));',
            '   occlusion_value += OcclusionGetSample(depthUV, texelSize * vec2(-0.4,  0.0));',
            '   occlusion_value += OcclusionGetSample(depthUV, texelSize * vec2(-0.2, -0.346));',
            '   occlusion_value += OcclusionGetSample(depthUV, texelSize * vec2( 0.2, -0.346));',
            '   // Outer ring: 6 samples at full radius, offset 30 degrees',
            '   occlusion_value += OcclusionGetSample(depthUV, texelSize * vec2( 0.866,  0.5));',
            '   occlusion_value += OcclusionGetSample(depthUV, texelSize * vec2( 0.0,    1.0));',
            '   occlusion_value += OcclusionGetSample(depthUV, texelSize * vec2(-0.866,  0.5));',
            '   occlusion_value += OcclusionGetSample(depthUV, texelSize * vec2(-0.866, -0.5));',
            '   occlusion_value += OcclusionGetSample(depthUV, texelSize * vec2( 0.0,   -1.0));',
            '   occlusion_value += OcclusionGetSample(depthUV, texelSize * vec2( 0.866, -0.5));',
            '   occlusion_value /= 13.0;',
            '  }',
            '  if (occlusion_value < 0.01) {',
            '      discard;',
            '  }',
            '  diffuseColor.a *= occlusion_value;',
            '}',
        ].join('\n'));
    }
    cleanup() {
        var _a;
        this.depthFeatureEnabled = undefined;
        this.cpuDepthData = [];
        this.gpuDepthData = [];
        (_a = this.preprocessingPass) === null || _a === void 0 ? void 0 : _a.dispose();
        this.preprocessingPass = undefined;
    }
    updateEnabledFeatures(xrSession) {
        if (!xrSession) {
            return;
        }
        const enabledFeatures = xrSession.enabledFeatures;
        this.depthFeatureEnabled = enabledFeatures === null || enabledFeatures === void 0 ? void 0 : enabledFeatures.includes('depth-sensing');
        if (!this.depthFeatureEnabled) {
            console.log('Warning: depth-sensing feature not enabled for WebXR session. Depth sensing features are disabled.');
        }
    }
    update() {
        if (!this.depthFeatureEnabled) {
            return;
        }
        const frame = this.xrFrame;
        if (frame) {
            this.updateLocalDepth(frame);
        }
        if (this.config.enableOcclusion.value) {
            this.runMinMaxPreprocessing();
            this.updateOcclusionUniforms();
        }
    }
    /**
     * Runs the MinMax depth preprocessing pass if any entity uses MinMaxSoftOcclusion.
     * Renders a fullscreen pass per eye that computes min/max/avg depth in a 4×4
     * neighborhood, outputting to per-view 2D render targets.
     */
    get isGPUDepth() {
        const session = this.xrManager.getSession();
        return (session === null || session === void 0 ? void 0 : session.depthUsage) === 'gpu-optimized';
    }
    getDepthTextureArray() {
        var _a, _b;
        return this.isGPUDepth
            ? (_a = this.depthTextures) === null || _a === void 0 ? void 0 : _a.getNativeTexture()
            : (_b = this.depthTextures) === null || _b === void 0 ? void 0 : _b.getDataArrayTexture();
    }
    runMinMaxPreprocessing() {
        var _a, _b;
        if (this.minMaxEntityCount === 0) {
            return;
        }
        const depthTextureArray = this.getDepthTextureArray();
        if (!depthTextureArray) {
            return;
        }
        const isGPUDepth = this.isGPUDepth;
        const depthNear = (_b = (_a = this.gpuDepthData[0]) === null || _a === void 0 ? void 0 : _a.depthNear) !== null && _b !== void 0 ? _b : 0;
        if (!this.preprocessingPass) {
            this.preprocessingPass = new DepthPreprocessingPass();
        }
        this.preprocessingPass.setDepthTexture(depthTextureArray, this.rawValueToMeters, isGPUDepth, depthNear);
        // Determine depth texture dimensions
        let depthWidth;
        let depthHeight;
        if (this.cpuDepthData[0]) {
            depthWidth = this.cpuDepthData[0].width;
            depthHeight = this.cpuDepthData[0].height;
        }
        else if (this.gpuDepthData[0]) {
            // GPU depth textures have their own resolution (typically ~256×192),
            // which is much smaller than the drawing buffer.
            depthWidth = this.gpuDepthData[0].width;
            depthHeight = this.gpuDepthData[0].height;
        }
        else {
            return; // No depth data available
        }
        // Render preprocessing for both eyes
        this.preprocessingPass.render(this.renderer, depthWidth, depthHeight, 0);
        this.preprocessingPass.render(this.renderer, depthWidth, depthHeight, 1);
    }
    /**
     * Updates depth data from the XR frame.
     */
    updateLocalDepth(frame) {
        const session = frame.session;
        const binding = this.renderer.xr.getBinding();
        const xrRefSpace = this.renderer.xr.getReferenceSpace();
        if (xrRefSpace) {
            const pose = frame.getViewerPose(xrRefSpace);
            if (pose) {
                if (session.depthUsage === 'gpu-optimized') {
                    // GPU path: the native texture is a texture array containing all
                    // views. We only need to update the ExternalTexture once using the
                    // first view's depth data.
                    const view = pose.views[0];
                    const depthData = binding.getDepthInformation(view);
                    if (depthData) {
                        this.updateGPUDepthData(depthData);
                    }
                }
                else {
                    // CPU path: each view has its own DataTexture.
                    for (let viewId = 0; viewId < pose.views.length; ++viewId) {
                        const view = pose.views[viewId];
                        const depthData = frame.getDepthInformation(view);
                        if (!depthData) {
                            return;
                        }
                        this.updateCPUDepthData(depthData, viewId);
                    }
                }
            }
        }
    }
    /**
     * Update with CPU-optimized depth data.
     */
    updateCPUDepthData(depthData, viewId = 0) {
        this.cpuDepthData[viewId] = depthData;
        if (this.config.enableDepthTexture.value && this.depthTextures) {
            this.depthTextures.updateData(depthData, viewId);
        }
    }
    /**
     * Update with GPU-optimized depth data.
     */
    updateGPUDepthData(depthData) {
        this.gpuDepthData[0] = depthData;
        if (this.config.enableDepthTexture.value && this.depthTextures) {
            this.depthTextures.updateNativeTexture(depthData, this.renderer);
        }
    }
    /**
     * Updates depth texture uniforms on all occludable materials each frame.
     * Both GPU and CPU paths set the texture array uniform; the shader uses
     * VIEW_ID to select the correct stereo layer.
     */
    updateOcclusionUniforms() {
        var _a, _b;
        const depthTextureArray = this.getDepthTextureArray();
        if (!depthTextureArray) {
            return;
        }
        const isGPUDepth = this.isGPUDepth;
        const depthNear = (_b = (_a = this.gpuDepthData[0]) === null || _a === void 0 ? void 0 : _a.depthNear) !== null && _b !== void 0 ? _b : 0;
        this.renderer.getDrawingBufferSize(this.viewportSize);
        for (const [entity, entityUniforms] of this.entityShaderMap) {
            const isHardMode = DepthOccludable.data.mode[entity.index] ===
                OcclusionShadersMode.HardOcclusion;
            const isMinMaxMode = DepthOccludable.data.mode[entity.index] ===
                OcclusionShadersMode.MinMaxSoftOcclusion;
            for (const uniforms of entityUniforms) {
                uniforms.uXRDepthTextureArray.value = depthTextureArray;
                uniforms.uRawValueToMeters.value = this.rawValueToMeters;
                uniforms.uIsGPUDepth.value = isGPUDepth;
                uniforms.uDepthNear.value = depthNear;
                uniforms.uViewportSize.value.copy(this.viewportSize);
                uniforms.uOcclusionBlurRadius.value = this.config.blurRadius.value;
                uniforms.uOcclusionHardMode.value = isHardMode;
                uniforms.uOcclusionMinMaxMode.value = isMinMaxMode;
                uniforms.occlusionEnabled.value = this.config.enableOcclusion.value;
                if (isMinMaxMode && this.preprocessingPass) {
                    uniforms.uMinMaxTexture0.value = this.preprocessingPass.getTexture(0);
                    uniforms.uMinMaxTexture1.value = this.preprocessingPass.getTexture(1);
                }
            }
        }
    }
}

export { DepthSensingSystem };
