// @iwer/devui/lib/components/controller.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { ControlPanel, Layout, PanelHeaderButton, SectionBreak, } from './styled.js';
import { ControlsMapper, useKeyMapStore } from './mapper.js';
import { Icon } from './icon.js';
import { Joystick } from './joystick.js';
import React from 'react';
import { ScrubControl } from './scrub.js';
import { Vector3Input } from './vec3.js';
const BUTTON_LABELS = {
    thumbstick: 'Stick',
    trigger: 'Trig',
    squeeze: 'Grip',
    'a-button': 'A',
    'b-button': 'B',
    'x-button': 'X',
    'y-button': 'Y',
    thumbrest: 'Rest',
    menu: 'Menu',
};
function labelFor(id) {
    if (BUTTON_LABELS[id])
        return BUTTON_LABELS[id];
    return id.charAt(0).toUpperCase() + id.slice(1).replace(/-button$/, '');
}
function transformGamepadConfig(gamepadConfig) {
    const axesSet = new Set();
    for (const axis of gamepadConfig.axes) {
        if (axis && axis.id)
            axesSet.add(axis.id);
    }
    const transformed = gamepadConfig.buttons
        .filter((button) => button !== null)
        .map((button) => ({
        id: button.id,
        type: button.type,
        hasAxes: axesSet.has(button.id),
    }));
    transformed.sort((a, b) => {
        if (a.hasAxes && !b.hasAxes)
            return -1;
        if (!a.hasAxes && b.hasAxes)
            return 1;
        return 0;
    });
    return transformed;
}
export const ControllerUI = ({ controller, handle, handedness, pointerLocked, }) => {
    const { keyMap } = useKeyMapStore();
    const [connected, setConnected] = React.useState(controller.connected);
    const [settingsOpen, setSettingsOpen] = React.useState(false);
    const transformedConfig = transformGamepadConfig(controller.gamepadConfig);
    const actions = transformedConfig.flatMap((config) => config.hasAxes
        ? [
            `${config.id}-left`,
            `${config.id}-right`,
            `${config.id}-up`,
            `${config.id}-down`,
            config.id,
        ]
        : config.id);
    React.useEffect(() => {
        if (pointerLocked)
            setSettingsOpen(false);
    }, [pointerLocked]);
    return (_jsxs(ControlPanel, { style: handedness === 'left'
            ? { left: '8px', bottom: '8px' }
            : { right: '8px', bottom: '8px' }, children: [!pointerLocked && (_jsxs("div", { style: {
                    display: 'flex',
                    flexDirection: 'row',
                    justifyContent: 'space-between',
                    alignItems: 'center',
                }, children: [_jsxs("div", { style: {
                            fontSize: '14px',
                            fontWeight: 600,
                            display: 'flex',
                            flexDirection: 'row',
                            alignItems: 'center',
                            gap: '6px',
                        }, children: [_jsx(Icon, { name: "gamepad-2", size: 15 }), "Controller [", handedness === 'left' ? 'L' : 'R', "]"] }), _jsx("div", { style: { display: 'flex', flexDirection: 'row', gap: '1px' }, children: connected ? (_jsxs(_Fragment, { children: [_jsx(PanelHeaderButton, { title: `Click to ${settingsOpen ? 'close' : 'change'} key bindings`, onClick: () => setSettingsOpen(!settingsOpen), children: _jsx(Icon, { name: "settings", size: 14 }) }), _jsx(PanelHeaderButton, { title: `Click to disconnect ${handedness} controller`, "$isRed": true, onClick: () => {
                                        controller.connected = false;
                                        setConnected(false);
                                    }, children: _jsx(Icon, { name: "circle-x", size: 14 }) })] })) : (_jsx(PanelHeaderButton, { title: `Click to reconnect ${handedness} controller`, onClick: () => {
                                controller.connected = true;
                                setConnected(true);
                            }, style: { marginLeft: '5px' }, children: _jsx(Icon, { name: "plug", size: 14 }) })) })] })), connected && !pointerLocked && (_jsxs(_Fragment, { children: [!settingsOpen && (_jsxs(_Fragment, { children: [_jsx(SectionBreak, {}), _jsx(Vector3Input, { vector: handle.position, label: "Pos", marginBottom: Layout.gap }), _jsx(Vector3Input, { vector: handle.rotation, label: "Rot" })] })), _jsx(SectionBreak, {})] })), connected &&
                (settingsOpen ? (_jsx(ControlsMapper, { handedness: handedness, actions: actions })) : (transformedConfig.map((buttonConfig) => {
                    const mapping = keyMap[handedness];
                    if (buttonConfig.hasAxes) {
                        return (_jsx(Joystick, { xrController: controller, pointerLocked: pointerLocked, buttonId: buttonConfig.id, mappedKeyUp: mapping[`${buttonConfig.id}-up`], mappedKeyDown: mapping[`${buttonConfig.id}-down`], mappedKeyLeft: mapping[`${buttonConfig.id}-left`], mappedKeyRight: mapping[`${buttonConfig.id}-right`], mappedKeyPressed: mapping[buttonConfig.id] }, buttonConfig.id));
                    }
                    return (_jsx(ScrubControl, { label: labelFor(buttonConfig.id), analog: buttonConfig.type === 'analog', supportsTouch: true, mappedKey: mapping[buttonConfig.id], pointerLocked: pointerLocked, onValue: (v) => controller.updateButtonValue(buttonConfig.id, v), onTouch: (t) => controller.updateButtonTouch(buttonConfig.id, t) }, buttonConfig.id));
                })))] }, handedness));
};
