// @iwer/devui/lib/components/hand.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { ControlPanel, Layout, PanelHeaderButton, SectionBreak, } from './styled.js';
import { ControlsMapper, useKeyMapStore } from './mapper.js';
import { Icon } from './icon.js';
import { PoseSelector } from './pose.js';
import React from 'react';
import { ScrubControl } from './scrub.js';
import { Vector3Input } from './vec3.js';
export const HandUI = ({ hand, handle, handedness, pointerLocked, }) => {
    const { keyMap } = useKeyMapStore();
    const [connected, setConnected] = React.useState(hand.connected);
    const [settingsOpen, setSettingsOpen] = React.useState(false);
    React.useEffect(() => {
        if (pointerLocked)
            setSettingsOpen(false);
    }, [pointerLocked]);
    return (_jsxs(ControlPanel, { style: handedness === 'left'
            ? { left: '8px', bottom: '8px' }
            : { right: '8px', bottom: '8px' }, children: [!pointerLocked && (_jsxs("div", { style: {
                    display: 'flex',
                    flexDirection: 'row',
                    justifyContent: 'space-between',
                    alignItems: 'center',
                }, children: [_jsxs("div", { style: {
                            fontSize: '14px',
                            fontWeight: 600,
                            display: 'flex',
                            flexDirection: 'row',
                            alignItems: 'center',
                            gap: '6px',
                        }, children: [_jsx(Icon, { name: "hand", size: 15 }), "Hand [", handedness === 'left' ? 'L' : 'R', "]"] }), _jsx("div", { style: { display: 'flex', flexDirection: 'row', gap: '1px' }, children: connected ? (_jsxs(_Fragment, { children: [_jsx(PanelHeaderButton, { title: `Click to ${settingsOpen ? 'close' : 'change'} key bindings`, onClick: () => setSettingsOpen(!settingsOpen), children: _jsx(Icon, { name: "settings", size: 14 }) }), _jsx(PanelHeaderButton, { title: `Click to disconnect ${handedness} hand`, "$isRed": true, onClick: () => {
                                        hand.connected = false;
                                        setConnected(false);
                                    }, children: _jsx(Icon, { name: "circle-x", size: 14 }) })] })) : (_jsx(PanelHeaderButton, { title: `Click to reconnect ${handedness} hand`, onClick: () => {
                                hand.connected = true;
                                setConnected(true);
                            }, style: { marginLeft: '5px' }, children: _jsx(Icon, { name: "plug", size: 14 }) })) })] })), connected && !pointerLocked && (_jsxs(_Fragment, { children: [!settingsOpen && (_jsxs(_Fragment, { children: [_jsx(SectionBreak, {}), _jsx(Vector3Input, { vector: handle.position, label: "Pos", marginBottom: Layout.gap }), _jsx(Vector3Input, { vector: handle.rotation, label: "Rot" })] })), _jsx(SectionBreak, {})] })), connected &&
                (settingsOpen ? (_jsx(ControlsMapper, { handedness: handedness, actions: ['pose', 'pinch'] })) : (_jsxs(_Fragment, { children: [_jsx(PoseSelector, { hand: hand, pointerLocked: pointerLocked, mappedKey: keyMap[handedness].pose }), _jsx(ScrubControl, { label: "Pinch", analog: true, supportsTouch: false, mappedKey: keyMap[handedness].pinch, pointerLocked: pointerLocked, onValue: (v) => hand.updatePinchValue(v) })] })))] }, handedness));
};
