// @iwer/devui/lib/components/header.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Button, Colors, ControlButtonStyles, ControlPanel, HeaderButton, HeaderButtonsContainer, PanelHeaderButton, SectionBreak, } from './styled.js';
import { Icon } from './icon.js';
import React from 'react';
import { Vector3Input } from './vec3.js';
import { create } from 'zustand';
import { emitPrefsPatch, getEnvironmentLoader } from '../prefs.js';
import { styled } from 'styled-components';
import { useInputModeStore } from './controls.js';
const VersionTableCol1 = styled.td `
  text-align: right;
  color: ${Colors.textWhite};
  padding: 0 8px 0 0;
  font-weight: bold;
`;
const VersionTableCol2 = styled.td `
  text-align: left;
  color: ${Colors.textGrey};
  padding: 0;
`;
const envNames = [
    'meeting_room',
    'living_room',
    'music_room',
    'office_large',
    'office_small',
    'none',
];
export const useHeaderStateStore = create((set) => ({
    infoPanelOpen: false,
    envDropDownOpen: false,
    headsetOpen: false,
    setInfoPanelOpen: (open) => set(() => ({ infoPanelOpen: open })),
    setEnvDropDownOpen: (open) => set(() => ({ envDropDownOpen: open })),
    setHeadsetOpen: (open) => set(() => ({ headsetOpen: open })),
}));
function underscoreToTitleCase(str) {
    return str
        .split('_')
        .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
        .join(' ');
}
export const HeaderUI = ({ xrDevice, inputLayer }) => {
    var _a, _b, _c;
    const [planesVisible, setPlanesVisible] = React.useState(Boolean((_a = xrDevice.sem) === null || _a === void 0 ? void 0 : _a.planesVisible));
    const [boxesVisible, setBoxesVisible] = React.useState(Boolean((_b = xrDevice.sem) === null || _b === void 0 ? void 0 : _b.boundingBoxesVisible));
    const [meshesVisible, setMeshesVisible] = React.useState(Boolean((_c = xrDevice.sem) === null || _c === void 0 ? void 0 : _c.meshesVisible));
    const { inputMode, setInputMode } = useInputModeStore();
    const { infoPanelOpen, setInfoPanelOpen, envDropDownOpen, setEnvDropDownOpen, headsetOpen, setHeadsetOpen, } = useHeaderStateStore();
    const semColor = (on) => (on ? Colors.accent : Colors.textGrey);
    return (_jsxs("div", { style: {
            display: 'flex',
            justifyContent: 'center',
            flexDirection: 'row',
            alignItems: 'center',
            gap: '6px',
            padding: '8px',
        }, children: [_jsxs("div", { style: { position: 'relative' }, children: [_jsxs(HeaderButtonsContainer, { children: [_jsx(HeaderButton, { title: "Reset device transforms", onClick: () => inputLayer.resetDeviceTransforms(), children: _jsx(Icon, { name: "rotate-ccw", size: 16 }) }), _jsx(HeaderButton, { title: "Play mode (lock pointer)", onClick: () => {
                                    inputLayer.lockPointer();
                                    setEnvDropDownOpen(false);
                                    setInfoPanelOpen(false);
                                }, children: _jsx(Icon, { name: "circle-play", size: 16 }) }), _jsx(HeaderButton, { title: "Toggle input mode", onClick: () => {
                                    const next = inputMode === 'controller' ? 'hand' : 'controller';
                                    setInputMode(next);
                                    xrDevice.primaryInputMode = next;
                                    emitPrefsPatch({ inputMode: next });
                                }, children: _jsx(Icon, { name: inputMode === 'controller' ? 'gamepad-2' : 'hand', size: 16 }) }), _jsx(HeaderButton, { title: "Headset position", onClick: () => {
                                    setHeadsetOpen(!headsetOpen);
                                    setEnvDropDownOpen(false);
                                    setInfoPanelOpen(false);
                                }, children: _jsx(Icon, { name: "glasses", size: 16, color: headsetOpen ? Colors.accent : Colors.textWhite }) }), xrDevice.sem && (_jsxs(_Fragment, { children: [_jsx(SectionBreak, { "$horizontal": false }), _jsx(HeaderButton, { title: "Select emulated environment", onClick: () => setEnvDropDownOpen(!envDropDownOpen), children: _jsx(Icon, { name: "armchair", size: 16 }) }), _jsx(HeaderButton, { title: "Toggle plane visibility", onClick: () => {
                                            xrDevice.sem.planesVisible = !planesVisible;
                                            setPlanesVisible(!planesVisible);
                                        }, children: _jsx(Icon, { name: "layers", size: 16, color: semColor(planesVisible) }) }), _jsx(HeaderButton, { title: "Toggle bounding-box visibility", onClick: () => {
                                            xrDevice.sem.boundingBoxesVisible = !boxesVisible;
                                            setBoxesVisible(!boxesVisible);
                                        }, children: _jsx(Icon, { name: "box", size: 16, color: semColor(boxesVisible) }) }), _jsx(HeaderButton, { title: "Toggle mesh visibility", onClick: () => {
                                            xrDevice.sem.meshesVisible = !meshesVisible;
                                            setMeshesVisible(!meshesVisible);
                                        }, children: _jsx(Icon, { name: "shapes", size: 16, color: semColor(meshesVisible) }) })] })), _jsx(SectionBreak, { "$horizontal": false }), _jsx(HeaderButton, { title: "Save current device pose as the default for this site", onClick: () => {
                                    const defaultPose = inputLayer.captureDefaultPose();
                                    inputLayer.applyDefaultPose(defaultPose);
                                    emitPrefsPatch({ defaultPose });
                                }, children: _jsx(Icon, { name: "save", size: 16 }) }), _jsx(SectionBreak, { "$horizontal": false }), _jsx(HeaderButton, { title: "Exit XR session", onClick: () => { var _a; return (_a = xrDevice.activeSession) === null || _a === void 0 ? void 0 : _a.end(); }, children: _jsx(Icon, { name: "log-out", size: 16 }) })] }), headsetOpen && (_jsxs(ControlPanel, { style: {
                            position: 'absolute',
                            top: 'calc(100% + 6px)',
                            left: 0,
                            gap: '6px',
                            padding: '10px',
                        }, children: [_jsxs("div", { style: {
                                    display: 'flex',
                                    alignItems: 'center',
                                    gap: '6px',
                                    fontSize: '13px',
                                    fontWeight: 600,
                                }, children: [_jsx(Icon, { name: "glasses", size: 15 }), xrDevice.name] }), _jsx(SectionBreak, {}), _jsx(Vector3Input, { vector: inputLayer.combinedCameraPosition, label: "Pos" })] })), envDropDownOpen && (_jsx(ControlPanel, { style: {
                            position: 'absolute',
                            top: 'calc(100% + 6px)',
                            left: 0,
                            gap: '2px',
                        }, children: envNames.map((name) => (_jsx(HeaderButton, { style: {
                                fontSize: '12px',
                                width: '100%',
                                justifyContent: 'start',
                                borderRadius: '8px',
                                padding: '6px 10px',
                            }, onClick: () => {
                                const loadEnvironment = getEnvironmentLoader();
                                if (loadEnvironment) {
                                    loadEnvironment(name);
                                }
                                else if (name === 'none') {
                                    xrDevice.sem.deleteAll();
                                }
                                else {
                                    // @ts-ignore
                                    xrDevice.sem.loadDefaultEnvironment(name);
                                }
                                emitPrefsPatch({ environment: name });
                            }, children: underscoreToTitleCase(name) }, name))) }))] }), _jsxs(HeaderButtonsContainer, { children: [_jsx(HeaderButton, { title: "About IWER", onClick: () => setInfoPanelOpen(!infoPanelOpen), children: _jsx(Icon, { name: "iwer", size: 16 }) }), _jsx(HeaderButton, { title: "Report issues", onClick: () => window.open('https://github.com/meta-quest/immersive-web-emulation-runtime/issues', '_blank'), children: _jsx(Icon, { name: "bug", size: 16 }) })] }), infoPanelOpen && (_jsxs(ControlPanel, { style: {
                    top: '50vh',
                    left: '50vw',
                    transform: 'translate(-50%, -50%)',
                    maxWidth: '240px',
                    gap: '4px',
                    padding: '16px',
                }, children: [_jsx("div", { style: { display: 'flex', justifyContent: 'end' }, children: _jsx(PanelHeaderButton, { "$isRed": true, onClick: () => setInfoPanelOpen(false), children: _jsx(Icon, { name: "circle-x", size: 14 }) }) }), _jsx("div", { style: { display: 'flex', justifyContent: 'center' }, children: _jsx(Icon, { name: "iwer", size: 72 }) }), _jsxs("p", { style: { textAlign: 'center', padding: '0 5px', margin: '0' }, children: [_jsx("b", { children: "Immersive Web Emulation Runtime" }), " (IWER) is a free, open-source WebXR developer tool created by Meta Platforms, Inc."] }), _jsx("table", { style: {
                            width: '100%',
                            borderCollapse: 'collapse',
                            display: 'flex',
                            justifyContent: 'center',
                            fontSize: '12px',
                            padding: '8px',
                        }, children: _jsxs("tbody", { children: [_jsxs("tr", { children: [_jsx(VersionTableCol1, { children: "IWER" }), _jsxs(VersionTableCol2, { children: ["v", xrDevice.version] })] }), _jsxs("tr", { children: [_jsx(VersionTableCol1, { children: "DevUI" }), _jsxs(VersionTableCol2, { children: ["v", xrDevice.devui.version] })] }), xrDevice.sem && (_jsxs("tr", { children: [_jsx(VersionTableCol1, { children: "SEM" }), _jsxs(VersionTableCol2, { children: ["v", xrDevice.sem.version] })] }))] }) }), _jsx(Button, { style: { borderRadius: ControlButtonStyles.radiusSolo }, onClick: () => window.open('https://github.com/meta-quest/immersive-web-emulation-runtime/blob/main/LICENSE', '_blank'), children: "MIT License" }), _jsx(Button, { style: { borderRadius: ControlButtonStyles.radiusSolo }, onClick: () => window.open('https://github.com/meta-quest/immersive-web-emulation-runtime', '_blank'), children: "View Source on GitHub" })] }))] }));
};
