// @iwer/devui/lib/components/keys.js
import { jsx as _jsx } from "react/jsx-runtime";
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Icon } from './icon.js';
export const MappedKeyDisplay = {
    KeyA: 'A',
    KeyB: 'B',
    KeyC: 'C',
    KeyD: 'D',
    KeyE: 'E',
    KeyF: 'F',
    KeyG: 'G',
    KeyH: 'H',
    KeyI: 'I',
    KeyJ: 'J',
    KeyK: 'K',
    KeyL: 'L',
    KeyM: 'M',
    KeyN: 'N',
    KeyO: 'O',
    KeyP: 'P',
    KeyQ: 'Q',
    KeyR: 'R',
    KeyS: 'S',
    KeyT: 'T',
    KeyU: 'U',
    KeyV: 'V',
    KeyW: 'W',
    KeyX: 'X',
    KeyY: 'Y',
    KeyZ: 'Z',
    Digit0: '0',
    Digit1: '1',
    Digit2: '2',
    Digit3: '3',
    Digit4: '4',
    Digit5: '5',
    Digit6: '6',
    Digit7: '7',
    Digit8: '8',
    Digit9: '9',
    Tab: _jsx(Icon, { name: "arrow-right-to-line", size: 15 }),
    Backspace: _jsx(Icon, { name: "delete", size: 15 }),
    Enter: _jsx(Icon, { name: "corner-down-left", size: 15 }),
    ShiftLeft: _jsx(Icon, { name: "arrow-big-up", size: 15 }),
    ShiftRight: _jsx(Icon, { name: "arrow-big-up", size: 15 }),
    Space: ' ',
    ArrowUp: _jsx(Icon, { name: "arrow-up", size: 15 }),
    ArrowDown: _jsx(Icon, { name: "arrow-down", size: 15 }),
    ArrowLeft: _jsx(Icon, { name: "arrow-left", size: 15 }),
    ArrowRight: _jsx(Icon, { name: "arrow-right", size: 15 }),
    Semicolon: ';',
    Equal: '=',
    Comma: ',',
    Minus: '-',
    Period: '.',
    Slash: '/',
    Backquote: '`',
    BracketLeft: '[',
    Backslash: '\\',
    BracketRight: ']',
    Quote: "'",
    MouseLeft: _jsx(Icon, { name: "mouse-left", size: 15 }),
    MouseRight: _jsx(Icon, { name: "mouse-right", size: 15 }),
};
