// @iwer/devui/lib/components/pose.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { CollapsedRow, Colors, GlyphCircle, Layout, MappedKeyBlock, RowLabel, } from './styled.js';
import { useEffect, useState } from 'react';
import { Icon } from './icon.js';
import { MappedKeyDisplay } from './keys.js';
import { emitPrefsPatch } from '../prefs.js';
import { styled } from 'styled-components';
const poses = ['default', 'pinch', 'point'];
const Row = styled.div `
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 0;
  height: ${Layout.rowH};
  margin-bottom: ${Layout.gap};
`;
// Fills the space after the label exactly like the scrub Cell, so the left
// chevron lines up with the pinch field's left edge (and the Pos/Rot column)
// and the right chevron lines up with the pinch pin column on the right.
const Stepper = styled.div `
  flex: 1 1 auto;
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: ${Layout.gap};
  height: ${Layout.rowH};
`;
const Step = styled.button `
  flex: none;
  width: 22px;
  height: ${Layout.rowH};
  display: flex;
  align-items: center;
  justify-content: center;
  border: none;
  border-radius: 6px;
  cursor: pointer;
  background-color: ${Layout.inset};
  color: ${Colors.textGrey};

  &:hover {
    color: ${Colors.textWhite};
  }
  &:focus {
    outline: none;
  }
`;
const Value = styled.div `
  flex: 1 1 auto;
  height: ${Layout.rowH};
  border-radius: ${Layout.cellRadius};
  background-color: ${Layout.inset};
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 11px;
  color: ${Colors.textWhite};
`;
export const PoseSelector = ({ hand, pointerLocked, mappedKey, }) => {
    const [poseId, setPoseId] = useState(hand.poseId);
    const [isKeyPressed, setIsKeyPressed] = useState(false);
    const handedness = hand.inputSource.handedness;
    const cyclePose = (delta) => {
        const idx = poses.indexOf(hand.poseId);
        const next = (idx + poses.length + delta) % poses.length;
        setPoseId(poses[next]);
        hand.poseId = poses[next];
        emitPrefsPatch({ handPoses: { [handedness]: poses[next] } });
    };
    useEffect(() => {
        // The pose-cycle binding can be a keyboard code or a mouse button.
        const isMouseMatch = (event) => (mappedKey === 'MouseLeft' && event.button === 0) ||
            (mappedKey === 'MouseRight' && event.button === 2);
        const down = (event) => {
            if (event.repeat)
                return;
            if (event.code === mappedKey) {
                cyclePose(1);
                setIsKeyPressed(true);
            }
        };
        const up = (event) => {
            if (event.code === mappedKey)
                setIsKeyPressed(false);
        };
        const mouseDown = (event) => {
            if (isMouseMatch(event)) {
                cyclePose(1);
                setIsKeyPressed(true);
            }
        };
        const mouseUp = (event) => {
            if (isMouseMatch(event))
                setIsKeyPressed(false);
        };
        if (pointerLocked) {
            window.addEventListener('keydown', down);
            window.addEventListener('keyup', up);
            window.addEventListener('mousedown', mouseDown);
            window.addEventListener('mouseup', mouseUp);
        }
        return () => {
            window.removeEventListener('keydown', down);
            window.removeEventListener('keyup', up);
            window.removeEventListener('mousedown', mouseDown);
            window.removeEventListener('mouseup', mouseUp);
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [mappedKey, pointerLocked, hand]);
    if (pointerLocked) {
        return (_jsxs(CollapsedRow, { children: [_jsx(GlyphCircle, { children: "Pose" }), _jsx(MappedKeyBlock, { "$pressed": isKeyPressed, children: MappedKeyDisplay[mappedKey] })] }));
    }
    return (_jsxs(Row, { children: [_jsx(RowLabel, { children: "Pose" }), _jsxs(Stepper, { children: [_jsx(Step, { type: "button", onClick: () => cyclePose(-1), title: "Previous pose", children: _jsx(Icon, { name: "chevron-left", size: 14 }) }), _jsx(Value, { children: poseId }), _jsx(Step, { type: "button", onClick: () => cyclePose(1), title: "Next pose", children: _jsx(Icon, { name: "chevron-right", size: 14 }) })] })] }));
};
