// iwer/lib/gamepad/Gamepad.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { P_GAMEPAD } from '../private.js';
export var GamepadMappingType;
(function (GamepadMappingType) {
    GamepadMappingType["None"] = "";
    GamepadMappingType["Standard"] = "standard";
    GamepadMappingType["XRStandard"] = "xr-standard";
})(GamepadMappingType || (GamepadMappingType = {}));
export class GamepadButton {
    constructor(type, eventTrigger) {
        this[P_GAMEPAD] = {
            type,
            eventTrigger,
            pressed: false,
            touched: false,
            value: 0,
            lastFrameValue: 0,
            pendingValue: null,
        };
    }
    get pressed() {
        if (this[P_GAMEPAD].type === 'manual') {
            return this[P_GAMEPAD].pressed;
        }
        else {
            return this[P_GAMEPAD].value > 0;
        }
    }
    get touched() {
        if (this[P_GAMEPAD].type === 'manual') {
            return this[P_GAMEPAD].touched;
        }
        else {
            return this[P_GAMEPAD].touched || this.pressed;
        }
    }
    get value() {
        return this[P_GAMEPAD].value;
    }
}
export class EmptyGamepadButton {
    constructor() {
        this.pressed = false;
        this.touched = false;
        this.value = 0;
    }
}
export class GamepadHapticActuator {
    constructor(type = 'vibration') {
        this[P_GAMEPAD] = {
            type,
            lastPulse: null,
        };
    }
    get type() {
        return this[P_GAMEPAD].type;
    }
    pulse(value, duration) {
        this[P_GAMEPAD].lastPulse = {
            value,
            duration,
            startTime: performance.now(),
        };
        return Promise.resolve(true);
    }
}
export class Gamepad {
    constructor(gamepadConfig, id = '', index = -1, numHapticActuators = 0) {
        this[P_GAMEPAD] = {
            id,
            index,
            connected: false,
            timestamp: performance.now(),
            mapping: gamepadConfig.mapping,
            buttonsMap: {},
            buttonsSequence: [],
            axesMap: {},
            axesSequence: [],
            hapticActuators: Array.from({ length: Math.max(0, numHapticActuators) }, () => new GamepadHapticActuator('vibration')),
            buttonsView: [],
            axesView: [],
            axesResolvers: [],
        };
        gamepadConfig.buttons.forEach((buttonConfig) => {
            var _a;
            if (buttonConfig === null) {
                this[P_GAMEPAD].buttonsSequence.push(null);
            }
            else {
                this[P_GAMEPAD].buttonsSequence.push(buttonConfig.id);
                this[P_GAMEPAD].buttonsMap[buttonConfig.id] = new GamepadButton(buttonConfig.type, (_a = buttonConfig.eventTrigger) !== null && _a !== void 0 ? _a : null);
            }
        });
        gamepadConfig.axes.forEach((axisConfig) => {
            if (axisConfig === null) {
                this[P_GAMEPAD].axesSequence.push(null);
            }
            else {
                this[P_GAMEPAD].axesSequence.push(axisConfig.id + axisConfig.type);
                if (!this[P_GAMEPAD].axesMap[axisConfig.id]) {
                    this[P_GAMEPAD].axesMap[axisConfig.id] = { x: 0, y: 0 };
                }
            }
        });
        // Build the cached buttons array once. The GamepadButton instances are
        // stable references mutated in place, so the cached array always reflects
        // current state; null slots get a shared empty button.
        this[P_GAMEPAD].buttonsView = this[P_GAMEPAD].buttonsSequence.map((id) => id === null ? new EmptyGamepadButton() : this[P_GAMEPAD].buttonsMap[id]);
        // Precompute axis resolvers and a reusable axes array (values written in
        // place on each read).
        this[P_GAMEPAD].axesResolvers = this[P_GAMEPAD].axesSequence.map((id) => id === null
            ? null
            : {
                axisId: id.substring(0, id.length - 6),
                isY: id.substring(id.length - 6) === 'y-axis',
            });
        this[P_GAMEPAD].axesView = this[P_GAMEPAD].axesResolvers.map(() => null);
    }
    get id() {
        return this[P_GAMEPAD].id;
    }
    get index() {
        return this[P_GAMEPAD].index;
    }
    get connected() {
        return this[P_GAMEPAD].connected;
    }
    get timestamp() {
        return this[P_GAMEPAD].timestamp;
    }
    get mapping() {
        return this[P_GAMEPAD].mapping;
    }
    get axes() {
        const { axesView, axesResolvers, axesMap } = this[P_GAMEPAD];
        for (let i = 0; i < axesResolvers.length; i++) {
            const resolver = axesResolvers[i];
            if (resolver === null) {
                axesView[i] = null;
            }
            else {
                const axis = axesMap[resolver.axisId];
                // manual axes fall through to the x value, matching the prior behavior
                axesView[i] = resolver.isY ? axis.y : axis.x;
            }
        }
        return axesView;
    }
    get buttons() {
        return this[P_GAMEPAD].buttonsView;
    }
    get hapticActuators() {
        return this[P_GAMEPAD].hapticActuators;
    }
    get vibrationActuator() {
        var _a;
        return (_a = this[P_GAMEPAD].hapticActuators[0]) !== null && _a !== void 0 ? _a : null;
    }
}
