// @iwsdk/core/dist/init/world-initializer.js
import { XRInputManager } from '@iwsdk/xr-input';
import { signal } from '@preact/signals-core';
import { AssetManager } from '../asset/asset-manager.js';
import '../asset/cache-manager.js';
import { AudioSource } from '../audio/audio.js';
import { AudioSystem } from '../audio/audio-system.js';
import '../runtime/three.js';
import '../camera/camera-utils.js';
import { CameraSource } from '../camera/camera-source.js';
import { CameraSystem } from '../camera/camera-system.js';
import { World, VisibilityState } from '../ecs/world.js';
import { DomeTexture } from '../environment/dome-texture.js';
import { DomeGradient } from '../environment/dome-gradient.js';
import { IBLTexture } from '../environment/ibl-texture.js';
import { IBLGradient } from '../environment/ibl-gradient.js';
import { EnvironmentSystem } from '../environment/environment-system.js';
import { EnvironmentRaycastSystem } from '../environment-raycast/environment-raycast.js';
import { EnvironmentRaycastTarget } from '../environment-raycast/raycast-target.js';
import '../grab/index.js';
import { InputSystem } from '../input/input-system.js';
import { InputManager } from '../input/input-manager.js';
import '../input/stateful-browser-gamepad.js';
import { CanvasPointerSystem } from '../input/canvas-pointer-system.js';
import { RayInteractable, PokeInteractable, Hovered, Pressed } from '../input/state-tags.js';
import { XRQuadLayer } from '../layers/xr-quad-layer.js';
import { XRCylinderLayer } from '../layers/xr-cylinder-layer.js';
import { XRLayerState, XRLayerSystem } from '../layers/xr-layer-system.js';
import { LevelSystem } from '../level/level-system.js';
import { LevelTag } from '../level/level-tag.js';
import '@iwsdk/glxf';
import 'elics';
import '@pmndrs/uikit';
import { PerspectiveCamera, WebGLRenderer, SRGBColorSpace, Scene, Clock } from 'three';
import { PanelUI, PanelUISystem } from '../ui/ui.js';
import { Follower, FollowSystem } from '../ui/follow.js';
import { ScreenSpace, ScreenSpaceUISystem } from '../ui/screenspace.js';
import { LevelRoot } from '../level/level-root.js';
import '../locomotion/slide.js';
import '../locomotion/teleport.js';
import '../locomotion/turn.js';
import { LocomotionSystem } from '../locomotion/locomotion.js';
import { MCPRuntime } from '../mcp/mcp-runtime.js';
import { SceneUnderstandingSystem } from '../scene-understanding/scene-understanding-system.js';
import { XRAnchor } from '../scene-understanding/anchor.js';
import { XRPlane } from '../scene-understanding/plane.js';
import { XRMesh } from '../scene-understanding/mesh.js';
import { Transform, TransformSystem } from '../transform/transform.js';
import '../transform/transform-utils.js';
import { Visibility, VisibilitySystem } from '../visibility/visibility.js';
import { attachBrowserCameraRestore } from './browser-camera.js';
import { attachCameraToPlayer } from './player-camera.js';
import { ReferenceSpaceType, SessionMode, buildSessionInit, normalizeReferenceSpec, resolveReferenceSpaceType } from './xr.js';
import { GrabSystem } from '../grab/grab-system.js';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
/**
 * Initialize a new WebXR world with all required systems and setup
 *
 * @param sceneContainer - HTML container for the renderer canvas
 * @param assets - Asset manifest for preloading
 * @param options - Configuration options for the world
 * @returns Promise that resolves to the initialized World instance
 */
/**
 * Initialize a new WebXR world with all required systems and setup.
 *
 * @param sceneContainer HTML container for the renderer canvas.
 * @param options Configuration options for the world.
 * @returns Promise that resolves to the initialized {@link World} instance.
 *
 * @remarks
 * This function powers {@link World.create}. Prefer using that static helper.
 */
async function initializeWorld(container, options = {}) {
    var _a;
    // Create and configure world instance
    const world = createWorldInstance();
    // Extract configuration options
    const config = extractConfiguration(options);
    // Setup core rendering components
    const { camera, renderer, scene } = setupRendering(container, config);
    assignRenderingToWorld(world, camera, renderer, scene);
    // Setup input management
    setupInputManagement(world, config.input);
    // Store XR defaults for later explicit launch/offer calls
    world.xrDefaults = {
        sessionMode: config.xr.sessionMode,
        referenceSpace: config.xr.referenceSpace,
        features: config.xr.features,
        restoreCameraOnExit: config.xr.restoreCameraOnExit,
    };
    // Register core systems (LevelSystem receives defaultLighting)
    registerCoreSystems(world, config);
    // Initialize asset manager
    initializeAssetManager(renderer, world);
    // Register additional systems (UI + Audio on by default)
    registerAdditionalSystems(world);
    // Register input and feature systems with explicit priorities.
    // Awaited because the physics branch dynamically imports its module so
    // `@babylonjs/havok` stays out of the static graph for non-physics builds.
    await registerFeatureSystems(world, config);
    // Setup render loop
    setupRenderLoop(world, renderer);
    // Setup resize handling
    setupResizeHandling(camera, renderer);
    // Manage XR offer flow if configured
    if (config.xr.offer && config.xr.offer !== 'none') {
        manageOfferFlow(world, config.xr.offer);
    }
    // Setup MCP runtime for framework-specific tools (dev only).
    // In production Vite builds, import.meta.env.DEV is false and this entire
    // code path is tree-shaken. In Node.js/tests, import.meta.env is undefined
    // so the check defaults to enabled.
    if (((_a = import.meta.env) === null || _a === void 0 ? void 0 : _a.DEV) !== false) {
        setupMCPRuntime(world);
    }
    // Return promise that resolves after asset preloading
    return finalizeInitialization(world, options.assets).then(async (w) => {
        var _a;
        // Load initial level or create empty level
        const levelUrl = typeof options.level === 'string' ? options.level : (_a = options.level) === null || _a === void 0 ? void 0 : _a.url;
        if (levelUrl) {
            await w.loadLevel(levelUrl);
        }
        else {
            await w.loadLevel();
        }
        return w;
    });
}
/**
 * Create a new World instance with basic ECS setup
 */
function createWorldInstance() {
    const world = new World();
    world
        .registerComponent(Transform)
        .registerComponent(Visibility)
        .registerComponent(LevelTag)
        .registerSystem(TransformSystem)
        .registerSystem(VisibilitySystem);
    return world;
}
/**
 * Extract and normalize configuration options
 */
function extractConfiguration(options) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7;
    const xrOptions = options.xr === false ? undefined : options.xr;
    const spatialUI = (_a = options.features) === null || _a === void 0 ? void 0 : _a.spatialUI;
    const legacyForwardHtmlEvents = typeof spatialUI === 'object' && spatialUI
        ? spatialUI.forwardHtmlEvents
        : undefined;
    const canvasPointerEvents = (_c = (_b = options.input) === null || _b === void 0 ? void 0 : _b.canvasPointerEvents) !== null && _c !== void 0 ? _c : legacyForwardHtmlEvents;
    return {
        cameraFov: (_e = (_d = options.render) === null || _d === void 0 ? void 0 : _d.fov) !== null && _e !== void 0 ? _e : 50,
        cameraNear: (_g = (_f = options.render) === null || _f === void 0 ? void 0 : _f.near) !== null && _g !== void 0 ? _g : 0.1,
        cameraFar: (_j = (_h = options.render) === null || _h === void 0 ? void 0 : _h.far) !== null && _j !== void 0 ? _j : 200,
        cameraPose: (_k = options.render) === null || _k === void 0 ? void 0 : _k.camera,
        defaultLighting: (_m = (_l = options.render) === null || _l === void 0 ? void 0 : _l.defaultLighting) !== null && _m !== void 0 ? _m : true,
        stencil: (_p = (_o = options.render) === null || _o === void 0 ? void 0 : _o.stencil) !== null && _p !== void 0 ? _p : false,
        xr: {
            enabled: options.xr !== false,
            sessionMode: (_q = xrOptions === null || xrOptions === void 0 ? void 0 : xrOptions.sessionMode) !== null && _q !== void 0 ? _q : SessionMode.ImmersiveVR,
            referenceSpace: (_r = xrOptions === null || xrOptions === void 0 ? void 0 : xrOptions.referenceSpace) !== null && _r !== void 0 ? _r : ReferenceSpaceType.LocalFloor,
            features: xrOptions === null || xrOptions === void 0 ? void 0 : xrOptions.features,
            offer: options.xr === false ? 'none' : ((_s = xrOptions === null || xrOptions === void 0 ? void 0 : xrOptions.offer) !== null && _s !== void 0 ? _s : 'always'),
            restoreCameraOnExit: (_t = xrOptions === null || xrOptions === void 0 ? void 0 : xrOptions.restoreCameraOnExit) !== null && _t !== void 0 ? _t : true,
        },
        input: {
            canvasPointerEvents,
        },
        features: {
            locomotion: (_v = (_u = options.features) === null || _u === void 0 ? void 0 : _u.locomotion) !== null && _v !== void 0 ? _v : false,
            grabbing: (_x = (_w = options.features) === null || _w === void 0 ? void 0 : _w.grabbing) !== null && _x !== void 0 ? _x : false,
            physics: (_z = (_y = options.features) === null || _y === void 0 ? void 0 : _y.physics) !== null && _z !== void 0 ? _z : false,
            sceneUnderstanding: (_1 = (_0 = options.features) === null || _0 === void 0 ? void 0 : _0.sceneUnderstanding) !== null && _1 !== void 0 ? _1 : false,
            environmentRaycast: (_3 = (_2 = options.features) === null || _2 === void 0 ? void 0 : _2.environmentRaycast) !== null && _3 !== void 0 ? _3 : false,
            camera: (_5 = (_4 = options.features) === null || _4 === void 0 ? void 0 : _4.camera) !== null && _5 !== void 0 ? _5 : false,
            spatialUI: (_7 = (_6 = options.features) === null || _6 === void 0 ? void 0 : _6.spatialUI) !== null && _7 !== void 0 ? _7 : true,
        },
    };
}
/**
 * Setup camera, renderer, and scene
 */
function setupRendering(sceneContainer, config) {
    // Camera Setup
    const camera = new PerspectiveCamera(config.cameraFov, window.innerWidth / window.innerHeight, config.cameraNear, config.cameraFar);
    camera.position.set(0, 1.7, 0);
    if (config.cameraPose) {
        applyCameraPose(camera, config.cameraPose);
    }
    // Renderer Setup
    const renderer = new WebGLRenderer({
        antialias: true,
        alpha: config.xr.enabled && config.xr.sessionMode === SessionMode.ImmersiveAR,
        // @ts-ignore
        multiviewStereo: true,
        stencil: config.stencil,
    });
    renderer.setPixelRatio(window.devicePixelRatio);
    renderer.setSize(window.innerWidth, window.innerHeight);
    renderer.outputColorSpace = SRGBColorSpace;
    renderer.xr.enabled = config.xr.enabled;
    sceneContainer.appendChild(renderer.domElement);
    // Scene Setup
    const scene = new Scene();
    return { camera, renderer, scene };
}
function applyCameraPose(camera, pose) {
    if (pose.position) {
        camera.position.fromArray(pose.position);
    }
    if (pose.rotation) {
        camera.rotation.fromArray(pose.rotation);
    }
    if (pose.quaternion) {
        camera.quaternion.fromArray(pose.quaternion);
    }
    if (pose.lookAt) {
        camera.lookAt(...pose.lookAt);
    }
}
/**
 * Assign rendering components to world instance
 */
function assignRenderingToWorld(world, camera, renderer, scene) {
    world.scene = scene;
    world.camera = camera;
    world.renderer = renderer;
    // Scene entity (wrap Scene in an entity for parenting convenience)
    world.sceneEntity = world.createTransformEntity(scene);
    // Create a default level root so activeLevel is always defined
    const levelRootEntity = world.createTransformEntity(undefined, {
        parent: world.sceneEntity,
    });
    levelRootEntity.object3D.name = 'LevelRoot';
    // @ts-ignore init signal now; LevelSystem will enforce identity each frame
    world.activeLevel = signal(levelRootEntity);
}
/**
 * Setup default lighting environment using Unity-style gradient ambient lighting
 */
// default lighting is attached per level by LevelSystem
/**
 * Setup XR input management
 */
function setupInputManagement(world, config) {
    const xrInputManager = new XRInputManager({
        camera: world.camera,
        scene: world.scene,
        assetLoader: AssetManager,
    });
    world.scene.add(xrInputManager.xrOrigin);
    attachCameraToPlayer(xrInputManager.xrOrigin, world.camera);
    world.player = xrInputManager.xrOrigin;
    world.input = new InputManager(xrInputManager, {
        canvasPointerEvents: config.canvasPointerEvents,
    });
    const xrOrigin = xrInputManager.xrOrigin;
    const playerEntity = world.createTransformEntity(xrOrigin, {
        parent: world.sceneEntity,
        persistent: true,
    });
    world.cameraEntity = world.createTransformEntity(world.camera, {
        parent: playerEntity,
        persistent: true,
    });
    const headEntity = world.createTransformEntity(xrOrigin.head, {
        parent: playerEntity,
        persistent: true,
    });
    const rayLeftEntity = world.createTransformEntity(xrOrigin.raySpaces.left, {
        parent: playerEntity,
        persistent: true,
    });
    const rayRightEntity = world.createTransformEntity(xrOrigin.raySpaces.right, {
        parent: playerEntity,
        persistent: true,
    });
    const gripLeftEntity = world.createTransformEntity(xrOrigin.gripSpaces.left, {
        parent: playerEntity,
        persistent: true,
    });
    const gripRightEntity = world.createTransformEntity(xrOrigin.gripSpaces.right, { parent: playerEntity, persistent: true });
    const indexTipLeftEntity = world.createTransformEntity(xrOrigin.indexTipSpaces.left, { parent: playerEntity, persistent: true });
    const indexTipRightEntity = world.createTransformEntity(xrOrigin.indexTipSpaces.right, { parent: playerEntity, persistent: true });
    world.playerEntity = playerEntity;
    world.playerHeadEntity = headEntity;
    world.playerSpaceEntities = {
        head: headEntity,
        raySpaces: { left: rayLeftEntity, right: rayRightEntity },
        gripSpaces: { left: gripLeftEntity, right: gripRightEntity },
        indexTipSpaces: { left: indexTipLeftEntity, right: indexTipRightEntity },
    };
    return xrInputManager;
}
/**
 * Manage offering XR sessions according to the configured offer policy.
 * - 'once': offer after init; no re-offer on end
 * - 'always': offer after init and re-offer whenever the session ends
 */
function manageOfferFlow(world, mode) {
    let offering = false;
    const offer = async () => {
        var _a, _b, _c, _d;
        if (offering || world.session) {
            return;
        }
        offering = true;
        try {
            const opts = (_a = world.xrDefaults) !== null && _a !== void 0 ? _a : { sessionMode: SessionMode.ImmersiveVR };
            const sessionInit = buildSessionInit(opts);
            const session = await ((_c = (_b = navigator.xr) === null || _b === void 0 ? void 0 : _b.offerSession) === null || _c === void 0 ? void 0 : _c.call(_b, (_d = opts.sessionMode) !== null && _d !== void 0 ? _d : SessionMode.ImmersiveVR, 
            // if the dynamic import failed, rebuild via launchXR path by calling request, but we only want offer
            sessionInit));
            if (!session) {
                return;
            }
            const refSpec = normalizeReferenceSpec(opts.referenceSpace);
            session.addEventListener('end', onEnd);
            try {
                // disable built-in occlusion
                world.renderer.xr.getDepthSensingMesh = function () {
                    return null;
                };
                const resolvedType = await resolveReferenceSpaceType(session, refSpec.type, refSpec.required ? [] : refSpec.fallbackOrder);
                world.renderer.xr.setReferenceSpaceType(resolvedType);
                if (opts.restoreCameraOnExit !== false) {
                    attachBrowserCameraRestore(world.camera, session);
                }
                await world.renderer.xr.setSession(session);
                world.session = session;
            }
            catch (err) {
                console.error('[XR] Failed to acquire reference space:', err);
                try {
                    await session.end();
                }
                catch (_e) { }
            }
        }
        finally {
            offering = false;
        }
    };
    const onEnd = () => {
        var _a;
        (_a = world.session) === null || _a === void 0 ? void 0 : _a.removeEventListener('end', onEnd);
        world.session = undefined;
        if (mode === 'always') {
            // re-offer after session ends
            offer();
        }
    };
    // initial offer once world is ready
    offer();
}
/**
 * Register core interaction systems
 */
function registerCoreSystems(world, config) {
    world
        .registerComponent(RayInteractable)
        .registerComponent(PokeInteractable)
        .registerComponent(Hovered)
        .registerComponent(Pressed)
        .registerComponent(LevelRoot)
        // New split components
        .registerComponent(DomeTexture)
        .registerComponent(DomeGradient)
        .registerComponent(IBLTexture)
        .registerComponent(IBLGradient)
        // Unified environment system (background + IBL)
        .registerSystem(EnvironmentSystem)
        .registerSystem(LevelSystem, {
        configData: { defaultLighting: config.defaultLighting },
    });
}
/**
 * Initialize the asset manager
 */
function initializeAssetManager(renderer, world) {
    AssetManager.init(renderer, world);
}
/**
 * Register optional systems based on configuration
 */
function registerAdditionalSystems(world) {
    // Audio system remains always-on
    world.registerComponent(AudioSource).registerSystem(AudioSystem);
}
async function registerFeatureSystems(world, config) {
    var _a;
    const locomotion = config.features.locomotion;
    const locomotionEnabled = !!locomotion;
    const grabbing = config.features.grabbing;
    const grabbingEnabled = !!grabbing;
    const physicsEnabled = !!config.features.physics;
    const sceneUnderstanding = config.features.sceneUnderstanding;
    const sceneUnderstandingEnabled = !!sceneUnderstanding;
    const environmentRaycastEnabled = !!config.features.environmentRaycast;
    const cameraEnabled = !!config.features.camera;
    const spatialUI = config.features.spatialUI;
    const spatialUIEnabled = !!spatialUI;
    if (locomotionEnabled) {
        const locOpts = typeof locomotion === 'object' && locomotion
            ? Object.fromEntries(Object.entries({
                useWorker: locomotion.useWorker,
                initialPlayerPosition: locomotion.initialPlayerPosition,
                comfortAssist: locomotion.comfortAssistLevel,
                turningMethod: locomotion.turningMethod,
                enableJumping: locomotion.enableJumping,
                browserControls: locomotion.browserControls,
            }).filter(([, v]) => v !== undefined))
            : undefined;
        world.registerSystem(LocomotionSystem, {
            priority: -5,
            configData: locOpts,
        });
    }
    world.registerSystem(InputSystem, {
        priority: -4,
        configData: {
            maintainScenePointers: world.input.canvasPointerEvents.enabled,
        },
    });
    if (world.input.canvasPointerEvents.enabled) {
        world.registerSystem(CanvasPointerSystem, {
            priority: -3.5,
            configData: world.input.canvasPointerEvents,
        });
    }
    if (grabbingEnabled) {
        const grabOpts = typeof grabbing === 'object' && grabbing
            ? Object.fromEntries(Object.entries({
                useHandPinchForGrab: grabbing.useHandPinchForGrab,
            }).filter(([, v]) => v !== undefined))
            : undefined;
        world.registerSystem(GrabSystem, { priority: -3, configData: grabOpts });
    }
    // Physics runs after Grab so it can respect Pressed overrides.
    // Dynamically imported so the physics module (and its ~2 MB Havok WASM)
    // stays out of the static module graph for non-physics projects.
    if (physicsEnabled) {
        const { PhysicsBody, PhysicsManipulation, PhysicsShape, PhysicsSystem } = await import('../physics/index.js');
        world
            .registerComponent(PhysicsBody)
            .registerComponent(PhysicsShape)
            .registerComponent(PhysicsManipulation)
            .registerSystem(PhysicsSystem, { priority: -2 });
    }
    // Scene Understanding updates plane/mesh/anchor debug after input/physics
    if (sceneUnderstandingEnabled) {
        const sceneOpts = typeof sceneUnderstanding === 'object' && sceneUnderstanding
            ? Object.fromEntries(Object.entries({
                showWireFrame: sceneUnderstanding.showWireFrame,
            }).filter(([, v]) => v !== undefined))
            : undefined;
        world
            .registerComponent(XRPlane)
            .registerComponent(XRMesh)
            .registerComponent(XRAnchor)
            .registerSystem(SceneUnderstandingSystem, {
            priority: -1,
            configData: sceneOpts,
        });
    }
    // Environment Raycast system - requires hit-test feature
    if (environmentRaycastEnabled) {
        world
            .registerComponent(EnvironmentRaycastTarget)
            .registerSystem(EnvironmentRaycastSystem, {
            priority: -1,
        });
    }
    // Camera system for video streaming
    if (cameraEnabled) {
        world.registerComponent(CameraSource).registerSystem(CameraSystem);
    }
    // WebXR composition layers (quad/cylinder)
    if (config.xr.enabled && ((_a = config.xr.features) === null || _a === void 0 ? void 0 : _a.layers)) {
        world
            .registerComponent(XRQuadLayer)
            .registerComponent(XRCylinderLayer)
            .registerComponent(XRLayerState)
            .registerSystem(XRLayerSystem, { priority: 1 });
    }
    // Spatial UI systems (Panel, ScreenSpace, Follow)
    if (spatialUIEnabled) {
        const kitsVal = typeof spatialUI === 'object' && spatialUI ? spatialUI.kits : undefined;
        const kitsObj = Array.isArray(kitsVal)
            ? Object.assign({}, ...kitsVal)
            : kitsVal;
        const preferredColorScheme = typeof spatialUI === 'object' && spatialUI
            ? spatialUI.preferredColorScheme
            : undefined;
        world
            .registerComponent(PanelUI)
            .registerComponent(ScreenSpace)
            .registerComponent(Follower)
            .registerSystem(PanelUISystem, {
            // Keep UIKit document updates ahead of screen-space layout publishing.
            priority: -3.8,
            configData: {
                ...(kitsObj ? { kits: kitsObj } : {}),
                ...(preferredColorScheme !== undefined
                    ? { preferredColorScheme }
                    : {}),
            },
        })
            .registerSystem(ScreenSpaceUISystem, {
            // Publish camera-local UI descendants after PanelUISystem updates
            // documents and before CanvasPointerSystem forwards DOM pointer events.
            priority: -3.75,
        })
            .registerSystem(FollowSystem);
    }
}
/**
 * Setup the main render loop
 */
function setupRenderLoop(world, renderer) {
    const clock = new Clock();
    const render = () => {
        var _a, _b;
        const delta = clock.getDelta();
        const elapsedTime = clock.elapsedTime;
        world.visibilityState.value = ((_b = (_a = world.session) === null || _a === void 0 ? void 0 : _a.visibilityState) !== null && _b !== void 0 ? _b : VisibilityState.NonImmersive);
        // Run ECS systems in priority order (InputSystem => LocomotionSystem => GrabSystem)
        world.update(delta, elapsedTime);
        renderer.render(world.scene, world.camera);
    };
    renderer.setAnimationLoop(render);
    // No explicit sessionend handling required on r177; WebXRManager handles
    // render target and canvas sizing restoration internally.
}
/**
 * Setup window resize handling
 */
function setupResizeHandling(camera, renderer) {
    const onWindowResize = () => {
        camera.aspect = window.innerWidth / window.innerHeight;
        camera.updateProjectionMatrix();
        renderer.setSize(window.innerWidth, window.innerHeight);
    };
    window.addEventListener('resize', onWindowResize, false);
}
/**
 * Finalize initialization with asset preloading
 */
function finalizeInitialization(world, assets) {
    return new Promise((resolve, reject) => {
        if (!assets || Object.keys(assets).length === 0) {
            return resolve(world);
        }
        AssetManager.preloadAssets(assets)
            .then(() => resolve(world))
            .catch(reject);
    });
}
/**
 * Setup MCP runtime for framework-specific tools.
 * This creates the MCPRuntime and exposes it on window.FRAMEWORK_MCP_RUNTIME
 * for vite-plugin-dev to route framework-specific tool calls.
 */
function setupMCPRuntime(world) {
    world.mcpRuntime = new MCPRuntime(world);
    // Expose globally for vite-plugin discovery
    // This allows the vite plugin to route framework-specific MCP tools
    // without having a direct dependency on @iwsdk/core
    if (typeof window !== 'undefined') {
        window.FRAMEWORK_MCP_RUNTIME = world.mcpRuntime;
    }
}

export { initializeWorld };
