// @iwsdk/core/dist/input/canvas-pointer-system.js
import { forwardHtmlEvents } from '@pmndrs/pointer-events';
import { Types } from 'elics';
import { createSystem } from '../ecs/system.js';
import '../runtime/three.js';
import { Object3D } from 'three';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
class CanvasPointerSystem extends createSystem({}, {
    /** Forward DOM pointer events from the renderer canvas into the Three scene. */
    enabled: { type: Types.Boolean, default: true },
    /** Continue forwarding pointer events while an immersive XR session is active. */
    activeDuringXR: { type: Types.Boolean, default: false },
}) {
    constructor() {
        super(...arguments);
        this.canvasPointerRoot = new Object3D();
        this.canvasPointerTraversalMarker = new Object3D();
        this.canvasPointerDescendants = [];
    }
    init() {
        // @pmndrs/pointer-events returns early for childless objects before
        // reading interactableDescendants, so keep this synthetic root traversable.
        this.canvasPointerRoot.add(this.canvasPointerTraversalMarker);
        this.cleanupFuncs.push(this.config.enabled.subscribe((enabled) => {
            var _a;
            (_a = this.htmlHandler) === null || _a === void 0 ? void 0 : _a.destroy();
            this.clearCanvasPointerRoot();
            if (enabled) {
                this.prepareCanvasPointerRoot();
                this.htmlHandler = forwardHtmlEvents(this.renderer.domElement, () => this.camera, this.canvasPointerRoot);
            }
            else {
                this.htmlHandler = undefined;
            }
        }));
    }
    update() {
        if (!this.htmlHandler) {
            return;
        }
        if (this.world.session && !this.config.activeDuringXR.value) {
            return;
        }
        this.prepareCanvasPointerRoot();
        this.htmlHandler.update();
    }
    prepareCanvasPointerRoot() {
        var _a;
        const scene = this.scene;
        const root = this.canvasPointerRoot;
        const rayDescendants = (_a = scene.rayDescendants) !== null && _a !== void 0 ? _a : scene.interactableDescendants;
        const screenSpaceDescendants = scene.screenSpaceDescendants;
        this.canvasPointerDescendants.length = 0;
        if (rayDescendants || screenSpaceDescendants) {
            this.pushDescendants(rayDescendants);
            this.pushDescendants(screenSpaceDescendants);
            root.interactableDescendants = this.canvasPointerDescendants;
            return;
        }
        this.canvasPointerDescendants.push(this.scene);
        root.interactableDescendants = this.canvasPointerDescendants;
    }
    pushDescendants(descendants) {
        if (!descendants) {
            return;
        }
        this.canvasPointerDescendants.push(...descendants);
    }
    clearCanvasPointerRoot() {
        const root = this.canvasPointerRoot;
        root.interactableDescendants = undefined;
        this.canvasPointerDescendants.length = 0;
    }
    destroy() {
        var _a;
        super.destroy();
        (_a = this.htmlHandler) === null || _a === void 0 ? void 0 : _a.destroy();
        this.htmlHandler = undefined;
        this.clearCanvasPointerRoot();
        this.canvasPointerRoot.remove(this.canvasPointerTraversalMarker);
    }
}

export { CanvasPointerSystem };
