// @iwsdk/core/dist/input/input-manager.js
import { InputActionManager } from './input-actions.js';
import { StatefulBrowserGamepad } from './stateful-browser-gamepad.js';
import { StatefulKeyboard } from './stateful-keyboard.js';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
function normalizeCanvasPointerEventsOptions(option) {
    var _a, _b;
    if (typeof option === 'boolean') {
        return { enabled: option, activeDuringXR: false };
    }
    return {
        enabled: (_a = option === null || option === void 0 ? void 0 : option.enabled) !== null && _a !== void 0 ? _a : true,
        activeDuringXR: (_b = option === null || option === void 0 ? void 0 : option.activeDuringXR) !== null && _b !== void 0 ? _b : false,
    };
}
class InputManager {
    constructor(xr, options = {}) {
        this.browserGamepads = [];
        this.xr = xr;
        this.keyboard = new StatefulKeyboard();
        this.actions = new InputActionManager();
        this.canvasPointerEvents = normalizeCanvasPointerEventsOptions(options.canvasPointerEvents);
    }
    /** @deprecated Use input.xr.gamepads instead. */
    get gamepads() {
        return this.xr.gamepads;
    }
    /** @deprecated Use input.xr.multiPointers instead. */
    get multiPointers() {
        return this.xr.multiPointers;
    }
    /** @deprecated Use input.xr.visualAdapters instead. */
    get visualAdapters() {
        return this.xr.visualAdapters;
    }
    /** @deprecated Use input.xr.isPrimary(...) instead. */
    isPrimary(...args) {
        return this.xr.isPrimary(...args);
    }
    /** @deprecated Use input.xr.getPrimaryInputSource(...) instead. */
    getPrimaryInputSource(...args) {
        return this.xr.getPrimaryInputSource(...args);
    }
    update(xrManager, delta, time) {
        this.keyboard.update();
        this.updateBrowserGamepads();
        this.xr.update(xrManager, delta, time);
        this.actions.update({
            keyboard: this.keyboard,
            browserGamepads: this.browserGamepads,
            xr: this.xr,
        });
    }
    destroy() {
        this.keyboard.destroy();
    }
    updateBrowserGamepads() {
        if (!navigator.getGamepads) {
            this.browserGamepads.length = 0;
            return;
        }
        const gamepads = navigator.getGamepads();
        this.browserGamepads.length = gamepads.length;
        for (let index = 0; index < gamepads.length; index++) {
            const gamepad = gamepads[index];
            if (!gamepad) {
                this.browserGamepads[index] = undefined;
                continue;
            }
            const state = this.browserGamepads[index];
            if (state) {
                state.refresh(gamepad);
            }
            else {
                const nextState = new StatefulBrowserGamepad(gamepad);
                nextState.refresh(gamepad);
                this.browserGamepads[index] = nextState;
            }
        }
    }
}

export { InputManager, normalizeCanvasPointerEventsOptions };
