// @iwsdk/core/dist/input/input-actions.js
import { InputComponent, AxesState } from '@iwsdk/xr-input';
import { BrowserGamepadAxis, BrowserGamepadButton } from './stateful-browser-gamepad.js';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const InputActions = {
    LocomotionMove: 'locomotion.move',
    LocomotionTurn: 'locomotion.turn',
    LocomotionJump: 'locomotion.jump',
    LocomotionTeleportAim: 'locomotion.teleportAim',
    LocomotionTeleportCommit: 'locomotion.teleportCommit',
    InteractionSelect: 'interaction.select',
};
const DEFAULT_THRESHOLD = 0.8;
function clamp(value, min = -1, max = 1) {
    return Math.min(max, Math.max(min, value));
}
function anyKeyPressed(keyboard, codes) {
    return !!(codes === null || codes === void 0 ? void 0 : codes.some((code) => keyboard.getKeyPressed(code)));
}
class InputActionManager {
    constructor() {
        this.bindings = [];
        this.buttonStates = new Map();
        this.axis1DStates = new Map();
        this.axis2DStates = new Map();
        this.enableDefaultBindings('xr');
    }
    enableDefaultBindings(profile, options = {}) {
        const bindings = this.createDefaultBindings(profile, options);
        for (const binding of bindings) {
            if (!this.bindings.some((existing) => bindingsEqual(existing, binding))) {
                this.bindings.push(binding);
            }
        }
    }
    clearBindings(action) {
        if (!action) {
            this.bindings.length = 0;
            return;
        }
        for (let index = this.bindings.length - 1; index >= 0; index--) {
            if (this.bindings[index].action === action) {
                this.bindings.splice(index, 1);
            }
        }
    }
    addBinding(binding) {
        this.bindings.push(binding);
    }
    addBindings(bindings) {
        this.bindings.push(...bindings);
    }
    removeBinding(binding) {
        const index = this.bindings.findIndex((existing) => bindingsEqual(existing, binding));
        if (index === -1) {
            return false;
        }
        this.bindings.splice(index, 1);
        return true;
    }
    getBindings(action) {
        return this.bindings
            .filter((binding) => !action || binding.action === action)
            .map(cloneBinding);
    }
    update(context) {
        this.prepareFrame();
        for (const binding of this.bindings) {
            if (binding.kind === 'axis2d') {
                this.applyAxis2DBinding(context, binding);
            }
            else if (binding.kind === 'axis1d') {
                this.applyAxis1DBinding(context, binding);
            }
            else {
                this.applyButtonBinding(context, binding);
            }
        }
        for (const state of this.axis1DStates.values()) {
            state.current = clamp(state.current);
        }
        for (const state of this.axis2DStates.values()) {
            state.x = clamp(state.x);
            state.y = clamp(state.y);
        }
    }
    getButtonPressed(action) {
        var _a, _b;
        return (_b = (_a = this.buttonStates.get(action)) === null || _a === void 0 ? void 0 : _a.pressed) !== null && _b !== void 0 ? _b : false;
    }
    getButtonDown(action) {
        var _a, _b;
        return (_b = (_a = this.buttonStates.get(action)) === null || _a === void 0 ? void 0 : _a.down) !== null && _b !== void 0 ? _b : false;
    }
    getButtonUp(action) {
        var _a, _b;
        return (_b = (_a = this.buttonStates.get(action)) === null || _a === void 0 ? void 0 : _a.up) !== null && _b !== void 0 ? _b : false;
    }
    getAxis1D(action) {
        var _a, _b;
        return (_b = (_a = this.axis1DStates.get(action)) === null || _a === void 0 ? void 0 : _a.current) !== null && _b !== void 0 ? _b : 0;
    }
    getAxis1DEnteringPositive(action, threshold = DEFAULT_THRESHOLD) {
        const state = this.axis1DStates.get(action);
        return !!state && state.current >= threshold && state.previous < threshold;
    }
    getAxis1DEnteringNegative(action, threshold = DEFAULT_THRESHOLD) {
        const state = this.axis1DStates.get(action);
        return (!!state && state.current <= -threshold && state.previous > -threshold);
    }
    getAxis2D(action, out = { x: 0, y: 0 }) {
        var _a, _b;
        const state = this.axis2DStates.get(action);
        out.x = (_a = state === null || state === void 0 ? void 0 : state.x) !== null && _a !== void 0 ? _a : 0;
        out.y = (_b = state === null || state === void 0 ? void 0 : state.y) !== null && _b !== void 0 ? _b : 0;
        return out;
    }
    prepareFrame() {
        for (const state of this.buttonStates.values()) {
            state.down = false;
            state.up = false;
            state.pressed = false;
        }
        for (const state of this.axis1DStates.values()) {
            state.previous = state.current;
            state.current = 0;
        }
        for (const state of this.axis2DStates.values()) {
            state.previousX = state.x;
            state.previousY = state.y;
            state.x = 0;
            state.y = 0;
        }
    }
    applyAxis2DBinding(context, binding) {
        const state = this.getAxis2DState(binding.action);
        if (binding.source === 'keyboard') {
            state.x += anyKeyPressed(context.keyboard, binding.positiveX) ? 1 : 0;
            state.x -= anyKeyPressed(context.keyboard, binding.negativeX) ? 1 : 0;
            state.y += anyKeyPressed(context.keyboard, binding.positiveY) ? 1 : 0;
            state.y -= anyKeyPressed(context.keyboard, binding.negativeY) ? 1 : 0;
            return;
        }
        const axisValues = this.getAxesValues(context, binding);
        if (axisValues) {
            state.x += axisValues.x;
            state.y += axisValues.y;
        }
    }
    applyAxis1DBinding(context, binding) {
        const state = this.getAxis1DState(binding.action);
        const axisValues = this.getAxesValues(context, binding);
        if (!axisValues) {
            return;
        }
        state.current += binding.component === 'y' ? axisValues.y : axisValues.x;
    }
    applyButtonBinding(context, binding) {
        const state = this.getButtonState(binding.action);
        const pressed = this.getBindingButtonPressed(context, binding);
        const down = this.getBindingButtonDown(context, binding);
        const up = this.getBindingButtonUp(context, binding);
        state.pressed || (state.pressed = pressed);
        state.down || (state.down = down);
        state.up || (state.up = up);
    }
    getAxesValues(context, binding) {
        var _a;
        if (binding.source === 'browserGamepad') {
            const gamepad = this.getBrowserGamepad(context, binding.gamepadIndex);
            return gamepad === null || gamepad === void 0 ? void 0 : gamepad.getAxesValues(binding.axis);
        }
        if (binding.source === 'xrGamepad') {
            return (_a = context.xr.gamepads[binding.handedness]) === null || _a === void 0 ? void 0 : _a.getAxesValues(binding.componentId);
        }
        return undefined;
    }
    getBindingButtonPressed(context, binding) {
        var _a, _b, _c, _d, _e;
        if (binding.source === 'keyboard') {
            return context.keyboard.getKeyPressed(binding.code);
        }
        if (binding.source === 'browserGamepad') {
            return ((_b = (_a = this.getBrowserGamepad(context, binding.gamepadIndex)) === null || _a === void 0 ? void 0 : _a.getButtonPressed(binding.button)) !== null && _b !== void 0 ? _b : false);
        }
        if (binding.source === 'xrGamepad') {
            return ((_d = (_c = context.xr.gamepads[binding.handedness]) === null || _c === void 0 ? void 0 : _c.getButtonPressed(binding.button)) !== null && _d !== void 0 ? _d : false);
        }
        return (((_e = context.xr.gamepads[binding.handedness]) === null || _e === void 0 ? void 0 : _e.getAxesState(binding.componentId)) === binding.state);
    }
    getBindingButtonDown(context, binding) {
        var _a, _b, _c, _d, _e, _f;
        if (binding.source === 'keyboard') {
            return context.keyboard.getKeyDown(binding.code);
        }
        if (binding.source === 'browserGamepad') {
            return ((_b = (_a = this.getBrowserGamepad(context, binding.gamepadIndex)) === null || _a === void 0 ? void 0 : _a.getButtonDown(binding.button)) !== null && _b !== void 0 ? _b : false);
        }
        if (binding.source === 'xrGamepad') {
            return ((_d = (_c = context.xr.gamepads[binding.handedness]) === null || _c === void 0 ? void 0 : _c.getButtonDown(binding.button)) !== null && _d !== void 0 ? _d : false);
        }
        return ((_f = (_e = context.xr.gamepads[binding.handedness]) === null || _e === void 0 ? void 0 : _e.getAxesEnteringState(binding.componentId, binding.state)) !== null && _f !== void 0 ? _f : false);
    }
    getBindingButtonUp(context, binding) {
        var _a, _b, _c, _d, _e, _f;
        if (binding.source === 'keyboard') {
            return context.keyboard.getKeyUp(binding.code);
        }
        if (binding.source === 'browserGamepad') {
            return ((_b = (_a = this.getBrowserGamepad(context, binding.gamepadIndex)) === null || _a === void 0 ? void 0 : _a.getButtonUp(binding.button)) !== null && _b !== void 0 ? _b : false);
        }
        if (binding.source === 'xrGamepad') {
            return ((_d = (_c = context.xr.gamepads[binding.handedness]) === null || _c === void 0 ? void 0 : _c.getButtonUp(binding.button)) !== null && _d !== void 0 ? _d : false);
        }
        return ((_f = (_e = context.xr.gamepads[binding.handedness]) === null || _e === void 0 ? void 0 : _e.getAxesLeavingState(binding.componentId, binding.state)) !== null && _f !== void 0 ? _f : false);
    }
    getBrowserGamepad(context, gamepadIndex) {
        if (gamepadIndex !== undefined) {
            return context.browserGamepads[gamepadIndex];
        }
        return context.browserGamepads.find((gamepad) => gamepad === null || gamepad === void 0 ? void 0 : gamepad.connected);
    }
    getButtonState(action) {
        let state = this.buttonStates.get(action);
        if (!state) {
            state = { pressed: false, down: false, up: false };
            this.buttonStates.set(action, state);
        }
        return state;
    }
    getAxis1DState(action) {
        let state = this.axis1DStates.get(action);
        if (!state) {
            state = { previous: 0, current: 0 };
            this.axis1DStates.set(action, state);
        }
        return state;
    }
    getAxis2DState(action) {
        let state = this.axis2DStates.get(action);
        if (!state) {
            state = { previousX: 0, previousY: 0, x: 0, y: 0 };
            this.axis2DStates.set(action, state);
        }
        return state;
    }
    createDefaultBindings(profile, options) {
        if (profile === 'browser-first-person') {
            return this.createBrowserFirstPersonBindings(options);
        }
        return [
            {
                source: 'xrGamepad',
                kind: 'axis2d',
                action: InputActions.LocomotionMove,
                handedness: 'left',
                componentId: InputComponent.Thumbstick,
            },
            {
                source: 'xrGamepad',
                kind: 'axis1d',
                action: InputActions.LocomotionTurn,
                handedness: 'right',
                componentId: InputComponent.Thumbstick,
                component: 'x',
            },
            {
                source: 'xrGamepad',
                kind: 'button',
                action: InputActions.LocomotionJump,
                handedness: 'right',
                button: InputComponent.A_Button,
            },
            {
                source: 'xrGamepadAxesDirection',
                kind: 'button',
                action: InputActions.LocomotionTeleportAim,
                handedness: 'right',
                componentId: InputComponent.Thumbstick,
                state: AxesState.Down,
            },
        ];
    }
    createBrowserFirstPersonBindings(options) {
        var _a, _b;
        const keyboardEnabled = (_a = options.keyboard) !== null && _a !== void 0 ? _a : true;
        const gamepadEnabled = (_b = options.gamepad) !== null && _b !== void 0 ? _b : true;
        const bindings = [];
        if (keyboardEnabled) {
            bindings.push({
                source: 'keyboard',
                kind: 'axis2d',
                action: InputActions.LocomotionMove,
                negativeX: ['KeyA', 'ArrowLeft'],
                positiveX: ['KeyD', 'ArrowRight'],
                negativeY: ['KeyW', 'ArrowUp'],
                positiveY: ['KeyS', 'ArrowDown'],
            }, {
                source: 'keyboard',
                kind: 'button',
                action: InputActions.LocomotionJump,
                code: 'Space',
            });
        }
        if (gamepadEnabled) {
            bindings.push({
                source: 'browserGamepad',
                kind: 'axis2d',
                action: InputActions.LocomotionMove,
                axis: BrowserGamepadAxis.LeftStick,
            }, {
                source: 'browserGamepad',
                kind: 'axis1d',
                action: InputActions.LocomotionTurn,
                axis: BrowserGamepadAxis.RightStick,
                component: 'x',
            }, {
                source: 'browserGamepad',
                kind: 'button',
                action: InputActions.LocomotionJump,
                button: BrowserGamepadButton.South,
            });
        }
        return bindings;
    }
}
function cloneBinding(binding) {
    if (binding.source === 'keyboard' && binding.kind === 'axis2d') {
        return {
            ...binding,
            negativeX: binding.negativeX ? [...binding.negativeX] : undefined,
            positiveX: binding.positiveX ? [...binding.positiveX] : undefined,
            negativeY: binding.negativeY ? [...binding.negativeY] : undefined,
            positiveY: binding.positiveY ? [...binding.positiveY] : undefined,
        };
    }
    return { ...binding };
}
function bindingsEqual(first, second) {
    return bindingKey(first) === bindingKey(second);
}
function bindingKey(binding) {
    const entries = Object.entries(binding).map(([key, value]) => [
        key,
        Array.isArray(value) ? [...value].sort() : value,
    ]);
    entries.sort(([firstKey], [secondKey]) => firstKey.localeCompare(secondKey));
    return JSON.stringify(entries);
}

export { InputActionManager, InputActions };
