// @iwsdk/core/dist/input/stateful-keyboard.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
class StatefulKeyboard {
    constructor(target = window) {
        this.pressedKeys = new Set();
        this.currentDownKeys = new Set();
        this.currentUpKeys = new Set();
        this.pendingDownKeys = new Set();
        this.pendingUpKeys = new Set();
        this.handleKeyDown = (event) => {
            if (!this.pressedKeys.has(event.code)) {
                this.pendingDownKeys.add(event.code);
            }
            this.pressedKeys.add(event.code);
        };
        this.handleKeyUp = (event) => {
            if (this.pressedKeys.has(event.code)) {
                this.pendingUpKeys.add(event.code);
            }
            this.pressedKeys.delete(event.code);
        };
        this.target = target;
        this.target.addEventListener('keydown', this.handleKeyDown);
        this.target.addEventListener('keyup', this.handleKeyUp);
    }
    update() {
        this.currentDownKeys.clear();
        this.currentUpKeys.clear();
        for (const code of this.pendingDownKeys) {
            this.currentDownKeys.add(code);
        }
        for (const code of this.pendingUpKeys) {
            this.currentUpKeys.add(code);
        }
        this.pendingDownKeys.clear();
        this.pendingUpKeys.clear();
    }
    getKeyPressed(code) {
        return this.pressedKeys.has(code);
    }
    getKeyDown(code) {
        return this.currentDownKeys.has(code);
    }
    getKeyUp(code) {
        return this.currentUpKeys.has(code);
    }
    destroy() {
        this.target.removeEventListener('keydown', this.handleKeyDown);
        this.target.removeEventListener('keyup', this.handleKeyUp);
        this.pressedKeys.clear();
        this.currentDownKeys.clear();
        this.currentUpKeys.clear();
        this.pendingDownKeys.clear();
        this.pendingUpKeys.clear();
    }
}

export { StatefulKeyboard };
