// @iwsdk/core/dist/input/stateful-browser-gamepad.js
import { StatefulButtonAxesDevice } from '@iwsdk/xr-input';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var BrowserGamepadButton;
(function (BrowserGamepadButton) {
    BrowserGamepadButton["South"] = "standard-south";
    BrowserGamepadButton["East"] = "standard-east";
    BrowserGamepadButton["West"] = "standard-west";
    BrowserGamepadButton["North"] = "standard-north";
    BrowserGamepadButton["LeftBumper"] = "standard-left-bumper";
    BrowserGamepadButton["RightBumper"] = "standard-right-bumper";
    BrowserGamepadButton["LeftTrigger"] = "standard-left-trigger";
    BrowserGamepadButton["RightTrigger"] = "standard-right-trigger";
    BrowserGamepadButton["Select"] = "standard-select";
    BrowserGamepadButton["Start"] = "standard-start";
    BrowserGamepadButton["LeftStick"] = "standard-left-stick";
    BrowserGamepadButton["RightStick"] = "standard-right-stick";
    BrowserGamepadButton["DPadUp"] = "standard-dpad-up";
    BrowserGamepadButton["DPadDown"] = "standard-dpad-down";
    BrowserGamepadButton["DPadLeft"] = "standard-dpad-left";
    BrowserGamepadButton["DPadRight"] = "standard-dpad-right";
    BrowserGamepadButton["Home"] = "standard-home";
})(BrowserGamepadButton || (BrowserGamepadButton = {}));
var BrowserGamepadAxis;
(function (BrowserGamepadAxis) {
    BrowserGamepadAxis["LeftStick"] = "standard-left-stick-axes";
    BrowserGamepadAxis["RightStick"] = "standard-right-stick-axes";
})(BrowserGamepadAxis || (BrowserGamepadAxis = {}));
const STANDARD_GAMEPAD_LAYOUT = {
    selectComponentId: BrowserGamepadButton.South,
    components: {
        [BrowserGamepadButton.South]: {
            type: 'button',
            gamepadIndices: { button: 0 },
        },
        [BrowserGamepadButton.East]: {
            type: 'button',
            gamepadIndices: { button: 1 },
        },
        [BrowserGamepadButton.West]: {
            type: 'button',
            gamepadIndices: { button: 2 },
        },
        [BrowserGamepadButton.North]: {
            type: 'button',
            gamepadIndices: { button: 3 },
        },
        [BrowserGamepadButton.LeftBumper]: {
            type: 'button',
            gamepadIndices: { button: 4 },
        },
        [BrowserGamepadButton.RightBumper]: {
            type: 'button',
            gamepadIndices: { button: 5 },
        },
        [BrowserGamepadButton.LeftTrigger]: {
            type: 'trigger',
            gamepadIndices: { button: 6 },
        },
        [BrowserGamepadButton.RightTrigger]: {
            type: 'trigger',
            gamepadIndices: { button: 7 },
        },
        [BrowserGamepadButton.Select]: {
            type: 'button',
            gamepadIndices: { button: 8 },
        },
        [BrowserGamepadButton.Start]: {
            type: 'button',
            gamepadIndices: { button: 9 },
        },
        [BrowserGamepadButton.LeftStick]: {
            type: 'button',
            gamepadIndices: { button: 10 },
        },
        [BrowserGamepadButton.RightStick]: {
            type: 'button',
            gamepadIndices: { button: 11 },
        },
        [BrowserGamepadButton.DPadUp]: {
            type: 'button',
            gamepadIndices: { button: 12 },
        },
        [BrowserGamepadButton.DPadDown]: {
            type: 'button',
            gamepadIndices: { button: 13 },
        },
        [BrowserGamepadButton.DPadLeft]: {
            type: 'button',
            gamepadIndices: { button: 14 },
        },
        [BrowserGamepadButton.DPadRight]: {
            type: 'button',
            gamepadIndices: { button: 15 },
        },
        [BrowserGamepadButton.Home]: {
            type: 'button',
            gamepadIndices: { button: 16 },
        },
        [BrowserGamepadAxis.LeftStick]: {
            type: 'thumbstick',
            gamepadIndices: { xAxis: 0, yAxis: 1 },
        },
        [BrowserGamepadAxis.RightStick]: {
            type: 'thumbstick',
            gamepadIndices: { xAxis: 2, yAxis: 3 },
        },
    },
};
class StatefulBrowserGamepad extends StatefulButtonAxesDevice {
    constructor(gamepad) {
        super(STANDARD_GAMEPAD_LAYOUT, gamepad.buttons.length);
        this.gamepad = gamepad;
    }
    get index() {
        return this.gamepad.index;
    }
    get id() {
        return this.gamepad.id;
    }
    get connected() {
        return this.gamepad.connected;
    }
    get mapping() {
        return this.gamepad.mapping;
    }
    refresh(gamepad = this.gamepad) {
        this.gamepad = gamepad;
        super.update(gamepad.buttons, gamepad.axes);
    }
}

export { BrowserGamepadAxis, BrowserGamepadButton, StatefulBrowserGamepad };
