// @iwsdk/core/dist/mcp/ecs-debug-tools.js
import { ComponentRegistry } from 'elics';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// ---------------------------------------------------------------------------
// Debug state (module-scoped singleton — only one World per page)
// ---------------------------------------------------------------------------
const debugState = {
    paused: false,
    stepResolve: null,
    stepDelta: 1 / 72,
    frameCount: 0,
    pausedAtFrame: 0,
};
// ---------------------------------------------------------------------------
// Install debug hook — monkey-patches world.update()
// ---------------------------------------------------------------------------
let originalUpdate = null;
let hookedWorld = null;
function installDebugHook(world) {
    if (hookedWorld === world) {
        return;
    } // already installed
    originalUpdate = world.update.bind(world);
    hookedWorld = world;
    world.update = (delta, time) => {
        debugState.frameCount++;
        if (debugState.paused && !debugState.stepResolve) {
            // Skip ECS updates; renderer.render() still runs after this
            return;
        }
        if (debugState.paused && debugState.stepResolve) {
            // Stepping — use fixed timestep
            originalUpdate(debugState.stepDelta, time);
            const resolve = debugState.stepResolve;
            debugState.stepResolve = null;
            resolve();
            return;
        }
        originalUpdate(delta, time);
    };
}
function ecsPause(world) {
    debugState.paused = true;
    debugState.pausedAtFrame = debugState.frameCount;
    return {
        paused: true,
        frame: debugState.frameCount,
        systemCount: world.getSystems().length,
    };
}
let resumeClampInstalled = false;
function ecsResume(world) {
    const framesWhilePaused = debugState.frameCount - debugState.pausedAtFrame;
    debugState.paused = false;
    debugState.stepResolve = null;
    // Cap first real delta after resume to avoid physics explosions
    if (!resumeClampInstalled && originalUpdate) {
        const currentUpdate = world.update;
        let clampNext = true;
        world.update = (delta, time) => {
            if (clampNext) {
                clampNext = false;
                const maxDelta = 1 / 30;
                currentUpdate(Math.min(delta, maxDelta), time);
                // Restore the regular patched update
                world.update = currentUpdate;
                return;
            }
            currentUpdate(delta, time);
        };
        // Only need to do this once per resume
        resumeClampInstalled = true;
        // Reset so the next resume can install again
        setTimeout(() => {
            resumeClampInstalled = false;
        }, 0);
    }
    return {
        paused: false,
        framesWhilePaused,
    };
}
async function ecsStep(_world, params) {
    const { count = 1, delta = 1 / 72 } = params;
    if (!debugState.paused) {
        throw new Error('Cannot step when ECS is not paused. Call ecs_pause first.');
    }
    const frameCount = Math.max(1, Math.min(120, Math.round(count)));
    debugState.stepDelta = delta;
    // Per-step timeout: if the render loop isn't running (tab backgrounded,
    // no active XR session, page hidden), the resolve callback from
    // world.update() never fires. A 5s timeout per step prevents the MCP
    // call from hanging indefinitely.
    const STEP_TIMEOUT_MS = 5000;
    for (let i = 0; i < frameCount; i++) {
        await new Promise((resolve, reject) => {
            debugState.stepResolve = () => {
                clearTimeout(timer);
                resolve();
            };
            const timer = setTimeout(() => {
                debugState.stepResolve = null;
                reject(new Error(`Step timeout after ${STEP_TIMEOUT_MS}ms — the render loop may not be running. ` +
                    `Ensure an XR session is active and the browser tab is visible.`));
            }, STEP_TIMEOUT_MS);
        });
    }
    return {
        framesAdvanced: frameCount,
        totalFrame: debugState.frameCount,
    };
}
// ---------------------------------------------------------------------------
// Component serialization helpers
// ---------------------------------------------------------------------------
const VECTOR_TYPES = new Set(['Vec2', 'Vec3', 'Vec4', 'Color']);
const MAX_STRING_LENGTH = 200;
const MAX_DEPTH = 3;
const MAX_TYPED_ARRAY_ELEMENTS = 16;
function safeSerialize(value, depth = 0, seen = new WeakSet()) {
    if (value === null || value === undefined) {
        return value;
    }
    const type = typeof value;
    if (type === 'number' || type === 'boolean') {
        return value;
    }
    if (type === 'string') {
        return value.length > MAX_STRING_LENGTH
            ? value.slice(0, MAX_STRING_LENGTH) + '...'
            : value;
    }
    if (type === 'function') {
        return '<function>';
    }
    if (type !== 'object') {
        return String(value);
    }
    const obj = value;
    // Circular reference check
    if (seen.has(obj)) {
        return '<circular>';
    }
    seen.add(obj);
    // Depth limit
    if (depth >= MAX_DEPTH) {
        return '<object>';
    }
    // TypedArray
    if (ArrayBuffer.isView(obj) && 'length' in obj) {
        const arr = obj;
        const len = Math.min(arr.length, MAX_TYPED_ARRAY_ELEMENTS);
        const result = Array.from({ length: len }, (_, i) => arr[i]);
        if (arr.length > MAX_TYPED_ARRAY_ELEMENTS) {
            return { values: result, truncated: true, totalLength: arr.length };
        }
        return result;
    }
    // Array
    if (Array.isArray(obj)) {
        return obj
            .slice(0, MAX_TYPED_ARRAY_ELEMENTS)
            .map((item) => safeSerialize(item, depth + 1, seen));
    }
    // Three.js sentinel detection — return type tag instead of full object
    if (obj.isObject3D) {
        return `<Object3D:${obj.name || obj.type || 'unnamed'}>`;
    }
    if (obj.isMaterial) {
        return `<Material:${obj.type || 'unknown'}>`;
    }
    if (obj.isTexture) {
        return `<Texture:${obj.name || 'unnamed'}>`;
    }
    if (obj.isBufferGeometry) {
        return `<BufferGeometry>`;
    }
    // Plain object
    const result = {};
    const keys = Object.keys(obj);
    for (const key of keys.slice(0, 20)) {
        result[key] = safeSerialize(obj[key], depth + 1, seen);
    }
    if (keys.length > 20) {
        result['...'] = `${keys.length - 20} more keys`;
    }
    return result;
}
function serializeComponentValue(entity, component, key, schemaField) {
    var _a;
    const type = schemaField.type;
    if (VECTOR_TYPES.has(type)) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return Array.from(entity.getVectorView(component, key));
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const raw = entity.getValue(component, key);
    if (type === 'Entity') {
        if (raw === null) {
            return null;
        }
        const ref = raw;
        return {
            entityIndex: ref.index,
            name: ((_a = ref.object3D) === null || _a === void 0 ? void 0 : _a.name) || undefined,
        };
    }
    if (type === 'Object') {
        return safeSerialize(raw);
    }
    return raw;
}
function ecsQueryEntity(world, params) {
    var _a;
    const { entityIndex, components } = params;
    if (entityIndex === undefined || entityIndex === null) {
        throw new Error('entityIndex is required. Use get_scene_hierarchy or ecs_find_entities to discover entity indices.');
    }
    const entity = world.entityManager.getEntityByIndex(entityIndex);
    if (!entity) {
        throw new Error(`Entity ${entityIndex} not found or has been destroyed. Use get_scene_hierarchy or ecs_find_entities to find active entities.`);
    }
    const entityComponents = entity.getComponents();
    // Filter to requested components if specified
    let targetComponents = entityComponents;
    if (components && components.length > 0) {
        const filterSet = new Set(components);
        targetComponents = entityComponents.filter((c) => filterSet.has(c.id));
    }
    const componentDataList = targetComponents.map((component) => {
        const values = {};
        const schema = component.schema;
        for (const key of Object.keys(schema)) {
            values[key] = serializeComponentValue(entity, component, key, schema[key]);
        }
        return {
            componentId: component.id,
            description: component.description,
            values,
        };
    });
    return {
        entityIndex: entity.index,
        active: entity.active,
        name: ((_a = entity.object3D) === null || _a === void 0 ? void 0 : _a.name) || undefined,
        components: componentDataList,
    };
}
function ecsFindEntities(world, params) {
    var _a, _b;
    const { withComponents, withoutComponents, namePattern, limit = 50, } = params;
    // Resolve component names to component objects
    const requiredComponents = [];
    if (withComponents) {
        for (const name of withComponents) {
            const comp = ComponentRegistry.getById(name);
            if (!comp) {
                throw new Error(`Component '${name}' not found in registry. Check the component ID.`);
            }
            requiredComponents.push(comp);
        }
    }
    const excludedComponents = [];
    if (withoutComponents) {
        for (const name of withoutComponents) {
            const comp = ComponentRegistry.getById(name);
            if (!comp) {
                throw new Error(`Component '${name}' not found in registry. Check the component ID.`);
            }
            excludedComponents.push(comp);
        }
    }
    let nameRegex = null;
    if (namePattern) {
        nameRegex = new RegExp(namePattern, 'i');
    }
    const maxResults = Math.max(1, Math.min(50, limit));
    const matches = [];
    // indexLookup is internal to ELICS (not part of the public API). If an
    // iteration API becomes public, migrate to that instead. For now, direct
    // access is the only way to enumerate all live entities efficiently.
    const lookup = world.entityManager.indexLookup;
    for (let i = 0; i < lookup.length; i++) {
        const entity = lookup[i];
        if (!entity || !entity.active) {
            continue;
        }
        // Check required components
        let match = true;
        for (const comp of requiredComponents) {
            if (!entity.hasComponent(comp)) {
                match = false;
                break;
            }
        }
        if (!match) {
            continue;
        }
        // Check excluded components
        for (const comp of excludedComponents) {
            if (entity.hasComponent(comp)) {
                match = false;
                break;
            }
        }
        if (!match) {
            continue;
        }
        // Check name pattern
        if (nameRegex) {
            const name = (_a = entity.object3D) === null || _a === void 0 ? void 0 : _a.name;
            if (!name || !nameRegex.test(name)) {
                continue;
            }
        }
        const componentIds = entity.getComponents().map((c) => c.id);
        matches.push({
            entityIndex: entity.index,
            name: ((_b = entity.object3D) === null || _b === void 0 ? void 0 : _b.name) || undefined,
            componentIds,
        });
    }
    const totalMatches = matches.length;
    const results = matches
        .sort((a, b) => getEntityDebugSortRank(a) - getEntityDebugSortRank(b))
        .slice(0, maxResults);
    return {
        entities: results,
        total: totalMatches,
        limited: totalMatches > maxResults,
    };
}
function getEntityDebugSortRank(entity) {
    // Keep infrastructure roots discoverable but list mutable level content first.
    // Tools and agents often pick the first match for quick probes; LevelRoot is
    // intentionally reset to identity every frame by LevelSystem.
    if (entity.componentIds.includes('LevelRoot')) {
        return 1;
    }
    return 0;
}
function ecsListSystems(world) {
    const systems = world.getSystems();
    const result = systems.map((system, index) => {
        var _a, _b, _c;
        const queryNames = Object.keys(system.queries || {});
        const entityCounts = {};
        for (const qName of queryNames) {
            entityCounts[qName] = (_c = (_b = (_a = system.queries[qName]) === null || _a === void 0 ? void 0 : _a.entities) === null || _b === void 0 ? void 0 : _b.size) !== null && _c !== void 0 ? _c : 0;
        }
        return {
            name: system.constructor.name,
            index,
            priority: system.priority,
            isPaused: system.isPaused,
            configKeys: Object.keys(system.config || {}),
            queryNames,
            entityCounts,
        };
    });
    return { systems: result, total: result.length };
}
function ecsListComponents() {
    const all = ComponentRegistry.getAllComponents();
    const result = all.map((comp) => {
        const fields = {};
        for (const [key, field] of Object.entries(comp.schema)) {
            fields[key] = {
                type: field.type,
                default: field.default,
            };
        }
        return {
            id: comp.id,
            description: comp.description,
            fields,
        };
    });
    return { components: result, total: result.length };
}
function ecsToggleSystem(world, params) {
    const { name, paused } = params;
    const systems = world.getSystems();
    const system = systems.find((s) => s.constructor.name === name);
    if (!system) {
        const available = systems.map((s) => s.constructor.name).join(', ');
        throw new Error(`System '${name}' not found. Available systems: ${available}`);
    }
    const targetPaused = paused !== undefined ? paused : !system.isPaused;
    if (targetPaused) {
        system.stop();
    }
    else {
        system.play();
    }
    return { name: system.constructor.name, isPaused: system.isPaused };
}
function ecsSetComponent(world, params) {
    const { entityIndex, componentId, field, value } = params;
    const entity = world.entityManager.getEntityByIndex(entityIndex);
    if (!entity) {
        throw new Error(`Entity ${entityIndex} not found.`);
    }
    const component = ComponentRegistry.getById(componentId);
    if (!component) {
        throw new Error(`Component '${componentId}' not found in registry.`);
    }
    if (!entity.hasComponent(component)) {
        throw new Error(`Entity ${entityIndex} does not have component '${componentId}'.`);
    }
    const schemaField = component.schema[field];
    if (!schemaField) {
        const available = Object.keys(component.schema).join(', ');
        throw new Error(`Field '${field}' not found on '${componentId}'. Available: ${available}`);
    }
    // Read previous value using existing serialization helper
    const previousValue = serializeComponentValue(entity, component, field, schemaField);
    const type = schemaField.type;
    if (VECTOR_TYPES.has(type)) {
        // Vector types: write via getVectorView
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const view = entity.getVectorView(component, field);
        // Value may arrive as a JSON string from MCP clients (e.g. "[0, 1.2, -1.5]")
        let arr = value;
        if (typeof arr === 'string') {
            try {
                arr = JSON.parse(arr);
            }
            catch (_a) {
                throw new Error(`${type} field '${field}': could not parse value string as JSON array.`);
            }
        }
        if (!Array.isArray(arr) || arr.length !== view.length) {
            throw new Error(`${type} field '${field}' requires an array of ${view.length} numbers.`);
        }
        for (let i = 0; i < view.length; i++) {
            view[i] = arr[i];
        }
    }
    else {
        // Scalar types: use setValue (handles validation, query updates)
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        entity.setValue(component, field, value);
    }
    // Read back new value
    const newValue = serializeComponentValue(entity, component, field, schemaField);
    return { entityIndex, componentId, field, previousValue, newValue };
}
// ---------------------------------------------------------------------------
// ecs_snapshot / ecs_diff
// ---------------------------------------------------------------------------
// Module-scoped snapshot storage
const snapshots = new Map();
function ecsSnapshot(world, params) {
    var _a;
    const { label = `snap-${snapshots.size}` } = params;
    // See comment in ecsFindEntities — indexLookup is internal to ELICS.
    const lookup = world.entityManager.indexLookup;
    const entities = [];
    let componentCount = 0;
    for (let i = 0; i < lookup.length; i++) {
        const entity = lookup[i];
        if (!entity || !entity.active) {
            continue;
        }
        const comps = entity.getComponents();
        const components = {};
        for (const comp of comps) {
            const values = {};
            for (const key of Object.keys(comp.schema)) {
                values[key] = serializeComponentValue(entity, comp, key, comp.schema[key]);
            }
            components[comp.id] = values;
            componentCount++;
        }
        entities.push({
            entityIndex: entity.index,
            name: ((_a = entity.object3D) === null || _a === void 0 ? void 0 : _a.name) || undefined,
            components,
        });
    }
    const snapshot = {
        label,
        timestamp: performance.now(),
        frame: debugState.frameCount,
        entities,
    };
    // Evict oldest if at capacity
    if (snapshots.size >= 2 && !snapshots.has(label)) {
        const oldest = snapshots.keys().next().value;
        snapshots.delete(oldest);
    }
    snapshots.set(label, snapshot);
    return {
        label,
        entityCount: entities.length,
        componentCount,
        storedSnapshots: Array.from(snapshots.keys()),
    };
}
function ecsDiff(_world, params) {
    const { from, to } = params;
    const snapFrom = snapshots.get(from);
    const snapTo = snapshots.get(to);
    if (!snapFrom) {
        throw new Error(`Snapshot '${from}' not found. Stored: ${Array.from(snapshots.keys()).join(', ')}`);
    }
    if (!snapTo) {
        throw new Error(`Snapshot '${to}' not found. Stored: ${Array.from(snapshots.keys()).join(', ')}`);
    }
    // Index by entityIndex
    const fromMap = new Map(snapFrom.entities.map((e) => [e.entityIndex, e]));
    const toMap = new Map(snapTo.entities.map((e) => [e.entityIndex, e]));
    const diffs = [];
    let added = 0, removed = 0, changed = 0;
    // Check removed + changed
    for (const [idx, fromEntity] of fromMap) {
        const toEntity = toMap.get(idx);
        if (!toEntity) {
            diffs.push({
                entityIndex: idx,
                name: fromEntity.name,
                status: 'removed',
            });
            removed++;
            continue;
        }
        // Diff components
        const compDiffs = [];
        const allCompIds = new Set([
            ...Object.keys(fromEntity.components),
            ...Object.keys(toEntity.components),
        ]);
        for (const compId of allCompIds) {
            const fromComp = fromEntity.components[compId];
            const toComp = toEntity.components[compId];
            if (!fromComp) {
                compDiffs.push({ componentId: compId, status: 'added' });
                continue;
            }
            if (!toComp) {
                compDiffs.push({ componentId: compId, status: 'removed' });
                continue;
            }
            // Diff fields
            const fieldDiffs = [];
            for (const field of Object.keys(fromComp)) {
                const fv = JSON.stringify(fromComp[field]);
                const tv = JSON.stringify(toComp[field]);
                if (fv !== tv) {
                    fieldDiffs.push({ field, from: fromComp[field], to: toComp[field] });
                }
            }
            // Check for fields added in toComp
            for (const field of Object.keys(toComp)) {
                if (!(field in fromComp)) {
                    fieldDiffs.push({
                        field,
                        from: undefined,
                        to: toComp[field],
                    });
                }
            }
            if (fieldDiffs.length > 0) {
                compDiffs.push({
                    componentId: compId,
                    status: 'changed',
                    fields: fieldDiffs,
                });
            }
        }
        if (compDiffs.length > 0) {
            diffs.push({
                entityIndex: idx,
                name: fromEntity.name,
                status: 'changed',
                components: compDiffs,
            });
            changed++;
        }
    }
    // Check added
    for (const [idx, toEntity] of toMap) {
        if (!fromMap.has(idx)) {
            diffs.push({
                entityIndex: idx,
                name: toEntity.name,
                status: 'added',
            });
            added++;
        }
    }
    return {
        from,
        to,
        frameDelta: snapTo.frame - snapFrom.frame,
        timeDelta: snapTo.timestamp - snapFrom.timestamp,
        entities: diffs,
        summary: { added, removed, changed },
    };
}

export { ecsDiff, ecsFindEntities, ecsListComponents, ecsListSystems, ecsPause, ecsQueryEntity, ecsResume, ecsSetComponent, ecsSnapshot, ecsStep, ecsToggleSystem, installDebugHook };
