// @iwsdk/core/dist/mcp/scene-tools.js
import '../runtime/three.js';
import { Vector3, Quaternion } from 'three';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
/** Default maximum number of children per node before truncation. */
const DEFAULT_MAX_CHILDREN = 50;
/**
 * Get the Three.js scene hierarchy as a JSON tree.
 * Returns object names, UUIDs, and entity indices where available.
 */
function getSceneHierarchy(world, params) {
    const { parentId, maxDepth = 5, maxChildren = DEFAULT_MAX_CHILDREN, } = params;
    let root;
    if (parentId) {
        root = world.scene.getObjectByProperty('uuid', parentId);
        if (!root) {
            throw new Error(`Object not found with UUID '${parentId}'. Use get_scene_hierarchy without parentId to see all available objects.`);
        }
    }
    else {
        root = world.scene;
    }
    return buildHierarchy(root, 0, maxDepth, maxChildren);
}
function buildHierarchy(obj, depth, maxDepth, maxChildren) {
    const node = {
        name: obj.name || '(unnamed)',
        uuid: obj.uuid,
    };
    // Check if Object3D has associated entity (entityIdx is set by Transform component)
    if ('entityIdx' in obj && typeof obj.entityIdx === 'number') {
        node.entityIndex = obj.entityIdx;
    }
    if (depth < maxDepth && obj.children.length > 0) {
        const total = obj.children.length;
        const limit = Math.max(1, maxChildren);
        const childrenToInclude = obj.children.slice(0, limit);
        node.children = childrenToInclude.map((child) => buildHierarchy(child, depth + 1, maxDepth, maxChildren));
        if (total > limit) {
            node.truncatedChildren = total - limit;
        }
    }
    return node;
}
/**
 * Get local and global transforms of an Object3D.
 * Includes positionRelativeToXROrigin which can be used directly with IWER look_at tool.
 */
function getObjectTransform(world, params) {
    const { uuid } = params;
    if (!uuid) {
        throw new Error('uuid parameter is required. Use get_scene_hierarchy to discover object UUIDs.');
    }
    const obj = world.scene.getObjectByProperty('uuid', uuid);
    if (!obj) {
        throw new Error(`Object not found with UUID '${uuid}'. Use get_scene_hierarchy to discover available objects.`);
    }
    // Ensure world matrix is up to date
    obj.updateWorldMatrix(true, false);
    // Local transform (direct properties)
    const localPosition = obj.position.toArray();
    const localQuaternion = obj.quaternion.toArray();
    const localScale = obj.scale.toArray();
    // Global transform (decompose from world matrix)
    const globalPosition = new Vector3();
    const globalQuaternion = new Quaternion();
    const globalScale = new Vector3();
    obj.matrixWorld.decompose(globalPosition, globalQuaternion, globalScale);
    // Position relative to XR origin
    let positionRelativeToXROrigin = null;
    if (world.player) {
        // Clone the global position and convert to XR origin local space
        const relativePos = globalPosition.clone();
        world.player.updateWorldMatrix(true, false);
        world.player.worldToLocal(relativePos);
        positionRelativeToXROrigin = relativePos.toArray();
    }
    return {
        localPosition,
        localQuaternion,
        localScale,
        globalPosition: globalPosition.toArray(),
        globalQuaternion: globalQuaternion.toArray(),
        globalScale: globalScale.toArray(),
        positionRelativeToXROrigin,
    };
}

export { getObjectTransform, getSceneHierarchy };
