// iwer/lib/device/XRDevice.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { GlobalSpace, XRSpace } from '../spaces/XRSpace.js';
import { P_DEVICE, P_REF_SPACE, P_SESSION, P_SPACE, P_SYSTEM, } from '../private.js';
import { Quaternion, Vector3 } from '../utils/Math.js';
import { RemoteControlInterface } from '../remote/RemoteControlInterface.js';
import { XRController } from './XRController.js';
import { XRSession, } from '../session/XRSession.js';
import { XREye, XRView } from '../views/XRView.js';
import { XRHandInput, oculusHandConfig } from './XRHandInput.js';
import { XRHandedness, XRInputSource, XRInputSourceArray, } from '../input/XRInputSource.js';
import { XRLayer, XRWebGLLayer } from '../layers/XRWebGLLayer.js';
import { XRReferenceSpace, XRReferenceSpaceType, } from '../spaces/XRReferenceSpace.js';
import { mat4, vec3 } from 'gl-matrix';
import { ActionPlayer } from '../action/ActionPlayer.js';
import { VERSION } from '../version.js';
import { XRFrame } from '../frameloop/XRFrame.js';
import { XRHand } from '../input/XRHand.js';
import { XRInputSourceEvent } from '../events/XRInputSourceEvent.js';
import { XRInputSourcesChangeEvent } from '../events/XRInputSourcesChangeEvent.js';
import { XRJointPose } from '../pose/XRJointPose.js';
import { XRJointSpace } from '../spaces/XRJointSpace.js';
import { XRPose } from '../pose/XRPose.js';
import { XRReferenceSpaceEvent } from '../events/XRReferenceSpaceEvent.js';
import { XRRenderState } from '../session/XRRenderState.js';
import { XRRigidTransform } from '../primitives/XRRigidTransform.js';
import { XRSessionEvent } from '../events/XRSessionEvent.js';
import { XRSystem } from '../initialization/XRSystem.js';
import { XRViewerPose } from '../pose/XRViewerPose.js';
import { XRViewport } from '../views/XRViewport.js';
import { XRWebGLBinding } from '../depth/XRWebGLBinding.js';
// @ts-ignore
import WebXRLayerPolyfill from 'webxr-layers-polyfill';
const DEFAULTS = {
    ipd: 0.063,
    fovy: Math.PI / 2,
    headsetPosition: new Vector3(0, 1.6, 0),
    headsetQuaternion: new Quaternion(),
    stereoEnabled: false,
};
const Z_INDEX_SEM_CANVAS = 1;
const Z_INDEX_APP_CANVAS = 2;
const Z_INDEX_DEVUI_CANVAS = 3;
const Z_INDEX_DEVUI_CONTAINER = 4;
// scratch buffers for per-frame view offset updates (consumed synchronously
// within updateViews via mat4.fromTranslation, never returned to callers)
const leftViewOffsetScratch = vec3.create();
const rightViewOffsetScratch = vec3.create();
/**
 * XRDevice is not a standard API class outlined in the WebXR Device API Specifications
 * Instead, it serves as an user-facing interface to control the emulated XR Device
 */
export class XRDevice {
    constructor(deviceConfig, deviceOptions = {}) {
        var _a, _b, _c, _d, _e, _f;
        this.version = VERSION;
        const globalSpace = new GlobalSpace();
        const viewerSpace = new XRReferenceSpace(XRReferenceSpaceType.Viewer, globalSpace);
        const viewSpaces = {
            [XREye.Left]: new XRSpace(viewerSpace),
            [XREye.Right]: new XRSpace(viewerSpace),
            [XREye.None]: new XRSpace(viewerSpace),
        };
        const controllerConfig = deviceConfig.controllerConfig;
        const controllers = {};
        if (controllerConfig) {
            Object.values(XRHandedness).forEach((handedness) => {
                if (controllerConfig.layout[handedness]) {
                    controllers[handedness] = new XRController(controllerConfig, handedness, globalSpace);
                }
            });
        }
        const hands = {
            [XRHandedness.Left]: new XRHandInput(oculusHandConfig, XRHandedness.Left, globalSpace),
            [XRHandedness.Right]: new XRHandInput(oculusHandConfig, XRHandedness.Right, globalSpace),
        };
        const canvasContainer = (_a = deviceOptions.canvasContainer) !== null && _a !== void 0 ? _a : document.createElement('div');
        canvasContainer.dataset.webxr_runtime = `Immersive Web Emulation Runtime v${VERSION}`;
        canvasContainer.style.position = 'fixed';
        canvasContainer.style.width = '100%';
        canvasContainer.style.height = '100%';
        canvasContainer.style.top = '0';
        canvasContainer.style.left = '0';
        canvasContainer.style.display = 'flex';
        canvasContainer.style.justifyContent = 'center';
        canvasContainer.style.alignItems = 'center';
        canvasContainer.style.overflow = 'hidden';
        canvasContainer.style.zIndex = '999';
        this[P_DEVICE] = {
            name: deviceConfig.name,
            supportedSessionModes: deviceConfig.supportedSessionModes,
            supportedFeatures: deviceConfig.supportedFeatures,
            supportedFrameRates: deviceConfig.supportedFrameRates,
            isSystemKeyboardSupported: deviceConfig.isSystemKeyboardSupported,
            internalNominalFrameRate: deviceConfig.internalNominalFrameRate,
            environmentBlendModes: deviceConfig.environmentBlendModes,
            interactionMode: deviceConfig.interactionMode,
            userAgent: deviceConfig.userAgent,
            position: (_b = deviceOptions.headsetPosition) !== null && _b !== void 0 ? _b : DEFAULTS.headsetPosition.clone(),
            quaternion: (_c = deviceOptions.headsetQuaternion) !== null && _c !== void 0 ? _c : DEFAULTS.headsetQuaternion.clone(),
            stereoEnabled: (_d = deviceOptions.stereoEnabled) !== null && _d !== void 0 ? _d : DEFAULTS.stereoEnabled,
            ipd: (_e = deviceOptions.ipd) !== null && _e !== void 0 ? _e : DEFAULTS.ipd,
            fovy: (_f = deviceOptions.fovy) !== null && _f !== void 0 ? _f : DEFAULTS.fovy,
            controllers,
            hands,
            primaryInputMode: 'controller',
            pendingReferenceSpaceReset: false,
            visibilityState: 'visible',
            pendingVisibilityState: null,
            xrSystem: null,
            installedGlobalObject: null,
            previousNavigatorXRDescriptor: null,
            previousUserAgentDescriptor: null,
            previousGlobals: null,
            previousMakeXRCompatible: null,
            matrix: mat4.create(),
            globalSpace,
            viewerSpace,
            viewSpaces,
            canvasContainer,
            getViewport: (layer, view) => {
                const canvas = layer.context.canvas;
                const { width, height } = canvas;
                const stereoEnabled = this[P_DEVICE].stereoEnabled;
                const eye = view.eye;
                let x;
                let y;
                let vpWidth;
                let vpHeight;
                switch (eye) {
                    case XREye.None:
                        x = 0;
                        y = 0;
                        vpWidth = width;
                        vpHeight = height;
                        break;
                    case XREye.Left:
                        x = 0;
                        y = 0;
                        vpWidth = stereoEnabled ? width / 2 : width;
                        vpHeight = height;
                        break;
                    case XREye.Right:
                        x = width / 2;
                        y = 0;
                        // zero-width right viewport in mono is deliberate
                        vpWidth = stereoEnabled ? width / 2 : 0;
                        vpHeight = height;
                        break;
                }
                // cache per (eye, width, height, stereoEnabled): return the cached
                // instance unless one of those changed
                const cached = this[P_DEVICE].viewportCache[eye];
                if (cached &&
                    cached.x === x &&
                    cached.y === y &&
                    cached.width === vpWidth &&
                    cached.height === vpHeight) {
                    return cached;
                }
                const viewport = new XRViewport(x, y, vpWidth, vpHeight);
                this[P_DEVICE].viewportCache[eye] = viewport;
                return viewport;
            },
            viewportCache: {},
            updateViews: () => {
                // update viewerSpace
                const viewerSpace = this[P_DEVICE].viewerSpace;
                mat4.fromRotationTranslation(viewerSpace[P_SPACE].offsetMatrix, this[P_DEVICE].quaternion.quat, this[P_DEVICE].position.vec3);
                // update viewSpaces
                vec3.set(leftViewOffsetScratch, -this[P_DEVICE].ipd / 2, 0, 0);
                mat4.fromTranslation(this[P_DEVICE].viewSpaces[XREye.Left][P_SPACE].offsetMatrix, leftViewOffsetScratch);
                vec3.set(rightViewOffsetScratch, this[P_DEVICE].ipd / 2, 0, 0);
                mat4.fromTranslation(this[P_DEVICE].viewSpaces[XREye.Right][P_SPACE].offsetMatrix, rightViewOffsetScratch);
            },
            onBaseLayerSet: (baseLayer) => {
                if (!baseLayer)
                    return;
                // backup canvas data
                const canvas = baseLayer.context.canvas;
                if (canvas.parentElement !== this[P_DEVICE].canvasContainer) {
                    const devui = this[P_DEVICE].devui;
                    if (devui) {
                        const { devUICanvas, devUIContainer } = devui;
                        devUICanvas.style.zIndex = Z_INDEX_DEVUI_CANVAS.toString();
                        devUIContainer.style.zIndex = Z_INDEX_DEVUI_CONTAINER.toString();
                        this[P_DEVICE].canvasContainer.appendChild(devui.devUICanvas);
                        this[P_DEVICE].canvasContainer.appendChild(devui.devUIContainer);
                    }
                    const sem = this[P_DEVICE].sem;
                    if (sem) {
                        sem.environmentCanvas.style.zIndex = Z_INDEX_SEM_CANVAS.toString();
                        this[P_DEVICE].canvasContainer.appendChild(sem.environmentCanvas);
                    }
                    this[P_DEVICE].canvasData = {
                        canvas,
                        parent: canvas.parentElement,
                        width: canvas.width,
                        height: canvas.height,
                        zIndex: canvas.style.zIndex,
                    };
                    canvas.style.zIndex = Z_INDEX_APP_CANVAS.toString();
                    this[P_DEVICE].canvasContainer.appendChild(canvas);
                    document.body.appendChild(this[P_DEVICE].canvasContainer);
                }
                canvas.width = window.innerWidth;
                canvas.height = window.innerHeight;
            },
            onSessionEnd: () => {
                if (this[P_DEVICE].canvasData) {
                    const { canvas, parent, width, height, zIndex } = this[P_DEVICE].canvasData;
                    canvas.width = width;
                    canvas.height = height;
                    canvas.style.zIndex = zIndex;
                    if (parent) {
                        parent.appendChild(canvas);
                    }
                    else {
                        this[P_DEVICE].canvasContainer.removeChild(canvas);
                    }
                    const devui = this[P_DEVICE].devui;
                    if (devui) {
                        this[P_DEVICE].canvasContainer.removeChild(devui.devUICanvas);
                        this[P_DEVICE].canvasContainer.removeChild(devui.devUIContainer);
                    }
                    const sem = this[P_DEVICE].sem;
                    if (sem) {
                        this[P_DEVICE].canvasContainer.removeChild(sem.environmentCanvas);
                    }
                    document.body.removeChild(this[P_DEVICE].canvasContainer);
                    this[P_DEVICE].canvasData = undefined;
                    window.dispatchEvent(new Event('resize'));
                }
            },
            onFrameStart: (frame) => {
                var _a;
                // Calculate delta time for remote control
                const now = performance.now();
                const deltaTimeMs = this[P_DEVICE].lastFrameTime > 0
                    ? now - this[P_DEVICE].lastFrameTime
                    : 16.67; // Default to ~60fps
                this[P_DEVICE].lastFrameTime = now;
                // Update remote control interface
                if (this[P_DEVICE].remote) {
                    this[P_DEVICE].remote.update(deltaTimeMs);
                }
                if ((_a = this[P_DEVICE].actionPlayer) === null || _a === void 0 ? void 0 : _a.playing) {
                    this[P_DEVICE].actionPlayer.playFrame();
                }
                else {
                    const session = frame.session;
                    if (this[P_DEVICE].pendingVisibilityState) {
                        this[P_DEVICE].visibilityState =
                            this[P_DEVICE].pendingVisibilityState;
                        this[P_DEVICE].pendingVisibilityState = null;
                        session.dispatchEvent(new XRSessionEvent('visibilitychange', { session }));
                    }
                    if (this[P_DEVICE].visibilityState === 'visible') {
                        this.activeInputs.forEach((activeInput) => {
                            activeInput.onFrameStart(frame);
                        });
                    }
                    if (this[P_DEVICE].pendingReferenceSpaceReset) {
                        session[P_SESSION].referenceSpaces.forEach((referenceSpace) => {
                            switch (referenceSpace[P_REF_SPACE].type) {
                                case XRReferenceSpaceType.Local:
                                case XRReferenceSpaceType.LocalFloor:
                                case XRReferenceSpaceType.BoundedFloor:
                                case XRReferenceSpaceType.Unbounded:
                                    referenceSpace.dispatchEvent(new XRReferenceSpaceEvent('reset', { referenceSpace }));
                                    break;
                            }
                        });
                        this[P_DEVICE].pendingReferenceSpaceReset = false;
                    }
                }
                this[P_DEVICE].updateViews();
            },
            // control mode for programmatic access
            controlMode: 'manual',
            controlModeListeners: new Set(),
            stateChangeListeners: new Set(),
            // remote control interface - initialized after this object
            remote: null,
            // frame timing for remote update
            lastFrameTime: 0,
        };
        // Initialize remote control interface
        this[P_DEVICE].remote = new RemoteControlInterface(this);
        this[P_DEVICE].updateViews();
        globalThis;
    }
    /**
     * Whether a native (non-emulated) WebXR runtime is already present on
     * `navigator.xr`. installRuntime uses this to avoid clobbering a real
     * runtime unless `forceInstall` is set.
     */
    isNativeXRAvailable() {
        var _a;
        const nativeXR = (_a = globalThis.navigator) === null || _a === void 0 ? void 0 : _a.xr;
        // An emulated runtime installed by IWER is an XRSystem instance; anything
        // else present is treated as a native runtime.
        return Boolean(nativeXR) && !(nativeXR instanceof XRSystem);
    }
    installRuntime(options) {
        var _a, _b, _c, _d;
        const globalObject = ((_a = options === null || options === void 0 ? void 0 : options.globalObject) !== null && _a !== void 0 ? _a : globalThis);
        const polyfillLayers = options === null || options === void 0 ? void 0 : options.polyfillLayers;
        // Skip clobbering a real WebXR runtime unless the caller forces it.
        if (this.isNativeXRAvailable() && !(options === null || options === void 0 ? void 0 : options.forceInstall)) {
            console.warn('IWER: a native WebXR runtime is already available on navigator.xr; ' +
                'skipping installRuntime. Pass { forceInstall: true } to override.');
            return;
        }
        // Track which globals/descriptors we overwrite so uninstallRuntime can
        // best-effort restore them later.
        const previousGlobals = new Map();
        const previousMakeXRCompatible = new Map();
        this[P_DEVICE].installedGlobalObject = globalObject;
        this[P_DEVICE].previousGlobals = previousGlobals;
        this[P_DEVICE].previousMakeXRCompatible = previousMakeXRCompatible;
        const defineMakeXRCompatible = (proto) => {
            if (!proto)
                return;
            previousMakeXRCompatible.set(proto, {
                existed: 'makeXRCompatible' in proto,
                descriptor: Object.getOwnPropertyDescriptor(proto, 'makeXRCompatible'),
            });
            Object.defineProperty(proto, 'makeXRCompatible', {
                value: function () {
                    return new Promise((resolve, _reject) => {
                        resolve(true);
                    });
                },
                configurable: true,
            });
        };
        defineMakeXRCompatible(WebGL2RenderingContext.prototype);
        // Also patch WebGL1 contexts, which can request XR compatibility too.
        defineMakeXRCompatible((_b = globalThis
            .WebGLRenderingContext) === null || _b === void 0 ? void 0 : _b.prototype);
        this[P_DEVICE].xrSystem = new XRSystem(this);
        // Capture the previous navigator.xr / userAgent descriptors before
        // overwriting so uninstallRuntime can restore them.
        this[P_DEVICE].previousNavigatorXRDescriptor =
            (_c = Object.getOwnPropertyDescriptor(globalThis.navigator, 'xr')) !== null && _c !== void 0 ? _c : null;
        this[P_DEVICE].previousUserAgentDescriptor =
            (_d = Object.getOwnPropertyDescriptor(navigator, 'userAgent')) !== null && _d !== void 0 ? _d : null;
        Object.defineProperty(globalThis.navigator, 'xr', {
            value: this[P_DEVICE].xrSystem,
            configurable: true,
        });
        // userAgent is configurable so uninstallRuntime can restore the original.
        Object.defineProperty(navigator, 'userAgent', {
            value: this[P_DEVICE].userAgent,
            writable: false,
            configurable: true,
            enumerable: true,
        });
        const setGlobal = (name, value) => {
            previousGlobals.set(name, {
                existed: name in globalObject,
                value: globalObject[name],
            });
            globalObject[name] = value;
        };
        setGlobal('XRSystem', XRSystem);
        setGlobal('XRSession', XRSession);
        setGlobal('XRRenderState', XRRenderState);
        setGlobal('XRFrame', XRFrame);
        setGlobal('XRSpace', XRSpace);
        setGlobal('XRReferenceSpace', XRReferenceSpace);
        setGlobal('XRJointSpace', XRJointSpace);
        setGlobal('XRView', XRView);
        setGlobal('XRViewport', XRViewport);
        setGlobal('XRRigidTransform', XRRigidTransform);
        setGlobal('XRPose', XRPose);
        setGlobal('XRViewerPose', XRViewerPose);
        setGlobal('XRJointPose', XRJointPose);
        setGlobal('XRInputSource', XRInputSource);
        setGlobal('XRInputSourceArray', XRInputSourceArray);
        setGlobal('XRHand', XRHand);
        setGlobal('XRLayer', XRLayer);
        setGlobal('XRWebGLLayer', XRWebGLLayer);
        setGlobal('XRSessionEvent', XRSessionEvent);
        setGlobal('XRInputSourceEvent', XRInputSourceEvent);
        setGlobal('XRInputSourcesChangeEvent', XRInputSourcesChangeEvent);
        setGlobal('XRReferenceSpaceEvent', XRReferenceSpaceEvent);
        if (polyfillLayers) {
            new WebXRLayerPolyfill();
        }
        else {
            setGlobal('XRMediaBinding', undefined);
        }
        setGlobal('XRWebGLBinding', XRWebGLBinding);
    }
    /**
     * Best-effort reversal of installRuntime: restores the previous
     * navigator.xr / navigator.userAgent descriptors, the WebGL makeXRCompatible
     * methods, and the overwritten global constructors. Safe to call even if
     * installRuntime was never invoked (no-op in that case).
     */
    uninstallRuntime() {
        var _a, _b, _c;
        const previousGlobals = this[P_DEVICE].previousGlobals;
        const globalObject = this[P_DEVICE].installedGlobalObject;
        // Restore overwritten global constructors.
        if (previousGlobals && globalObject) {
            previousGlobals.forEach(({ existed, value }, name) => {
                if (existed) {
                    globalObject[name] = value;
                }
                else {
                    delete globalObject[name];
                }
            });
        }
        // Restore navigator.xr.
        const navXrDescriptor = this[P_DEVICE].previousNavigatorXRDescriptor;
        if (navXrDescriptor) {
            Object.defineProperty(globalThis.navigator, 'xr', navXrDescriptor);
        }
        else if ((_a = Object.getOwnPropertyDescriptor(globalThis.navigator, 'xr')) === null || _a === void 0 ? void 0 : _a.configurable) {
            delete globalThis.navigator.xr;
        }
        // Restore navigator.userAgent (now configurable, so this can succeed).
        const userAgentDescriptor = this[P_DEVICE].previousUserAgentDescriptor;
        if (userAgentDescriptor &&
            ((_b = Object.getOwnPropertyDescriptor(navigator, 'userAgent')) === null || _b === void 0 ? void 0 : _b.configurable)) {
            Object.defineProperty(navigator, 'userAgent', userAgentDescriptor);
        }
        // Restore makeXRCompatible on the WebGL prototypes.
        (_c = this[P_DEVICE].previousMakeXRCompatible) === null || _c === void 0 ? void 0 : _c.forEach(({ existed, descriptor }, proto) => {
            if (existed && descriptor) {
                Object.defineProperty(proto, 'makeXRCompatible', descriptor);
            }
            else {
                delete proto.makeXRCompatible;
            }
        });
        this[P_DEVICE].xrSystem = null;
        this[P_DEVICE].installedGlobalObject = null;
        this[P_DEVICE].previousGlobals = null;
        this[P_DEVICE].previousMakeXRCompatible = null;
        this[P_DEVICE].previousNavigatorXRDescriptor = null;
        this[P_DEVICE].previousUserAgentDescriptor = null;
    }
    installDevUI(devUIConstructor) {
        this[P_DEVICE].devui = new devUIConstructor(this);
    }
    installSEM(semConstructor) {
        this[P_DEVICE].sem = new semConstructor(this);
    }
    get supportedSessionModes() {
        return this[P_DEVICE].supportedSessionModes;
    }
    get supportedFeatures() {
        return this[P_DEVICE].supportedFeatures;
    }
    get supportedFrameRates() {
        return this[P_DEVICE].supportedFrameRates;
    }
    get isSystemKeyboardSupported() {
        return this[P_DEVICE].isSystemKeyboardSupported;
    }
    get internalNominalFrameRate() {
        return this[P_DEVICE].internalNominalFrameRate;
    }
    get stereoEnabled() {
        return this[P_DEVICE].stereoEnabled;
    }
    set stereoEnabled(value) {
        this[P_DEVICE].stereoEnabled = value;
    }
    get ipd() {
        return this[P_DEVICE].ipd;
    }
    set ipd(value) {
        this[P_DEVICE].ipd = value;
    }
    get fovy() {
        return this[P_DEVICE].fovy;
    }
    set fovy(value) {
        this[P_DEVICE].fovy = value;
    }
    get position() {
        return this[P_DEVICE].position;
    }
    get quaternion() {
        return this[P_DEVICE].quaternion;
    }
    get viewerSpace() {
        var _a;
        if ((_a = this[P_DEVICE].actionPlayer) === null || _a === void 0 ? void 0 : _a.playing) {
            return this[P_DEVICE].actionPlayer.viewerSpace;
        }
        else {
            return this[P_DEVICE].viewerSpace;
        }
    }
    get viewSpaces() {
        var _a;
        if ((_a = this[P_DEVICE].actionPlayer) === null || _a === void 0 ? void 0 : _a.playing) {
            return this[P_DEVICE].actionPlayer.viewSpaces;
        }
        else {
            return this[P_DEVICE].viewSpaces;
        }
    }
    get controllers() {
        return this[P_DEVICE].controllers;
    }
    get hands() {
        return this[P_DEVICE].hands;
    }
    get primaryInputMode() {
        return this[P_DEVICE].primaryInputMode;
    }
    set primaryInputMode(mode) {
        if (mode !== 'controller' && mode !== 'hand') {
            console.warn('primary input mode can only be "controller" or "hand"');
            return;
        }
        this[P_DEVICE].primaryInputMode = mode;
    }
    get activeInputs() {
        if (this[P_DEVICE].visibilityState !== 'visible') {
            return [];
        }
        const activeInputs = this[P_DEVICE].primaryInputMode === 'controller'
            ? Object.values(this[P_DEVICE].controllers)
            : Object.values(this[P_DEVICE].hands);
        return activeInputs.filter((input) => input.connected);
    }
    get inputSources() {
        var _a;
        if ((_a = this[P_DEVICE].actionPlayer) === null || _a === void 0 ? void 0 : _a.playing) {
            return this[P_DEVICE].actionPlayer.inputSources;
        }
        else {
            return this.activeInputs.map((input) => input.inputSource);
        }
    }
    get canvasContainer() {
        return this[P_DEVICE].canvasContainer;
    }
    get canvasDimensions() {
        if (this[P_DEVICE].canvasData) {
            const { width, height } = this[P_DEVICE].canvasData.canvas;
            return { width, height };
        }
        return;
    }
    /**
     * Get the app canvas when an XR session is active.
     * Returns undefined if no session is active or no canvas is available.
     */
    get appCanvas() {
        var _a;
        return (_a = this[P_DEVICE].canvasData) === null || _a === void 0 ? void 0 : _a.canvas;
    }
    get activeSession() {
        var _a;
        return (_a = this[P_DEVICE].xrSystem) === null || _a === void 0 ? void 0 : _a[P_SYSTEM].activeSession;
    }
    get sessionOffered() {
        var _a;
        return Boolean((_a = this[P_DEVICE].xrSystem) === null || _a === void 0 ? void 0 : _a[P_SYSTEM].offeredSessionConfig);
    }
    get name() {
        return this[P_DEVICE].name;
    }
    grantOfferedSession() {
        const xrSystem = this[P_DEVICE].xrSystem;
        const pSystem = xrSystem === null || xrSystem === void 0 ? void 0 : xrSystem[P_SYSTEM];
        if (pSystem && pSystem.offeredSessionConfig) {
            const { resolve, reject, mode, options } = pSystem.offeredSessionConfig;
            // Clear the offered session config first
            pSystem.offeredSessionConfig = undefined;
            // Use the same requestSession flow to ensure identical behavior
            xrSystem.requestSession(mode, options).then(resolve).catch(reject);
        }
    }
    recenter() {
        const deltaVec = new Vector3(-this.position.x, 0, -this.position.z);
        const forward = new Vector3(0, 0, -1).applyQuaternion(this.quaternion);
        forward.y = 0;
        forward.normalize();
        const angle = Math.atan2(forward.x, -forward.z);
        const deltaQuat = new Quaternion().setFromAxisAngle(new Vector3(0, 1, 0), angle);
        this.position.add(deltaVec);
        this.quaternion.multiply(deltaQuat);
        [
            ...Object.values(this[P_DEVICE].controllers),
            ...Object.values(this[P_DEVICE].hands),
        ].forEach((activeInput) => {
            activeInput.position.add(deltaVec);
            activeInput.quaternion.multiply(deltaQuat);
            activeInput.position.applyQuaternion(deltaQuat);
        });
        this[P_DEVICE].pendingReferenceSpaceReset = true;
    }
    get visibilityState() {
        return this[P_DEVICE].visibilityState;
    }
    // visibility state updates are queued until the XRSession produces frames
    updateVisibilityState(state) {
        if (!Object.values(['visible', 'visible-blurred', 'hidden']).includes(state)) {
            throw new DOMException('Invalid XRVisibilityState value', 'NotSupportedError');
        }
        if (state !== this[P_DEVICE].visibilityState) {
            this[P_DEVICE].pendingVisibilityState = state;
        }
    }
    createActionPlayer(refSpace, recording) {
        this[P_DEVICE].actionPlayer = new ActionPlayer(refSpace, recording, this[P_DEVICE].ipd);
        return this[P_DEVICE].actionPlayer;
    }
    get devui() {
        return this[P_DEVICE].devui;
    }
    get sem() {
        return this[P_DEVICE].sem;
    }
    get remote() {
        return this[P_DEVICE].remote;
    }
    // =============================================================================
    // Control Mode API
    // =============================================================================
    /**
     * Get the current control mode
     * - 'manual': User controls device via DevUI (default)
     * - 'programmatic': External API controls device
     */
    get controlMode() {
        return this[P_DEVICE].controlMode;
    }
    /**
     * Set the control mode
     * Notifies all registered listeners of the change
     */
    set controlMode(mode) {
        if (mode !== 'manual' && mode !== 'programmatic') {
            console.warn('control mode can only be "manual" or "programmatic"');
            return;
        }
        const prevMode = this[P_DEVICE].controlMode;
        if (prevMode !== mode) {
            this[P_DEVICE].controlMode = mode;
            this[P_DEVICE].controlModeListeners.forEach((listener) => listener(mode));
        }
    }
    /**
     * Register a listener to be notified when control mode changes
     * @param listener - Callback function that receives the new mode
     * @returns Unsubscribe function to remove the listener
     */
    onControlModeChange(listener) {
        this[P_DEVICE].controlModeListeners.add(listener);
        return () => {
            this[P_DEVICE].controlModeListeners.delete(listener);
        };
    }
    /**
     * Register a listener to be notified when device state changes
     * Called after programmatic state modifications
     * @param listener - Callback function
     * @returns Unsubscribe function to remove the listener
     */
    onStateChange(listener) {
        this[P_DEVICE].stateChangeListeners.add(listener);
        return () => {
            this[P_DEVICE].stateChangeListeners.delete(listener);
        };
    }
    /**
     * Notify all state change listeners that device state has been modified
     * Should be called after programmatic state modifications
     */
    notifyStateChange() {
        this[P_DEVICE].stateChangeListeners.forEach((listener) => listener());
    }
}
